{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:56:49.989678Z",
     "iopub.status.busy": "2024-12-02T20:56:49.989256Z",
     "iopub.status.idle": "2024-12-02T20:56:51.967793Z",
     "shell.execute_reply": "2024-12-02T20:56:51.965954Z"
    }
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import pandas as pd\n",
    "import re\n",
    "import glob\n",
    "import os\n",
    "import statsmodels.api as sm\n",
    "\n",
    "PREFIX = 'calculate-functional-enrichments-set-level-logistic-regression.'\n",
    "\n",
    "\n",
    "topmed_to_roadmap = {\n",
    "    'Whole_blood': 'E062',\n",
    "    'PBMC': 'E062',\n",
    "    'Lung': 'E096',\n",
    "    'Monocyte': 'E029',\n",
    "    'T_cell': 'E034',\n",
    "    'Nasal_epithelial': 'E114'\n",
    "}\n",
    "\n",
    "roadmap_state_names = {\n",
    "    '1_TssA': 'Active_TSS',\n",
    "    '2_TssAFlnk': 'Flanking Active TSS',\n",
    "    '3_TxFlnk':\t\"Transcr. at gene 5' and 3'\",\n",
    "    '4_Tx':\t'Strong transcription',\n",
    "    '5_TxWk': 'Weak transcription',\n",
    "    '6_EnhG': 'Genic enhancers',\n",
    "    '7_Enh': 'Enhancers',\n",
    "    '8_ZNF/Rpts': 'ZNF genes & repeats',\n",
    "    '9_Het': 'Heterochromatin',\n",
    "    '10_TssBiv': 'Bivalent/Poised TSS',\n",
    "    '11_BivFlnk': 'Flanking Bivalent TSS/Enh',\n",
    "    '12_EnhBiv': 'Bivalent Enhancer',\n",
    "    '13_ReprPC': 'Repressed PolyComb',\n",
    "    '14_ReprPCWk': 'Weak Repressed PolyComb',\n",
    "    '15_Quies': 'Quiescent/Low'\n",
    "}\n",
    "\n",
    "def format_modality(modality):\n",
    "    modality_lower = modality.lower()\n",
    "    assert('trans' in modality_lower or 'cis' in modality_lower)\n",
    "    cis_trans = 'trans' if 'trans' in modality_lower else 'cis'\n",
    "    assert('eqtl' in modality_lower or 'sqtl' in modality_lower)\n",
    "    e_sqtl = 'eQTL' if 'eqtl' in modality_lower else 'sQTL'\n",
    "    return cis_trans + '-' + e_sqtl\n",
    "\n",
    "\n",
    "def parse_modality(s):\n",
    "    RE = '((cis|trans)-?([es]qtl))'\n",
    "    m = re.search(RE, s, flags=re.IGNORECASE)\n",
    "    assert(m is not None)\n",
    "    return m.group(0)\n",
    "\n",
    "\n",
    "def parse_tissue(s):\n",
    "    tissues = ['Whole_blood', 'Lung', 'Nasal_epithelial', 'PBMC', 'Monocyte', 'T_cell']\n",
    "    RE = '(' + '|'.join(tissues) + ')'\n",
    "    m = re.search(RE, s)\n",
    "    assert(m is not None)\n",
    "    return m.group(0)\n",
    "\n",
    "\n",
    "def parse_maf(s):\n",
    "    mafs = ['maf001', 'maf0001', 'maf005']\n",
    "    RE = '(' + '|'.join(mafs) + ')'\n",
    "    m = re.search(RE, s)\n",
    "    assert(m is not None)\n",
    "    return m.group(0)\n",
    "\n",
    "\n",
    "def format_maf(s):\n",
    "    x = {\n",
    "        'maf001': '1%',\n",
    "        'maf0001': '0.1%',\n",
    "        'maf005': '5%'\n",
    "         }\n",
    "    if s not in x:\n",
    "        raise ValueError(\"Cannot format MAF: {}\".format(s))\n",
    "    return x[s]\n",
    "\n",
    "\n",
    "def variant_annotation_to_credible_set_annotation(credible_sets, annotation_table):\n",
    "    tmp = annotation_table.loc[credible_sets.variant_id.unique()].reset_index().rename(columns={'index': 'variant_id'})\n",
    "    tmp = credible_sets[['phenotype_id', 'variant_id', 'cs_id']].merge(tmp, on='variant_id').drop(columns=['variant_id'])\n",
    "    return tmp.groupby(['phenotype_id', 'cs_id']).max()\n",
    "\n",
    "\n",
    "def logistic_regression(credible_sets, control_credible_sets, annotation_table):\n",
    "    x_hit = variant_annotation_to_credible_set_annotation(credible_sets, annotation_table)\n",
    "    x_control = variant_annotation_to_credible_set_annotation(control_credible_sets, annotation_table)\n",
    "    x = pd.concat([x_hit.assign(observed=1), x_control.assign(observed=0)]).reset_index(drop=True)\n",
    "    y = x.pop('observed')\n",
    "\n",
    "    # drop annotations w/ no variance\n",
    "    DROP_ANNOTATIONS = x.nunique().where(lambda y: y==1).dropna().index.to_list()\n",
    "\n",
    "    # drop annotations w/ few overlaps (eliminates some singular matrix issues)\n",
    "    # DROP_ANNOTATIONS += x.sum().where(lambda y: y<5).dropna().index.to_list()\n",
    "    threshold = len(x) * 0.01 # at least 1% of true + control credible sets need to overlap the annotation\n",
    "    DROP_ANNOTATIONS += x.sum().where(lambda y: y<threshold).dropna().index.to_list()\n",
    "\n",
    "    # drop annotations w/ no overlaps in either case or control\n",
    "    # DROP_ANNOTATIONS += x.groupby(y).mean().min().where(lambda y: y<=0).dropna().index.to_list()\n",
    "    x = x.drop(columns=list(set(DROP_ANNOTATIONS)))\n",
    "\n",
    "    logit_model = sm.Logit(y, sm.add_constant(x))\n",
    "    result = logit_model.fit(disp=0)\n",
    "    coefs = result.params\n",
    "    pvalues = result.pvalues\n",
    "    converged = result.mle_retvals['converged']\n",
    "    assert(all(coefs.index == pvalues.index))\n",
    "\n",
    "    tmp = result.summary2().tables[1]\n",
    "    tmp['converged'] = converged\n",
    "    tmp = tmp[tmp.index!='const']\n",
    "\n",
    "    hits_in_annotation = x_hit.sum()\n",
    "    hits_not_in_annotation = (x_hit==0).sum()\n",
    "    controls_in_annotation = x_control.sum()\n",
    "    controls_not_in_annotation = (x_control==0).sum()\n",
    "    tmp['n_hits_in_annotation'] = hits_in_annotation.loc[tmp.index].to_list()\n",
    "    tmp['n_hits_not_in_annotation'] = hits_not_in_annotation.loc[tmp.index].to_list()\n",
    "    tmp['n_controls_in_annotation'] = controls_in_annotation.loc[tmp.index].to_list()\n",
    "    tmp['n_controls_not_in_annotation'] = controls_not_in_annotation.loc[tmp.index].to_list()\n",
    "\n",
    "    return tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:56:51.971802Z",
     "iopub.status.busy": "2024-12-02T20:56:51.971065Z",
     "iopub.status.idle": "2024-12-02T20:57:45.206344Z",
     "shell.execute_reply": "2024-12-02T20:57:45.205356Z"
    }
   },
   "outputs": [],
   "source": [
    "ANNOTATION_FILES = glob.glob('../work/variant-annotation-matrices/annotations.*.txt')\n",
    "annotations = {os.path.basename(f).split('.')[1]: pd.read_csv(f, sep='\\t', index_col=0) for f in ANNOTATION_FILES}\n",
    "\n",
    "# rename Roadmap states\n",
    "for a in [i for i in annotations.keys() if 'roadmap' in i]:\n",
    "    annotations[a] = annotations[a].rename(columns=roadmap_state_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:57:45.212854Z",
     "iopub.status.busy": "2024-12-02T20:57:45.212120Z",
     "iopub.status.idle": "2024-12-02T20:57:45.216229Z",
     "shell.execute_reply": "2024-12-02T20:57:45.215635Z"
    }
   },
   "outputs": [],
   "source": [
    "# del annotations['regulatory_build']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:57:45.219318Z",
     "iopub.status.busy": "2024-12-02T20:57:45.218791Z",
     "iopub.status.idle": "2024-12-02T20:58:03.217806Z",
     "shell.execute_reply": "2024-12-02T20:58:03.216866Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>maf</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>modality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000168350.8</td>\n",
       "      <td>chr14_69324247_T_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.32150</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000100711.13</td>\n",
       "      <td>chr14_76492603_C_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.05732</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000258913.1</td>\n",
       "      <td>chr14_104810307_C_T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03292</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000179627.10</td>\n",
       "      <td>chr14_20870069_A_G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.08056</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000211923.1</td>\n",
       "      <td>chr14_105495192_G_A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.08211</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id           variant_id  pip      maf  cs_id tissue  \\\n",
       "0   ENSG00000168350.8   chr14_69324247_T_G  1.0  0.32150      2   Lung   \n",
       "1  ENSG00000100711.13   chr14_76492603_C_G  1.0  0.05732      1   Lung   \n",
       "2   ENSG00000258913.1  chr14_104810307_C_T  1.0  0.03292      1   Lung   \n",
       "3  ENSG00000179627.10   chr14_20870069_A_G  1.0  0.08056      1   Lung   \n",
       "4   ENSG00000211923.1  chr14_105495192_G_A  1.0  0.08211      1   Lung   \n",
       "\n",
       "  maf_threshold  modality  \n",
       "0            1%  cis-eQTL  \n",
       "1            1%  cis-eQTL  \n",
       "2            1%  cis-eQTL  \n",
       "3            1%  cis-eQTL  \n",
       "4            1%  cis-eQTL  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_credible_sets = glob.glob('../work/control-credible-sets/cis-*/maf*/results/controls/*.control-credible-sets.tsv')\n",
    "control_credible_sets = pd.concat([pd.read_csv(f, sep='\\t').assign(f=f) for f in control_credible_sets])\n",
    "control_credible_sets['tissue'] = control_credible_sets.f.map(lambda x: os.path.basename(x).split('.')[0])\n",
    "control_credible_sets['maf_threshold'] = control_credible_sets.f.map(lambda x: re.search('(maf\\d+)', x).group(0))\n",
    "control_credible_sets['modality'] = control_credible_sets.f.map(lambda x: re.search('(cis-[a-z]+)', x).group(0))\n",
    "control_credible_sets = control_credible_sets.drop(columns=['f'])\n",
    "control_credible_sets.modality = control_credible_sets.modality.map(format_modality)\n",
    "control_credible_sets.maf_threshold = control_credible_sets.maf_threshold.map({'maf0001': '0.1%', 'maf001': '1%', 'maf005': '5%'})\n",
    "control_credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:03.254396Z",
     "iopub.status.busy": "2024-12-02T20:58:03.253928Z",
     "iopub.status.idle": "2024-12-02T20:58:03.281262Z",
     "shell.execute_reply": "2024-12-02T20:58:03.280445Z"
    }
   },
   "outputs": [],
   "source": [
    "to_load = glob.glob('../../manuscript-intermediate-processing/data/scan-results/joint/cis-eqtl/susie/maf*/*.cs.txt') + glob.glob('../../manuscript-intermediate-processing/data/scan-results/joint/cis-sqtl/susie/maf*/postprocessed/*.by-gene.cs.txt')\n",
    "to_load = pd.DataFrame({'file': to_load})\n",
    "to_load['tissue'] = to_load.file.apply(parse_tissue)\n",
    "to_load['modality'] = to_load.file.apply(parse_modality).map(format_modality)\n",
    "to_load['maf'] = to_load.file.apply(parse_maf).map(format_maf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:03.286578Z",
     "iopub.status.busy": "2024-12-02T20:58:03.285757Z",
     "iopub.status.idle": "2024-12-02T20:58:08.774442Z",
     "shell.execute_reply": "2024-12-02T20:58:08.773922Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold  \n",
       "0  cis-eQTL            1%  \n",
       "1  cis-eQTL            1%  \n",
       "2  cis-eQTL            1%  \n",
       "3  cis-eQTL            1%  \n",
       "4  cis-eQTL            1%  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "credible_sets = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=tissue, modality=modality, maf_threshold=maf_threshold) for f, tissue, modality, maf_threshold in zip(to_load.file, to_load.tissue, to_load.modality, to_load.maf)])\n",
    "credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:08.776680Z",
     "iopub.status.busy": "2024-12-02T20:58:08.776476Z",
     "iopub.status.idle": "2024-12-02T20:58:17.428270Z",
     "shell.execute_reply": "2024-12-02T20:58:17.427604Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>has_control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold  has_control  \n",
       "0  cis-eQTL            1%            1  \n",
       "1  cis-eQTL            1%            1  \n",
       "2  cis-eQTL            1%            1  \n",
       "3  cis-eQTL            1%            1  \n",
       "4  cis-eQTL            1%            1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove credible sets without a matching control credible set\n",
    "credible_sets = credible_sets.merge(control_credible_sets[['phenotype_id', 'cs_id', 'tissue', 'modality', 'maf_threshold']].assign(has_control=1).drop_duplicates(), how='left')\n",
    "credible_sets['has_control'] = credible_sets.has_control.fillna(0).astype(int)\n",
    "credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:17.433948Z",
     "iopub.status.busy": "2024-12-02T20:58:17.432896Z",
     "iopub.status.idle": "2024-12-02T20:58:21.265950Z",
     "shell.execute_reply": "2024-12-02T20:58:21.265486Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    364693\n",
       "0       476\n",
       "Name: has_control, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "credible_sets[['phenotype_id', 'cs_id', 'tissue', 'modality', 'maf_threshold', 'has_control']].drop_duplicates().has_control.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:21.268994Z",
     "iopub.status.busy": "2024-12-02T20:58:21.268361Z",
     "iopub.status.idle": "2024-12-02T20:58:22.408696Z",
     "shell.execute_reply": "2024-12-02T20:58:22.408009Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold  \n",
       "0  cis-eQTL            1%  \n",
       "1  cis-eQTL            1%  \n",
       "2  cis-eQTL            1%  \n",
       "3  cis-eQTL            1%  \n",
       "4  cis-eQTL            1%  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "credible_sets = credible_sets[credible_sets.has_control==1].drop(columns=['has_control'])\n",
    "credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:22.411332Z",
     "iopub.status.busy": "2024-12-02T20:58:22.410953Z",
     "iopub.status.idle": "2024-12-02T20:58:22.456104Z",
     "shell.execute_reply": "2024-12-02T20:58:22.455403Z"
    }
   },
   "outputs": [],
   "source": [
    "# add trans-e/sQTL top hits, reshaping to resemble credible sets\n",
    "input_variant_files = glob.glob('../work/control-snps/trans-*qtl/maf005/results/controls/*.controls.txt')\n",
    "RE = re.compile('.*/work/control-snps/(.*)/(maf\\d+)/results/controls/(.*).controls.txt')\n",
    "input_variants = pd.concat([pd.read_csv(f, sep='\\t').assign(modality=RE.match(f).group(1), tissue=RE.match(f).group(3), maf=RE.match(f).group(2)) for f in input_variant_files])\n",
    "input_variants.maf = input_variants.maf.map({'maf001': '1%', 'maf0001': '0.1%', 'maf005': '5%'})\n",
    "input_variants.modality = input_variants.modality.map(format_modality)\n",
    "input_variants = input_variants.groupby(['variant', 'modality', 'tissue', 'maf']).head(1)\n",
    "\n",
    "input_variants['phenotype_id'] = [f'fake_{i}' for i in range(1, len(input_variants)+ 1)]\n",
    "input_variants['cs_id'] = 1\n",
    "input_variants['pip'] = 1.0\n",
    "trans_credible_sets = input_variants[['phenotype_id', 'variant', 'pip', 'variant_maf', 'cs_id', 'tissue', 'modality', 'maf']].rename(columns={'variant': 'variant_id', 'variant_maf': 'af', 'maf': 'maf_threshold'})\n",
    "trans_control_credible_sets = input_variants[['phenotype_id', 'control_variant', 'pip', 'control_variant_maf', 'cs_id', 'tissue', 'modality', 'maf']].rename(columns={'control_variant': 'variant_id', 'control_variant_maf': 'maf', 'maf': 'maf_threshold'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:22.458671Z",
     "iopub.status.busy": "2024-12-02T20:58:22.458242Z",
     "iopub.status.idle": "2024-12-02T20:58:22.469139Z",
     "shell.execute_reply": "2024-12-02T20:58:22.468428Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>control_variant</th>\n",
       "      <th>variant_maf</th>\n",
       "      <th>control_variant_maf</th>\n",
       "      <th>modality</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>pip</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19_37475759_A_G</td>\n",
       "      <td>chr19_40989528_A_G</td>\n",
       "      <td>0.09328</td>\n",
       "      <td>0.09209</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>5%</td>\n",
       "      <td>fake_1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr19_16325451_T_C</td>\n",
       "      <td>chr19_43696515_C_A</td>\n",
       "      <td>0.17980</td>\n",
       "      <td>0.18260</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>5%</td>\n",
       "      <td>fake_2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19_36330296_T_G</td>\n",
       "      <td>chr19_2894839_C_T</td>\n",
       "      <td>0.18460</td>\n",
       "      <td>0.18180</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>5%</td>\n",
       "      <td>fake_3</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr19_6077048_A_G</td>\n",
       "      <td>chr19_28418435_C_A</td>\n",
       "      <td>0.25490</td>\n",
       "      <td>0.25490</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>5%</td>\n",
       "      <td>fake_4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr19_12477305_T_C</td>\n",
       "      <td>chr19_20056962_A_G</td>\n",
       "      <td>0.26640</td>\n",
       "      <td>0.26760</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>5%</td>\n",
       "      <td>fake_5</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               variant     control_variant  variant_maf  control_variant_maf  \\\n",
       "0   chr19_37475759_A_G  chr19_40989528_A_G      0.09328              0.09209   \n",
       "5   chr19_16325451_T_C  chr19_43696515_C_A      0.17980              0.18260   \n",
       "10  chr19_36330296_T_G   chr19_2894839_C_T      0.18460              0.18180   \n",
       "15   chr19_6077048_A_G  chr19_28418435_C_A      0.25490              0.25490   \n",
       "20  chr19_12477305_T_C  chr19_20056962_A_G      0.26640              0.26760   \n",
       "\n",
       "      modality tissue maf phenotype_id  cs_id  pip  \n",
       "0   trans-eQTL   PBMC  5%       fake_1      1  1.0  \n",
       "5   trans-eQTL   PBMC  5%       fake_2      1  1.0  \n",
       "10  trans-eQTL   PBMC  5%       fake_3      1  1.0  \n",
       "15  trans-eQTL   PBMC  5%       fake_4      1  1.0  \n",
       "20  trans-eQTL   PBMC  5%       fake_5      1  1.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_variants.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:22.471737Z",
     "iopub.status.busy": "2024-12-02T20:58:22.471326Z",
     "iopub.status.idle": "2024-12-02T20:58:24.165803Z",
     "shell.execute_reply": "2024-12-02T20:58:24.165072Z"
    }
   },
   "outputs": [],
   "source": [
    "credible_sets = pd.concat([credible_sets, trans_credible_sets])\n",
    "control_credible_sets = pd.concat([control_credible_sets, trans_control_credible_sets])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:24.168988Z",
     "iopub.status.busy": "2024-12-02T20:58:24.168610Z",
     "iopub.status.idle": "2024-12-02T20:58:24.177110Z",
     "shell.execute_reply": "2024-12-02T20:58:24.176673Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold  \n",
       "0  cis-eQTL            1%  \n",
       "1  cis-eQTL            1%  \n",
       "2  cis-eQTL            1%  \n",
       "3  cis-eQTL            1%  \n",
       "4  cis-eQTL            1%  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T20:58:24.179198Z",
     "iopub.status.busy": "2024-12-02T20:58:24.178916Z",
     "iopub.status.idle": "2024-12-02T21:00:24.117316Z",
     "shell.execute_reply": "2024-12-02T21:00:24.116424Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "      <th>converged</th>\n",
       "      <th>n_hits_in_annotation</th>\n",
       "      <th>n_hits_not_in_annotation</th>\n",
       "      <th>n_controls_in_annotation</th>\n",
       "      <th>n_controls_not_in_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequence_feature</th>\n",
       "      <td>-0.078589</td>\n",
       "      <td>0.048185</td>\n",
       "      <td>-1.630979</td>\n",
       "      <td>1.028947e-01</td>\n",
       "      <td>-0.173030</td>\n",
       "      <td>0.015852</td>\n",
       "      <td>True</td>\n",
       "      <td>1335</td>\n",
       "      <td>23744</td>\n",
       "      <td>838</td>\n",
       "      <td>24241</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <td>0.648394</td>\n",
       "      <td>0.039557</td>\n",
       "      <td>16.391570</td>\n",
       "      <td>2.196965e-60</td>\n",
       "      <td>0.570864</td>\n",
       "      <td>0.725923</td>\n",
       "      <td>True</td>\n",
       "      <td>3798</td>\n",
       "      <td>21281</td>\n",
       "      <td>1315</td>\n",
       "      <td>23764</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <td>0.326662</td>\n",
       "      <td>0.028733</td>\n",
       "      <td>11.369052</td>\n",
       "      <td>5.963155e-30</td>\n",
       "      <td>0.270347</td>\n",
       "      <td>0.382976</td>\n",
       "      <td>True</td>\n",
       "      <td>6034</td>\n",
       "      <td>19045</td>\n",
       "      <td>2990</td>\n",
       "      <td>22089</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_variant</th>\n",
       "      <td>0.166290</td>\n",
       "      <td>0.027745</td>\n",
       "      <td>5.993512</td>\n",
       "      <td>2.053564e-09</td>\n",
       "      <td>0.111911</td>\n",
       "      <td>0.220670</td>\n",
       "      <td>True</td>\n",
       "      <td>20532</td>\n",
       "      <td>4547</td>\n",
       "      <td>17018</td>\n",
       "      <td>8061</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <td>0.272838</td>\n",
       "      <td>0.022563</td>\n",
       "      <td>12.092274</td>\n",
       "      <td>1.160295e-33</td>\n",
       "      <td>0.228615</td>\n",
       "      <td>0.317060</td>\n",
       "      <td>True</td>\n",
       "      <td>16615</td>\n",
       "      <td>8464</td>\n",
       "      <td>11696</td>\n",
       "      <td>13383</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Coef.  Std.Err.          z         P>|z|  \\\n",
       "sequence_feature        -0.078589  0.048185  -1.630979  1.028947e-01   \n",
       "5_prime_UTR_variant      0.648394  0.039557  16.391570  2.196965e-60   \n",
       "3_prime_UTR_variant      0.326662  0.028733  11.369052  5.963155e-30   \n",
       "intron_variant           0.166290  0.027745   5.993512  2.053564e-09   \n",
       "downstream_gene_variant  0.272838  0.022563  12.092274  1.160295e-33   \n",
       "\n",
       "                           [0.025    0.975]  converged  n_hits_in_annotation  \\\n",
       "sequence_feature        -0.173030  0.015852       True                  1335   \n",
       "5_prime_UTR_variant      0.570864  0.725923       True                  3798   \n",
       "3_prime_UTR_variant      0.270347  0.382976       True                  6034   \n",
       "intron_variant           0.111911  0.220670       True                 20532   \n",
       "downstream_gene_variant  0.228615  0.317060       True                 16615   \n",
       "\n",
       "                         n_hits_not_in_annotation  n_controls_in_annotation  \\\n",
       "sequence_feature                            23744                       838   \n",
       "5_prime_UTR_variant                         21281                      1315   \n",
       "3_prime_UTR_variant                         19045                      2990   \n",
       "intron_variant                               4547                     17018   \n",
       "downstream_gene_variant                      8464                     11696   \n",
       "\n",
       "                         n_controls_not_in_annotation tissue  modality  \\\n",
       "sequence_feature                                24241   Lung  cis-eQTL   \n",
       "5_prime_UTR_variant                             23764   Lung  cis-eQTL   \n",
       "3_prime_UTR_variant                             22089   Lung  cis-eQTL   \n",
       "intron_variant                                   8061   Lung  cis-eQTL   \n",
       "downstream_gene_variant                         13383   Lung  cis-eQTL   \n",
       "\n",
       "                        maf_threshold annotation  \n",
       "sequence_feature                   1%     snpeff  \n",
       "5_prime_UTR_variant                1%     snpeff  \n",
       "3_prime_UTR_variant                1%     snpeff  \n",
       "intron_variant                     1%     snpeff  \n",
       "downstream_gene_variant            1%     snpeff  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for (tissue, modality, maf_threshold), df in credible_sets.groupby(['tissue', 'modality', 'maf_threshold']):\n",
    "    if 'trans' in modality and tissue != 'Whole_blood' or modality == 'trans-sQTL':\n",
    "        # matrix singularity / convergence issues likely\n",
    "        continue\n",
    "    control_df = control_credible_sets[(control_credible_sets.tissue==tissue) & (control_credible_sets.modality==modality) & (control_credible_sets.maf_threshold==maf_threshold)]\n",
    "    annotations_to_test = ['snpeff', 'regulatory_build', f'roadmap_{topmed_to_roadmap[tissue]}']\n",
    "    for a in annotations_to_test:\n",
    "        lr = logistic_regression(df, control_df, annotations[a])\n",
    "        if lr.converged.values[0] == False:\n",
    "            print(tissue, modality, maf_threshold, a, 'converged: ', lr.converged.values[0])\n",
    "        results.append(lr.assign(tissue=tissue, modality=modality, maf_threshold=maf_threshold, annotation=a))\n",
    "results = pd.concat(results)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:00:24.120212Z",
     "iopub.status.busy": "2024-12-02T21:00:24.119773Z",
     "iopub.status.idle": "2024-12-02T21:00:24.146423Z",
     "shell.execute_reply": "2024-12-02T21:00:24.145707Z"
    }
   },
   "outputs": [],
   "source": [
    "results.to_csv(f'tables/{PREFIX}enrichments-vs-control-variants.tsv', sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:00:24.148833Z",
     "iopub.status.busy": "2024-12-02T21:00:24.148392Z",
     "iopub.status.idle": "2024-12-02T21:00:45.052974Z",
     "shell.execute_reply": "2024-12-02T21:00:45.052286Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3369953/1042608307.py:2: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  cis_sharing = pd.read_csv('tables/summarize-cis-eqtl-coloc-with-cis-sqtl.cis-eqtl-and-sqtl-overlap.tsv', sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-eQTL      2388661\n",
      "cis-sQTL      1621388\n",
      "trans-eQTL        700\n",
      "trans-sQTL         67\n",
      "Name: modality, dtype: int64\n",
      "cis-eQTL    2388661\n",
      "cis-sQTL    1621388\n",
      "Name: modality, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# enrichments for cis-eQTL that are cis-sQTL or are not\n",
    "cis_sharing = pd.read_csv('tables/summarize-cis-eqtl-coloc-with-cis-sqtl.cis-eqtl-and-sqtl-overlap.tsv', sep='\\t')\n",
    "cis_sharing = cis_sharing.rename(columns={'coloced_with_other_modality_signal': 'colocs_with_other_modality'})\n",
    "# losing trans cs here -- but that's fine as we don't use them anymore\n",
    "print(credible_sets.modality.value_counts())\n",
    "credible_sets = credible_sets.merge(cis_sharing.rename(columns={'maf': 'maf_threshold'}))\n",
    "print(credible_sets.modality.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:00:45.056226Z",
     "iopub.status.busy": "2024-12-02T21:00:45.055819Z",
     "iopub.status.idle": "2024-12-02T21:00:45.066321Z",
     "shell.execute_reply": "2024-12-02T21:00:45.065876Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>colocing_phenotypes</th>\n",
       "      <th>colocs_with_other_modality</th>\n",
       "      <th>coloced_with_other_modality_signal_from_same_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:637861:693019:clu_45255_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold                                colocing_phenotypes  \\\n",
       "0  cis-eQTL            1%  chr6:637861:693019:clu_45255_-:ENSG00000112685.14   \n",
       "1  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "2  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "3  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "4  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "\n",
       "   colocs_with_other_modality  \\\n",
       "0                        True   \n",
       "1                        True   \n",
       "2                        True   \n",
       "3                        True   \n",
       "4                        True   \n",
       "\n",
       "   coloced_with_other_modality_signal_from_same_gene  \n",
       "0                                               True  \n",
       "1                                               True  \n",
       "2                                               True  \n",
       "3                                               True  \n",
       "4                                               True  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:00:45.068677Z",
     "iopub.status.busy": "2024-12-02T21:00:45.068223Z",
     "iopub.status.idle": "2024-12-02T21:01:23.909226Z",
     "shell.execute_reply": "2024-12-02T21:01:23.907739Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "      <th>converged</th>\n",
       "      <th>n_hits_in_annotation</th>\n",
       "      <th>n_hits_not_in_annotation</th>\n",
       "      <th>n_controls_in_annotation</th>\n",
       "      <th>n_controls_not_in_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>colocs_with_other_modality</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequence_feature</th>\n",
       "      <td>-0.102893</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>-2.983861</td>\n",
       "      <td>2.846360e-03</td>\n",
       "      <td>-0.170479</td>\n",
       "      <td>-0.035307</td>\n",
       "      <td>True</td>\n",
       "      <td>2414</td>\n",
       "      <td>74107</td>\n",
       "      <td>1682</td>\n",
       "      <td>74839</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>False</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <td>0.891838</td>\n",
       "      <td>0.028040</td>\n",
       "      <td>31.805785</td>\n",
       "      <td>5.384317e-222</td>\n",
       "      <td>0.836881</td>\n",
       "      <td>0.946796</td>\n",
       "      <td>True</td>\n",
       "      <td>8573</td>\n",
       "      <td>67948</td>\n",
       "      <td>2266</td>\n",
       "      <td>74255</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>False</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <td>0.317298</td>\n",
       "      <td>0.018947</td>\n",
       "      <td>16.746864</td>\n",
       "      <td>5.969097e-63</td>\n",
       "      <td>0.280163</td>\n",
       "      <td>0.354433</td>\n",
       "      <td>True</td>\n",
       "      <td>12203</td>\n",
       "      <td>64318</td>\n",
       "      <td>5949</td>\n",
       "      <td>70572</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>False</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_variant</th>\n",
       "      <td>-0.115431</td>\n",
       "      <td>0.015474</td>\n",
       "      <td>-7.459702</td>\n",
       "      <td>8.671868e-14</td>\n",
       "      <td>-0.145759</td>\n",
       "      <td>-0.085102</td>\n",
       "      <td>True</td>\n",
       "      <td>57413</td>\n",
       "      <td>19108</td>\n",
       "      <td>49873</td>\n",
       "      <td>26648</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>False</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <td>0.312512</td>\n",
       "      <td>0.012160</td>\n",
       "      <td>25.699438</td>\n",
       "      <td>1.185919e-145</td>\n",
       "      <td>0.288679</td>\n",
       "      <td>0.336346</td>\n",
       "      <td>True</td>\n",
       "      <td>43676</td>\n",
       "      <td>32845</td>\n",
       "      <td>30687</td>\n",
       "      <td>45834</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>False</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Coef.  Std.Err.          z          P>|z|  \\\n",
       "sequence_feature        -0.102893  0.034483  -2.983861   2.846360e-03   \n",
       "5_prime_UTR_variant      0.891838  0.028040  31.805785  5.384317e-222   \n",
       "3_prime_UTR_variant      0.317298  0.018947  16.746864   5.969097e-63   \n",
       "intron_variant          -0.115431  0.015474  -7.459702   8.671868e-14   \n",
       "downstream_gene_variant  0.312512  0.012160  25.699438  1.185919e-145   \n",
       "\n",
       "                           [0.025    0.975]  converged  n_hits_in_annotation  \\\n",
       "sequence_feature        -0.170479 -0.035307       True                  2414   \n",
       "5_prime_UTR_variant      0.836881  0.946796       True                  8573   \n",
       "3_prime_UTR_variant      0.280163  0.354433       True                 12203   \n",
       "intron_variant          -0.145759 -0.085102       True                 57413   \n",
       "downstream_gene_variant  0.288679  0.336346       True                 43676   \n",
       "\n",
       "                         n_hits_not_in_annotation  n_controls_in_annotation  \\\n",
       "sequence_feature                            74107                      1682   \n",
       "5_prime_UTR_variant                         67948                      2266   \n",
       "3_prime_UTR_variant                         64318                      5949   \n",
       "intron_variant                              19108                     49873   \n",
       "downstream_gene_variant                     32845                     30687   \n",
       "\n",
       "                         n_controls_not_in_annotation       tissue  modality  \\\n",
       "sequence_feature                                74839  Whole_blood  cis-eQTL   \n",
       "5_prime_UTR_variant                             74255  Whole_blood  cis-eQTL   \n",
       "3_prime_UTR_variant                             70572  Whole_blood  cis-eQTL   \n",
       "intron_variant                                  26648  Whole_blood  cis-eQTL   \n",
       "downstream_gene_variant                         45834  Whole_blood  cis-eQTL   \n",
       "\n",
       "                        maf_threshold  colocs_with_other_modality annotation  \n",
       "sequence_feature                 0.1%                       False     snpeff  \n",
       "5_prime_UTR_variant              0.1%                       False     snpeff  \n",
       "3_prime_UTR_variant              0.1%                       False     snpeff  \n",
       "intron_variant                   0.1%                       False     snpeff  \n",
       "downstream_gene_variant          0.1%                       False     snpeff  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for (tissue, modality, maf_threshold, colocs_with_other_modality), df in credible_sets.groupby(['tissue', 'modality', 'maf_threshold', 'colocs_with_other_modality']):\n",
    "    if tissue != 'Whole_blood':\n",
    "        continue\n",
    "    control_df = control_credible_sets.merge(df[['phenotype_id', 'cs_id', 'tissue', 'modality', 'maf_threshold', 'colocs_with_other_modality']].drop_duplicates())\n",
    "    annotations_to_test = ['snpeff', 'regulatory_build', f'roadmap_{topmed_to_roadmap[tissue]}']\n",
    "    for a in annotations_to_test:\n",
    "        lr = logistic_regression(df, control_df, annotations[a])\n",
    "        if lr.converged.values[0] == False:\n",
    "            print(tissue, modality, maf_threshold, colocs_with_other_modality, a, 'converged: ', lr.converged.values[0])\n",
    "        results.append(lr.assign(tissue=tissue, modality=modality, maf_threshold=maf_threshold, colocs_with_other_modality=colocs_with_other_modality, annotation=a))\n",
    "results = pd.concat(results)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:23.912467Z",
     "iopub.status.busy": "2024-12-02T21:01:23.912062Z",
     "iopub.status.idle": "2024-12-02T21:01:23.919800Z",
     "shell.execute_reply": "2024-12-02T21:01:23.919177Z"
    }
   },
   "outputs": [],
   "source": [
    "results.to_csv(f'tables/{PREFIX}cis-sharing.tsv', sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:23.922058Z",
     "iopub.status.busy": "2024-12-02T21:01:23.921802Z",
     "iopub.status.idle": "2024-12-02T21:01:24.150157Z",
     "shell.execute_reply": "2024-12-02T21:01:24.149397Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>in_whole_blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>chrX_100649875_A_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.518977</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>chr1_169788346_A_C</td>\n",
       "      <td>0.852837</td>\n",
       "      <td>0.508908</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>chr1_169757980_T_A</td>\n",
       "      <td>0.009195</td>\n",
       "      <td>0.091402</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>chr1_169796758_G_A</td>\n",
       "      <td>0.057029</td>\n",
       "      <td>0.300542</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000971.15</td>\n",
       "      <td>chr1_196391465_A_T</td>\n",
       "      <td>0.405031</td>\n",
       "      <td>0.560418</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id          variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000000003.14  chrX_100649875_A_G  1.000000  0.518977      1   Lung   \n",
       "1  ENSG00000000457.14  chr1_169788346_A_C  0.852837  0.508908      1   Lung   \n",
       "2  ENSG00000000460.17  chr1_169757980_T_A  0.009195  0.091402      1   Lung   \n",
       "3  ENSG00000000460.17  chr1_169796758_G_A  0.057029  0.300542      2   Lung   \n",
       "4  ENSG00000000971.15  chr1_196391465_A_T  0.405031  0.560418      1   Lung   \n",
       "\n",
       "   modality  in_whole_blood  \n",
       "0  cis-eQTL               0  \n",
       "1  cis-eQTL               1  \n",
       "2  cis-eQTL               1  \n",
       "3  cis-eQTL               0  \n",
       "4  cis-eQTL               0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare Lung CS that are in blood or aren't (and same for nasal epithelial)\n",
    "\n",
    "blood_as_proxy = pd.read_csv('tables/blood-as-proxy.credible_set_in_blood.tsv', sep='\\t')\n",
    "blood_as_proxy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:24.153489Z",
     "iopub.status.busy": "2024-12-02T21:01:24.152689Z",
     "iopub.status.idle": "2024-12-02T21:01:24.156222Z",
     "shell.execute_reply": "2024-12-02T21:01:24.155438Z"
    }
   },
   "outputs": [],
   "source": [
    "# blood_as_proxy.groupby(['tissue', 'modality', 'in_whole_blood']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:24.158458Z",
     "iopub.status.busy": "2024-12-02T21:01:24.158271Z",
     "iopub.status.idle": "2024-12-02T21:01:24.160928Z",
     "shell.execute_reply": "2024-12-02T21:01:24.160428Z"
    }
   },
   "outputs": [],
   "source": [
    "# blood_as_proxy.loc[blood_as_proxy.variant_id=='chr1_196891605_G_A',['phenotype_id', 'variant_id', 'tissue', 'modality']].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:24.162904Z",
     "iopub.status.busy": "2024-12-02T21:01:24.162729Z",
     "iopub.status.idle": "2024-12-02T21:01:24.165722Z",
     "shell.execute_reply": "2024-12-02T21:01:24.165054Z"
    }
   },
   "outputs": [],
   "source": [
    "# credible_sets.loc[credible_sets.variant_id=='chr1_196891605_G_A',['phenotype_id', 'variant_id', 'tissue', 'modality']].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:24.167913Z",
     "iopub.status.busy": "2024-12-02T21:01:24.167488Z",
     "iopub.status.idle": "2024-12-02T21:01:24.170800Z",
     "shell.execute_reply": "2024-12-02T21:01:24.169910Z"
    }
   },
   "outputs": [],
   "source": [
    "# WTF? cluster IDs different\n",
    "# Ah -- because the blood as proxy analysis was using EUR credible sets\n",
    "# need to update analyses accordingly\n",
    "# print(credible_sets.loc[credible_sets.variant_id=='chr1_196891605_G_A',['phenotype_id', 'variant_id', 'tissue', 'modality']].values[0][0])\n",
    "# print(blood_as_proxy.loc[blood_as_proxy.variant_id=='chr1_196891605_G_A',['phenotype_id', 'variant_id', 'tissue', 'modality']].values[0][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:24.172846Z",
     "iopub.status.busy": "2024-12-02T21:01:24.172678Z",
     "iopub.status.idle": "2024-12-02T21:01:25.320241Z",
     "shell.execute_reply": "2024-12-02T21:01:25.319543Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>colocing_phenotypes</th>\n",
       "      <th>colocs_with_other_modality</th>\n",
       "      <th>coloced_with_other_modality_signal_from_same_gene</th>\n",
       "      <th>in_whole_blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:637861:693019:clu_45255_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold                                colocing_phenotypes  \\\n",
       "0  cis-eQTL            1%  chr6:637861:693019:clu_45255_-:ENSG00000112685.14   \n",
       "1  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "2  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "3  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "4  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "\n",
       "   colocs_with_other_modality  \\\n",
       "0                        True   \n",
       "1                        True   \n",
       "2                        True   \n",
       "3                        True   \n",
       "4                        True   \n",
       "\n",
       "   coloced_with_other_modality_signal_from_same_gene  in_whole_blood  \n",
       "0                                               True               1  \n",
       "1                                               True               1  \n",
       "2                                               True               1  \n",
       "3                                               True               1  \n",
       "4                                               True               1  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blood_as_proxy = credible_sets.merge(blood_as_proxy[['phenotype_id', 'cs_id', 'tissue', 'modality', 'in_whole_blood']].drop_duplicates())\n",
    "blood_as_proxy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:25.324056Z",
     "iopub.status.busy": "2024-12-02T21:01:25.323247Z",
     "iopub.status.idle": "2024-12-02T21:01:25.327150Z",
     "shell.execute_reply": "2024-12-02T21:01:25.326297Z"
    }
   },
   "outputs": [],
   "source": [
    "# blood_as_proxy.groupby(['tissue', 'modality', 'in_whole_blood']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:25.330160Z",
     "iopub.status.busy": "2024-12-02T21:01:25.329436Z",
     "iopub.status.idle": "2024-12-02T21:01:46.713412Z",
     "shell.execute_reply": "2024-12-02T21:01:46.712706Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "      <th>converged</th>\n",
       "      <th>n_hits_in_annotation</th>\n",
       "      <th>n_hits_not_in_annotation</th>\n",
       "      <th>n_controls_in_annotation</th>\n",
       "      <th>n_controls_not_in_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>in_whole_blood</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequence_feature</th>\n",
       "      <td>0.092739</td>\n",
       "      <td>0.074010</td>\n",
       "      <td>1.253060</td>\n",
       "      <td>2.101839e-01</td>\n",
       "      <td>-0.052318</td>\n",
       "      <td>0.237795</td>\n",
       "      <td>True</td>\n",
       "      <td>588</td>\n",
       "      <td>10275</td>\n",
       "      <td>328</td>\n",
       "      <td>10535</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>0</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <td>0.461738</td>\n",
       "      <td>0.061782</td>\n",
       "      <td>7.473646</td>\n",
       "      <td>7.800276e-14</td>\n",
       "      <td>0.340647</td>\n",
       "      <td>0.582828</td>\n",
       "      <td>True</td>\n",
       "      <td>1389</td>\n",
       "      <td>9474</td>\n",
       "      <td>570</td>\n",
       "      <td>10293</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>0</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <td>0.364939</td>\n",
       "      <td>0.043569</td>\n",
       "      <td>8.376027</td>\n",
       "      <td>5.474443e-17</td>\n",
       "      <td>0.279544</td>\n",
       "      <td>0.450333</td>\n",
       "      <td>True</td>\n",
       "      <td>2586</td>\n",
       "      <td>8277</td>\n",
       "      <td>1279</td>\n",
       "      <td>9584</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>0</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_variant</th>\n",
       "      <td>0.304533</td>\n",
       "      <td>0.042868</td>\n",
       "      <td>7.103917</td>\n",
       "      <td>1.212693e-12</td>\n",
       "      <td>0.220512</td>\n",
       "      <td>0.388553</td>\n",
       "      <td>True</td>\n",
       "      <td>8978</td>\n",
       "      <td>1885</td>\n",
       "      <td>7313</td>\n",
       "      <td>3550</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>0</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <td>0.339265</td>\n",
       "      <td>0.034059</td>\n",
       "      <td>9.960947</td>\n",
       "      <td>2.259000e-23</td>\n",
       "      <td>0.272509</td>\n",
       "      <td>0.406020</td>\n",
       "      <td>True</td>\n",
       "      <td>7217</td>\n",
       "      <td>3646</td>\n",
       "      <td>5047</td>\n",
       "      <td>5816</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>0</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Coef.  Std.Err.         z         P>|z|    [0.025  \\\n",
       "sequence_feature         0.092739  0.074010  1.253060  2.101839e-01 -0.052318   \n",
       "5_prime_UTR_variant      0.461738  0.061782  7.473646  7.800276e-14  0.340647   \n",
       "3_prime_UTR_variant      0.364939  0.043569  8.376027  5.474443e-17  0.279544   \n",
       "intron_variant           0.304533  0.042868  7.103917  1.212693e-12  0.220512   \n",
       "downstream_gene_variant  0.339265  0.034059  9.960947  2.259000e-23  0.272509   \n",
       "\n",
       "                           0.975]  converged  n_hits_in_annotation  \\\n",
       "sequence_feature         0.237795       True                   588   \n",
       "5_prime_UTR_variant      0.582828       True                  1389   \n",
       "3_prime_UTR_variant      0.450333       True                  2586   \n",
       "intron_variant           0.388553       True                  8978   \n",
       "downstream_gene_variant  0.406020       True                  7217   \n",
       "\n",
       "                         n_hits_not_in_annotation  n_controls_in_annotation  \\\n",
       "sequence_feature                            10275                       328   \n",
       "5_prime_UTR_variant                          9474                       570   \n",
       "3_prime_UTR_variant                          8277                      1279   \n",
       "intron_variant                               1885                      7313   \n",
       "downstream_gene_variant                      3646                      5047   \n",
       "\n",
       "                         n_controls_not_in_annotation tissue  modality  \\\n",
       "sequence_feature                                10535   Lung  cis-eQTL   \n",
       "5_prime_UTR_variant                             10293   Lung  cis-eQTL   \n",
       "3_prime_UTR_variant                              9584   Lung  cis-eQTL   \n",
       "intron_variant                                   3550   Lung  cis-eQTL   \n",
       "downstream_gene_variant                          5816   Lung  cis-eQTL   \n",
       "\n",
       "                        maf_threshold  in_whole_blood annotation  \n",
       "sequence_feature                   1%               0     snpeff  \n",
       "5_prime_UTR_variant                1%               0     snpeff  \n",
       "3_prime_UTR_variant                1%               0     snpeff  \n",
       "intron_variant                     1%               0     snpeff  \n",
       "downstream_gene_variant            1%               0     snpeff  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for (tissue, modality, maf_threshold, in_whole_blood), df in blood_as_proxy.groupby(['tissue', 'modality', 'maf_threshold', 'in_whole_blood']):\n",
    "    if tissue not in ['Lung', 'Nasal_epithelial']:\n",
    "        continue\n",
    "    control_df = control_credible_sets.merge(df[['phenotype_id', 'cs_id', 'tissue', 'modality', 'maf_threshold', 'in_whole_blood']].drop_duplicates())\n",
    "    annotations_to_test = ['snpeff', 'regulatory_build', f'roadmap_{topmed_to_roadmap[tissue]}']\n",
    "    for a in annotations_to_test:\n",
    "        if a == 'regulatory_build':\n",
    "            continue\n",
    "        lr = logistic_regression(df, control_df, annotations[a])\n",
    "        if lr.converged.values[0] == False:\n",
    "            print(tissue, modality, maf_threshold, in_whole_blood, a, 'converged: ', lr.converged.values[0])\n",
    "        results.append(lr.assign(tissue=tissue, modality=modality, maf_threshold=maf_threshold, in_whole_blood=in_whole_blood, annotation=a))\n",
    "results = pd.concat(results)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:46.715664Z",
     "iopub.status.busy": "2024-12-02T21:01:46.715473Z",
     "iopub.status.idle": "2024-12-02T21:01:46.721597Z",
     "shell.execute_reply": "2024-12-02T21:01:46.721181Z"
    }
   },
   "outputs": [],
   "source": [
    "results.to_csv(f'tables/{PREFIX}blood-as-proxy.tsv', sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:46.723797Z",
     "iopub.status.busy": "2024-12-02T21:01:46.723410Z",
     "iopub.status.idle": "2024-12-02T21:01:47.138605Z",
     "shell.execute_reply": "2024-12-02T21:01:47.138133Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>modality</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>variant</th>\n",
       "      <th>rnk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000179344.16</td>\n",
       "      <td>1</td>\n",
       "      <td>chr6_32667428_C_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000179344.16</td>\n",
       "      <td>1</td>\n",
       "      <td>chr6_32667428_C_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000013573.17</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12_31076435_A_C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000013573.17</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12_31076435_A_C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000113504.21</td>\n",
       "      <td>1</td>\n",
       "      <td>chr5_1104823_C_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        tissue   maf  modality        phenotype_id  cs_id             variant  \\\n",
       "0  Whole_blood  0.1%  cis-eQTL  ENSG00000179344.16      1   chr6_32667428_C_T   \n",
       "1  Whole_blood    1%  cis-eQTL  ENSG00000179344.16      1   chr6_32667428_C_T   \n",
       "2  Whole_blood    1%  cis-eQTL  ENSG00000013573.17      1  chr12_31076435_A_C   \n",
       "3  Whole_blood  0.1%  cis-eQTL  ENSG00000013573.17      1  chr12_31076435_A_C   \n",
       "4  Whole_blood    1%  cis-eQTL  ENSG00000113504.21      1    chr5_1104823_C_T   \n",
       "\n",
       "   rnk  \n",
       "0    1  \n",
       "1    1  \n",
       "2    1  \n",
       "3    1  \n",
       "4    1  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# primary vs secondary\n",
    "cs_ranks = pd.read_csv(f'tables/rank-signals.cis-ranks.tsv', sep='\\t').rename(columns={'variant_id': 'variant', 'rank': 'rnk'})\n",
    "#cs_ranks.modality = cs_ranks.modality.map({'cis-eQTL': 'cis-eqtl', 'cis-sQTL': 'cis-sqtl'})\n",
    "#tissue_and_variant_to_min_rank = cs_ranks.groupby(['tissue', 'variant', 'modality', 'maf']).rnk.min().rename('min_rank').reset_index()\n",
    "#tissue_and_variant_to_max_rank = cs_ranks.groupby(['tissue', 'variant', 'modality', 'maf']).rnk.max().rename('max_rank').reset_index()\n",
    "#tissue_and_variant_to_rank = tissue_and_variant_to_max_rank.merge(tissue_and_variant_to_min_rank)\n",
    "#tissue_and_variant_to_rank.head()\n",
    "cs_ranks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:47.141175Z",
     "iopub.status.busy": "2024-12-02T21:01:47.140751Z",
     "iopub.status.idle": "2024-12-02T21:01:49.183344Z",
     "shell.execute_reply": "2024-12-02T21:01:49.182438Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>colocing_phenotypes</th>\n",
       "      <th>colocs_with_other_modality</th>\n",
       "      <th>coloced_with_other_modality_signal_from_same_gene</th>\n",
       "      <th>rnk</th>\n",
       "      <th>primary_secondary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:637861:693019:clu_45255_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>secondary</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold                                colocing_phenotypes  \\\n",
       "0  cis-eQTL            1%  chr6:637861:693019:clu_45255_-:ENSG00000112685.14   \n",
       "1  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "2  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "3  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "4  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "\n",
       "   colocs_with_other_modality  \\\n",
       "0                        True   \n",
       "1                        True   \n",
       "2                        True   \n",
       "3                        True   \n",
       "4                        True   \n",
       "\n",
       "   coloced_with_other_modality_signal_from_same_gene  rnk primary_secondary  \n",
       "0                                               True    1           primary  \n",
       "1                                               True    2         secondary  \n",
       "2                                               True    2         secondary  \n",
       "3                                               True    2         secondary  \n",
       "4                                               True    2         secondary  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: validate this merge\n",
    "cs_ranks = credible_sets.merge(cs_ranks[['tissue', 'maf', 'modality', 'phenotype_id', 'cs_id', 'rnk']].rename(columns={'maf': 'maf_threshold'}))\n",
    "cs_ranks['primary_secondary'] = ['primary' if i == 1 else 'secondary' for i in cs_ranks.rnk]\n",
    "cs_ranks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:01:49.186150Z",
     "iopub.status.busy": "2024-12-02T21:01:49.185750Z",
     "iopub.status.idle": "2024-12-02T21:02:27.597539Z",
     "shell.execute_reply": "2024-12-02T21:02:27.596878Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "      <th>converged</th>\n",
       "      <th>n_hits_in_annotation</th>\n",
       "      <th>n_hits_not_in_annotation</th>\n",
       "      <th>n_controls_in_annotation</th>\n",
       "      <th>n_controls_not_in_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>primary_secondary</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequence_feature</th>\n",
       "      <td>-0.007981</td>\n",
       "      <td>0.075604</td>\n",
       "      <td>-0.105564</td>\n",
       "      <td>9.159281e-01</td>\n",
       "      <td>-0.156163</td>\n",
       "      <td>0.140201</td>\n",
       "      <td>True</td>\n",
       "      <td>552</td>\n",
       "      <td>17842</td>\n",
       "      <td>346</td>\n",
       "      <td>18048</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>primary</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <td>1.056094</td>\n",
       "      <td>0.057625</td>\n",
       "      <td>18.327005</td>\n",
       "      <td>5.039021e-75</td>\n",
       "      <td>0.943151</td>\n",
       "      <td>1.169037</td>\n",
       "      <td>True</td>\n",
       "      <td>2536</td>\n",
       "      <td>15858</td>\n",
       "      <td>506</td>\n",
       "      <td>17888</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>primary</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <td>0.353018</td>\n",
       "      <td>0.039567</td>\n",
       "      <td>8.922056</td>\n",
       "      <td>4.577125e-19</td>\n",
       "      <td>0.275469</td>\n",
       "      <td>0.430568</td>\n",
       "      <td>True</td>\n",
       "      <td>3053</td>\n",
       "      <td>15341</td>\n",
       "      <td>1362</td>\n",
       "      <td>17032</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>primary</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_variant</th>\n",
       "      <td>-0.281058</td>\n",
       "      <td>0.032653</td>\n",
       "      <td>-8.607325</td>\n",
       "      <td>7.478543e-18</td>\n",
       "      <td>-0.345058</td>\n",
       "      <td>-0.217059</td>\n",
       "      <td>True</td>\n",
       "      <td>13589</td>\n",
       "      <td>4805</td>\n",
       "      <td>11629</td>\n",
       "      <td>6765</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>primary</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <td>0.332567</td>\n",
       "      <td>0.025395</td>\n",
       "      <td>13.095649</td>\n",
       "      <td>3.486791e-39</td>\n",
       "      <td>0.282794</td>\n",
       "      <td>0.382341</td>\n",
       "      <td>True</td>\n",
       "      <td>10460</td>\n",
       "      <td>7934</td>\n",
       "      <td>6982</td>\n",
       "      <td>11412</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>primary</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Coef.  Std.Err.          z         P>|z|  \\\n",
       "sequence_feature        -0.007981  0.075604  -0.105564  9.159281e-01   \n",
       "5_prime_UTR_variant      1.056094  0.057625  18.327005  5.039021e-75   \n",
       "3_prime_UTR_variant      0.353018  0.039567   8.922056  4.577125e-19   \n",
       "intron_variant          -0.281058  0.032653  -8.607325  7.478543e-18   \n",
       "downstream_gene_variant  0.332567  0.025395  13.095649  3.486791e-39   \n",
       "\n",
       "                           [0.025    0.975]  converged  n_hits_in_annotation  \\\n",
       "sequence_feature        -0.156163  0.140201       True                   552   \n",
       "5_prime_UTR_variant      0.943151  1.169037       True                  2536   \n",
       "3_prime_UTR_variant      0.275469  0.430568       True                  3053   \n",
       "intron_variant          -0.345058 -0.217059       True                 13589   \n",
       "downstream_gene_variant  0.282794  0.382341       True                 10460   \n",
       "\n",
       "                         n_hits_not_in_annotation  n_controls_in_annotation  \\\n",
       "sequence_feature                            17842                       346   \n",
       "5_prime_UTR_variant                         15858                       506   \n",
       "3_prime_UTR_variant                         15341                      1362   \n",
       "intron_variant                               4805                     11629   \n",
       "downstream_gene_variant                      7934                      6982   \n",
       "\n",
       "                         n_controls_not_in_annotation       tissue  modality  \\\n",
       "sequence_feature                                18048  Whole_blood  cis-eQTL   \n",
       "5_prime_UTR_variant                             17888  Whole_blood  cis-eQTL   \n",
       "3_prime_UTR_variant                             17032  Whole_blood  cis-eQTL   \n",
       "intron_variant                                   6765  Whole_blood  cis-eQTL   \n",
       "downstream_gene_variant                         11412  Whole_blood  cis-eQTL   \n",
       "\n",
       "                        maf_threshold primary_secondary annotation  \n",
       "sequence_feature                 0.1%           primary     snpeff  \n",
       "5_prime_UTR_variant              0.1%           primary     snpeff  \n",
       "3_prime_UTR_variant              0.1%           primary     snpeff  \n",
       "intron_variant                   0.1%           primary     snpeff  \n",
       "downstream_gene_variant          0.1%           primary     snpeff  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for (tissue, modality, maf_threshold, primary_secondary), df in cs_ranks.groupby(['tissue', 'modality', 'maf_threshold', 'primary_secondary']):\n",
    "    if tissue != 'Whole_blood':\n",
    "        continue\n",
    "    control_df = control_credible_sets.merge(df[['phenotype_id', 'cs_id', 'tissue', 'modality', 'maf_threshold', 'primary_secondary']].drop_duplicates())\n",
    "    annotations_to_test = ['snpeff', 'regulatory_build', f'roadmap_{topmed_to_roadmap[tissue]}']\n",
    "    for a in annotations_to_test:\n",
    "        lr = logistic_regression(df, control_df, annotations[a])\n",
    "        if lr.converged.values[0] == False:\n",
    "            print(tissue, modality, maf_threshold, in_whole_blood, a, 'converged: ', lr.converged.values[0])\n",
    "        results.append(lr.assign(tissue=tissue, modality=modality, maf_threshold=maf_threshold, primary_secondary=primary_secondary, annotation=a))\n",
    "results = pd.concat(results)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:27.600906Z",
     "iopub.status.busy": "2024-12-02T21:02:27.600212Z",
     "iopub.status.idle": "2024-12-02T21:02:27.609891Z",
     "shell.execute_reply": "2024-12-02T21:02:27.608620Z"
    }
   },
   "outputs": [],
   "source": [
    "results.to_csv(f'tables/{PREFIX}primary_secondary.tsv', sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:27.612638Z",
     "iopub.status.busy": "2024-12-02T21:02:27.612203Z",
     "iopub.status.idle": "2024-12-02T21:02:28.356964Z",
     "shell.execute_reply": "2024-12-02T21:02:28.355806Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>unique_cs_id</th>\n",
       "      <th>AFR_MAF</th>\n",
       "      <th>AMR_MAF</th>\n",
       "      <th>EUR_MAF</th>\n",
       "      <th>variant_maf</th>\n",
       "      <th>max_maf</th>\n",
       "      <th>driven_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.231720</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.967864</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.231720</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10:100275493:100280124:clu_5973_-:ENSG00000...</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>7</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-sQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10:100275493:100280124:clu_5973_-:ENSG00000...</td>\n",
       "      <td>0.231720</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000235823.2</td>\n",
       "      <td>chr10_100365714_A_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.179966</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000235823.2</td>\n",
       "      <td>ENSG00000235823.2___L1</td>\n",
       "      <td>0.241398</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>0.179180</td>\n",
       "      <td>0.179966</td>\n",
       "      <td>0.241398</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000107937.19</td>\n",
       "      <td>chr10_1009926_A_T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.874574</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107937.19</td>\n",
       "      <td>ENSG00000107937.19___L1</td>\n",
       "      <td>0.240860</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.098423</td>\n",
       "      <td>0.125426</td>\n",
       "      <td>0.240860</td>\n",
       "      <td>AFR,EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         phenotype_id           variant_id  \\\n",
       "2                                  ENSG00000196072.12  chr10_100278884_G_A   \n",
       "3                                  ENSG00000107566.14  chr10_100278884_G_A   \n",
       "5   chr10:100275493:100280124:clu_5973_-:ENSG00000...  chr10_100278884_G_A   \n",
       "7                                   ENSG00000235823.2  chr10_100365714_A_G   \n",
       "13                                 ENSG00000107937.19    chr10_1009926_A_T   \n",
       "\n",
       "         pip        af  cs_id       tissue  modality             gene_id  \\\n",
       "2   1.000000  0.363108      1  Whole_blood  cis-eQTL  ENSG00000196072.12   \n",
       "3   0.967864  0.363108      2  Whole_blood  cis-eQTL  ENSG00000107566.14   \n",
       "5   1.000000  0.363108      7  Whole_blood  cis-sQTL  ENSG00000196072.12   \n",
       "7   1.000000  0.179966      1  Whole_blood  cis-eQTL   ENSG00000235823.2   \n",
       "13  1.000000  0.874574      1  Whole_blood  cis-eQTL  ENSG00000107937.19   \n",
       "\n",
       "                                         unique_cs_id   AFR_MAF   AMR_MAF  \\\n",
       "2                             ENSG00000196072.12___L1  0.231720  0.231707   \n",
       "3                             ENSG00000107566.14___L2  0.231720  0.231707   \n",
       "5   chr10:100275493:100280124:clu_5973_-:ENSG00000...  0.231720  0.231707   \n",
       "7                              ENSG00000235823.2___L1  0.241398  0.097561   \n",
       "13                            ENSG00000107937.19___L1  0.240860  0.041667   \n",
       "\n",
       "     EUR_MAF  variant_maf   max_maf    driven_by  \n",
       "2   0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "3   0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "5   0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "7   0.179180     0.179966  0.241398  AFR,AMR,EUR  \n",
       "13  0.098423     0.125426  0.240860      AFR,EUR  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rare vs common\n",
    "mafs = pd.read_csv('tables/cis-rare-variants-summary.maf-breakdown.tsv', sep='\\t')\n",
    "mafs = mafs[mafs.maf=='0.1%'].drop(columns=['maf'])\n",
    "mafs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:28.359889Z",
     "iopub.status.busy": "2024-12-02T21:02:28.359454Z",
     "iopub.status.idle": "2024-12-02T21:02:29.253607Z",
     "shell.execute_reply": "2024-12-02T21:02:29.252343Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>unique_cs_id</th>\n",
       "      <th>AFR_MAF</th>\n",
       "      <th>AMR_MAF</th>\n",
       "      <th>EUR_MAF</th>\n",
       "      <th>variant_maf</th>\n",
       "      <th>max_maf</th>\n",
       "      <th>driven_by</th>\n",
       "      <th>rare_in_all</th>\n",
       "      <th>common</th>\n",
       "      <th>frequency_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.231720</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "      <td>not_rare_in_all</td>\n",
       "      <td>common</td>\n",
       "      <td>common_in_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.967864</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.231720</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "      <td>not_rare_in_all</td>\n",
       "      <td>common</td>\n",
       "      <td>common_in_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10:100275493:100280124:clu_5973_-:ENSG00000...</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>7</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-sQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10:100275493:100280124:clu_5973_-:ENSG00000...</td>\n",
       "      <td>0.231720</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "      <td>not_rare_in_all</td>\n",
       "      <td>common</td>\n",
       "      <td>common_in_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000235823.2</td>\n",
       "      <td>chr10_100365714_A_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.179966</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000235823.2</td>\n",
       "      <td>ENSG00000235823.2___L1</td>\n",
       "      <td>0.241398</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>0.179180</td>\n",
       "      <td>0.179966</td>\n",
       "      <td>0.241398</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "      <td>not_rare_in_all</td>\n",
       "      <td>common</td>\n",
       "      <td>common_in_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000107937.19</td>\n",
       "      <td>chr10_1009926_A_T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.874574</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107937.19</td>\n",
       "      <td>ENSG00000107937.19___L1</td>\n",
       "      <td>0.240860</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.098423</td>\n",
       "      <td>0.125426</td>\n",
       "      <td>0.240860</td>\n",
       "      <td>AFR,EUR</td>\n",
       "      <td>not_rare_in_all</td>\n",
       "      <td>common</td>\n",
       "      <td>common_in_all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         phenotype_id           variant_id  \\\n",
       "2                                  ENSG00000196072.12  chr10_100278884_G_A   \n",
       "3                                  ENSG00000107566.14  chr10_100278884_G_A   \n",
       "5   chr10:100275493:100280124:clu_5973_-:ENSG00000...  chr10_100278884_G_A   \n",
       "7                                   ENSG00000235823.2  chr10_100365714_A_G   \n",
       "13                                 ENSG00000107937.19    chr10_1009926_A_T   \n",
       "\n",
       "         pip        af  cs_id       tissue  modality             gene_id  \\\n",
       "2   1.000000  0.363108      1  Whole_blood  cis-eQTL  ENSG00000196072.12   \n",
       "3   0.967864  0.363108      2  Whole_blood  cis-eQTL  ENSG00000107566.14   \n",
       "5   1.000000  0.363108      7  Whole_blood  cis-sQTL  ENSG00000196072.12   \n",
       "7   1.000000  0.179966      1  Whole_blood  cis-eQTL   ENSG00000235823.2   \n",
       "13  1.000000  0.874574      1  Whole_blood  cis-eQTL  ENSG00000107937.19   \n",
       "\n",
       "                                         unique_cs_id   AFR_MAF   AMR_MAF  \\\n",
       "2                             ENSG00000196072.12___L1  0.231720  0.231707   \n",
       "3                             ENSG00000107566.14___L2  0.231720  0.231707   \n",
       "5   chr10:100275493:100280124:clu_5973_-:ENSG00000...  0.231720  0.231707   \n",
       "7                              ENSG00000235823.2___L1  0.241398  0.097561   \n",
       "13                            ENSG00000107937.19___L1  0.240860  0.041667   \n",
       "\n",
       "     EUR_MAF  variant_maf   max_maf    driven_by      rare_in_all  common  \\\n",
       "2   0.444479     0.363108  0.444479  AFR,AMR,EUR  not_rare_in_all  common   \n",
       "3   0.444479     0.363108  0.444479  AFR,AMR,EUR  not_rare_in_all  common   \n",
       "5   0.444479     0.363108  0.444479  AFR,AMR,EUR  not_rare_in_all  common   \n",
       "7   0.179180     0.179966  0.241398  AFR,AMR,EUR  not_rare_in_all  common   \n",
       "13  0.098423     0.125426  0.240860      AFR,EUR  not_rare_in_all  common   \n",
       "\n",
       "   frequency_class  \n",
       "2    common_in_all  \n",
       "3    common_in_all  \n",
       "5    common_in_all  \n",
       "7    common_in_all  \n",
       "13   common_in_all  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# break up by modality, rare, common\n",
    "def make_label(maf_series):\n",
    "    if max(maf_series) < 0.01:\n",
    "        return 'rare_in_all'\n",
    "    elif min(maf_series) >= 0.01:\n",
    "        return 'common_in_all'\n",
    "    elif min(maf_series) < 0.01 and max(maf_series) >= 0.01:\n",
    "        return 'rare_in_some'\n",
    "    else:\n",
    "        raise ValueError('This should not occur')\n",
    "\n",
    "mafs['rare_in_all'] = (mafs.max_maf<0.01).map({True: 'rare_in_all', False: 'not_rare_in_all'})\n",
    "mafs['common'] = (mafs.variant_maf>=0.01).map({True: 'common', False: 'rare'})\n",
    "mafs['frequency_class'] = mafs[['AFR_MAF', 'AMR_MAF', 'EUR_MAF']].apply(make_label, axis=1)\n",
    "mafs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:29.256964Z",
     "iopub.status.busy": "2024-12-02T21:02:29.256133Z",
     "iopub.status.idle": "2024-12-02T21:02:29.267472Z",
     "shell.execute_reply": "2024-12-02T21:02:29.266924Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>maf_threshold</th>\n",
       "      <th>colocing_phenotypes</th>\n",
       "      <th>colocs_with_other_modality</th>\n",
       "      <th>coloced_with_other_modality_signal_from_same_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.682804</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:637861:693019:clu_45255_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_506781_A_G</td>\n",
       "      <td>0.043571</td>\n",
       "      <td>0.578234</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518078_G_A</td>\n",
       "      <td>0.038384</td>\n",
       "      <td>0.595662</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_518480_A_T</td>\n",
       "      <td>0.071511</td>\n",
       "      <td>0.594888</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>chr6_528568_T_C</td>\n",
       "      <td>0.058727</td>\n",
       "      <td>0.595275</td>\n",
       "      <td>2</td>\n",
       "      <td>Lung</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>1%</td>\n",
       "      <td>chr6:532610:549175:clu_45252_-:ENSG00000112685.14</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       variant_id       pip        af  cs_id tissue  \\\n",
       "0  ENSG00000112685.14  chr6_692805_C_G  1.000000  0.682804      1   Lung   \n",
       "1  ENSG00000112685.14  chr6_506781_A_G  0.043571  0.578234      2   Lung   \n",
       "2  ENSG00000112685.14  chr6_518078_G_A  0.038384  0.595662      2   Lung   \n",
       "3  ENSG00000112685.14  chr6_518480_A_T  0.071511  0.594888      2   Lung   \n",
       "4  ENSG00000112685.14  chr6_528568_T_C  0.058727  0.595275      2   Lung   \n",
       "\n",
       "   modality maf_threshold                                colocing_phenotypes  \\\n",
       "0  cis-eQTL            1%  chr6:637861:693019:clu_45255_-:ENSG00000112685.14   \n",
       "1  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "2  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "3  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "4  cis-eQTL            1%  chr6:532610:549175:clu_45252_-:ENSG00000112685.14   \n",
       "\n",
       "   colocs_with_other_modality  \\\n",
       "0                        True   \n",
       "1                        True   \n",
       "2                        True   \n",
       "3                        True   \n",
       "4                        True   \n",
       "\n",
       "   coloced_with_other_modality_signal_from_same_gene  \n",
       "0                                               True  \n",
       "1                                               True  \n",
       "2                                               True  \n",
       "3                                               True  \n",
       "4                                               True  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "credible_sets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:29.270637Z",
     "iopub.status.busy": "2024-12-02T21:02:29.269924Z",
     "iopub.status.idle": "2024-12-02T21:02:30.222354Z",
     "shell.execute_reply": "2024-12-02T21:02:30.221172Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-eQTL    531239\n",
      "cis-sQTL    331781\n",
      "Name: modality, dtype: int64\n",
      "cis-eQTL    531239\n",
      "cis-sQTL    331781\n",
      "Name: modality, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(credible_sets[credible_sets.maf_threshold=='0.1%'].modality.value_counts())\n",
    "mafs = credible_sets[credible_sets.maf_threshold=='0.1%'].merge(mafs[['phenotype_id', 'cs_id', 'tissue', 'modality', 'frequency_class']].drop_duplicates())\n",
    "print(mafs.modality.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:30.225300Z",
     "iopub.status.busy": "2024-12-02T21:02:30.224910Z",
     "iopub.status.idle": "2024-12-02T21:02:46.594209Z",
     "shell.execute_reply": "2024-12-02T21:02:46.592993Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coef.</th>\n",
       "      <th>Std.Err.</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "      <th>converged</th>\n",
       "      <th>n_hits_in_annotation</th>\n",
       "      <th>n_hits_not_in_annotation</th>\n",
       "      <th>n_controls_in_annotation</th>\n",
       "      <th>n_controls_not_in_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>frequency_class</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sequence_feature</th>\n",
       "      <td>-0.102950</td>\n",
       "      <td>0.047722</td>\n",
       "      <td>-2.157280</td>\n",
       "      <td>3.098384e-02</td>\n",
       "      <td>-0.196483</td>\n",
       "      <td>-0.009416</td>\n",
       "      <td>True</td>\n",
       "      <td>1268</td>\n",
       "      <td>41610</td>\n",
       "      <td>873</td>\n",
       "      <td>42005</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>common_in_all</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <td>0.783749</td>\n",
       "      <td>0.039572</td>\n",
       "      <td>19.805706</td>\n",
       "      <td>2.658115e-87</td>\n",
       "      <td>0.706190</td>\n",
       "      <td>0.861309</td>\n",
       "      <td>True</td>\n",
       "      <td>4053</td>\n",
       "      <td>38825</td>\n",
       "      <td>1165</td>\n",
       "      <td>41713</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>common_in_all</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <td>0.376418</td>\n",
       "      <td>0.026146</td>\n",
       "      <td>14.396637</td>\n",
       "      <td>5.432340e-47</td>\n",
       "      <td>0.325172</td>\n",
       "      <td>0.427663</td>\n",
       "      <td>True</td>\n",
       "      <td>6556</td>\n",
       "      <td>36322</td>\n",
       "      <td>3019</td>\n",
       "      <td>39859</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>common_in_all</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron_variant</th>\n",
       "      <td>-0.092860</td>\n",
       "      <td>0.021402</td>\n",
       "      <td>-4.338940</td>\n",
       "      <td>1.431715e-05</td>\n",
       "      <td>-0.134806</td>\n",
       "      <td>-0.050914</td>\n",
       "      <td>True</td>\n",
       "      <td>31723</td>\n",
       "      <td>11155</td>\n",
       "      <td>26913</td>\n",
       "      <td>15965</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>common_in_all</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <td>0.344206</td>\n",
       "      <td>0.016254</td>\n",
       "      <td>21.176851</td>\n",
       "      <td>1.561257e-99</td>\n",
       "      <td>0.312349</td>\n",
       "      <td>0.376063</td>\n",
       "      <td>True</td>\n",
       "      <td>23840</td>\n",
       "      <td>19038</td>\n",
       "      <td>16164</td>\n",
       "      <td>26714</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>common_in_all</td>\n",
       "      <td>snpeff</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Coef.  Std.Err.          z         P>|z|  \\\n",
       "sequence_feature        -0.102950  0.047722  -2.157280  3.098384e-02   \n",
       "5_prime_UTR_variant      0.783749  0.039572  19.805706  2.658115e-87   \n",
       "3_prime_UTR_variant      0.376418  0.026146  14.396637  5.432340e-47   \n",
       "intron_variant          -0.092860  0.021402  -4.338940  1.431715e-05   \n",
       "downstream_gene_variant  0.344206  0.016254  21.176851  1.561257e-99   \n",
       "\n",
       "                           [0.025    0.975]  converged  n_hits_in_annotation  \\\n",
       "sequence_feature        -0.196483 -0.009416       True                  1268   \n",
       "5_prime_UTR_variant      0.706190  0.861309       True                  4053   \n",
       "3_prime_UTR_variant      0.325172  0.427663       True                  6556   \n",
       "intron_variant          -0.134806 -0.050914       True                 31723   \n",
       "downstream_gene_variant  0.312349  0.376063       True                 23840   \n",
       "\n",
       "                         n_hits_not_in_annotation  n_controls_in_annotation  \\\n",
       "sequence_feature                            41610                       873   \n",
       "5_prime_UTR_variant                         38825                      1165   \n",
       "3_prime_UTR_variant                         36322                      3019   \n",
       "intron_variant                              11155                     26913   \n",
       "downstream_gene_variant                     19038                     16164   \n",
       "\n",
       "                         n_controls_not_in_annotation       tissue  modality  \\\n",
       "sequence_feature                                42005  Whole_blood  cis-eQTL   \n",
       "5_prime_UTR_variant                             41713  Whole_blood  cis-eQTL   \n",
       "3_prime_UTR_variant                             39859  Whole_blood  cis-eQTL   \n",
       "intron_variant                                  15965  Whole_blood  cis-eQTL   \n",
       "downstream_gene_variant                         26714  Whole_blood  cis-eQTL   \n",
       "\n",
       "                        frequency_class annotation  \n",
       "sequence_feature          common_in_all     snpeff  \n",
       "5_prime_UTR_variant       common_in_all     snpeff  \n",
       "3_prime_UTR_variant       common_in_all     snpeff  \n",
       "intron_variant            common_in_all     snpeff  \n",
       "downstream_gene_variant   common_in_all     snpeff  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for (tissue, modality, frequency_class), df in mafs.groupby(['tissue', 'modality', 'frequency_class']):\n",
    "    if tissue != 'Whole_blood':\n",
    "        continue\n",
    "    control_df = control_credible_sets[control_credible_sets.maf_threshold=='0.1%'].merge(df[['phenotype_id', 'cs_id', 'tissue', 'modality', 'frequency_class']].drop_duplicates())\n",
    "    annotations_to_test = ['snpeff', 'regulatory_build', f'roadmap_{topmed_to_roadmap[tissue]}']\n",
    "    for a in annotations_to_test:\n",
    "        lr = logistic_regression(df, control_df, annotations[a])\n",
    "        if lr.converged.values[0] == False:\n",
    "            print(tissue, modality, frequency_class, a, 'converged: ', lr.converged.values[0])\n",
    "        results.append(lr.assign(tissue=tissue, modality=modality, frequency_class=frequency_class, annotation=a))\n",
    "results = pd.concat(results)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T21:02:46.597522Z",
     "iopub.status.busy": "2024-12-02T21:02:46.597283Z",
     "iopub.status.idle": "2024-12-02T21:02:46.604932Z",
     "shell.execute_reply": "2024-12-02T21:02:46.604167Z"
    }
   },
   "outputs": [],
   "source": [
    "results.to_csv(f'tables/{PREFIX}frequency_class.tsv', sep='\\t', index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
