{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:07:31.814655Z",
     "iopub.status.busy": "2024-12-02T19:07:31.814490Z",
     "iopub.status.idle": "2024-12-02T19:09:49.897858Z",
     "shell.execute_reply": "2024-12-02T19:09:49.895844Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "PREFIX = 'cis-rare-variants-summary.'\n",
    "\n",
    "def phenotype_id_to_gene_id(x):\n",
    "    # first, try to match with a version\n",
    "    ENSEMBL_RE_WITH_VERSION = r'ENSG\\d+\\.\\d+'\n",
    "    ENSEMBL_RE_WITHOUT_VERSION = r'ENSG\\d+'\n",
    "    with_version = re.search(ENSEMBL_RE_WITH_VERSION, x)\n",
    "    without_version = re.search(ENSEMBL_RE_WITHOUT_VERSION, x)\n",
    "    if with_version:\n",
    "        return with_version.group(0)\n",
    "    elif without_version:\n",
    "        return without_version.group(0)\n",
    "    else:\n",
    "        raise ValueError(f'Not able to infer gene ID from {x}')\n",
    "\n",
    "\n",
    "def compare_credible_sets(scan_1_cs, scan_2_cs, summarize=True):\n",
    "    \"\"\"\n",
    "    Given a dataframe representing scan_1_cs and scan_2_cs (each having columns ['phenotype_id', 'variant_id', 'cs_id']; can have other columns too),\n",
    "    return a dataframe showing, for each CS in scan_1_cs, whether it overlaps a CS in scan_2_cs (for the same phenotype_id)\n",
    "    If summarize = False, simply returns scan_1_cs with an added column indicating whether each credible set SNP is a credible set SNP \n",
    "    for the same phenotype in the other scan\n",
    "    \"\"\"\n",
    "    # Validate input\n",
    "    if not isinstance(scan_1_cs, pd.DataFrame):\n",
    "        raise TypeError('scan_1_cs must be a DataFrame')\n",
    "    if not isinstance(scan_2_cs, pd.DataFrame):\n",
    "        raise TypeError('scan_2_cs must be a DataFrame')\n",
    "    for i in ['phenotype_id', 'variant_id', 'cs_id']:\n",
    "        if not i in scan_1_cs.columns.to_list():\n",
    "            raise ValueError(f'scan_1_cs must include column {i}')\n",
    "        if not i in scan_2_cs.columns.to_list():\n",
    "            raise ValueError(f'scan_2_cs must include column {i}')\n",
    "\n",
    "    results = scan_1_cs.merge(scan_2_cs[['phenotype_id', 'variant_id']].drop_duplicates().assign(in_other_scan_cs=1), how='left')\n",
    "    results.in_other_scan_cs = results.in_other_scan_cs.fillna(0).astype(int)\n",
    "    assert(len(results) == len(scan_1_cs))\n",
    "\n",
    "    if summarize:\n",
    "        return results.groupby(['phenotype_id', 'cs_id']).in_other_scan_cs.max().reset_index()\n",
    "    else:\n",
    "        return results\n",
    "\n",
    "\n",
    "def top_pip_variants(cs):\n",
    "    \"\"\"\n",
    "    Given a dataframe representing CS (having columns ['phenotype_id', 'variant_id', 'cs_id', 'pip']; can have other columns too),\n",
    "    return a dataframe containing only the top PIP variant per credible set (credible set defined by phenotype_id + csID pair)\n",
    "    \"\"\"\n",
    "    # Validate input\n",
    "    if not isinstance(cs, pd.DataFrame):\n",
    "        raise TypeError('cs must be a DataFrame')\n",
    "    for i in ['phenotype_id', 'variant_id', 'cs_id', 'pip']:\n",
    "        if not i in cs.columns.to_list():\n",
    "            raise ValueError(f'cs must include column {i}')\n",
    "\n",
    "    return cs.sort_values(['pip', 'variant_id'], ascending=[False, True]).groupby(['phenotype_id', 'cs_id']).head(1)\n",
    "\n",
    "\n",
    "CIS_EQTL_PCS = {\n",
    "    'Lung': 75,\n",
    "    'Monocyte': 30,\n",
    "    'Nasal_epithelial': 30,\n",
    "    'PBMC': 30,\n",
    "    'T_cell': 30,\n",
    "    'Whole_blood': 100,\n",
    "}\n",
    "\n",
    "CIS_SQTL_PCS = {\n",
    "    'Lung': 10,\n",
    "    'Monocyte': 10,\n",
    "    'Nasal_epithelial': 10,\n",
    "    'PBMC': 10,\n",
    "    'T_cell': 10,\n",
    "    'Whole_blood': 10,\n",
    "}\n",
    "\n",
    "METADATA = '../data/metadata/metadata.tm.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:09:49.902122Z",
     "iopub.status.busy": "2024-12-02T19:09:49.901327Z",
     "iopub.status.idle": "2024-12-02T19:09:55.636241Z",
     "shell.execute_reply": "2024-12-02T19:09:55.635255Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>num_var</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>true_df</th>\n",
       "      <th>pval_true_df</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>...</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>tissue</th>\n",
       "      <th>pcs</th>\n",
       "      <th>maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>4927</td>\n",
       "      <td>1.02051</td>\n",
       "      <td>994.008</td>\n",
       "      <td>5807.91</td>\n",
       "      <td>5.121050e-54</td>\n",
       "      <td>chr6_184740_C_T</td>\n",
       "      <td>401</td>\n",
       "      <td>351</td>\n",
       "      <td>366</td>\n",
       "      <td>...</td>\n",
       "      <td>1.180640e-59</td>\n",
       "      <td>-0.892507</td>\n",
       "      <td>0.054226</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>4.693430e-52</td>\n",
       "      <td>2.136940e-51</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000112679.14</td>\n",
       "      <td>5553</td>\n",
       "      <td>1.04192</td>\n",
       "      <td>1091.760</td>\n",
       "      <td>5735.69</td>\n",
       "      <td>4.436730e-16</td>\n",
       "      <td>chr6_250373_T_TGAG</td>\n",
       "      <td>-41257</td>\n",
       "      <td>592</td>\n",
       "      <td>623</td>\n",
       "      <td>...</td>\n",
       "      <td>7.488500e-18</td>\n",
       "      <td>0.377827</td>\n",
       "      <td>0.043769</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>1.449020e-13</td>\n",
       "      <td>3.235320e-13</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000137265.15</td>\n",
       "      <td>6093</td>\n",
       "      <td>1.04731</td>\n",
       "      <td>1128.120</td>\n",
       "      <td>5717.04</td>\n",
       "      <td>1.010130e-18</td>\n",
       "      <td>chr6_396321_C_T</td>\n",
       "      <td>4569</td>\n",
       "      <td>1464</td>\n",
       "      <td>1642</td>\n",
       "      <td>...</td>\n",
       "      <td>7.095080e-21</td>\n",
       "      <td>0.250111</td>\n",
       "      <td>0.026594</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>2.192040e-16</td>\n",
       "      <td>5.316970e-16</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000112685.14</td>\n",
       "      <td>7690</td>\n",
       "      <td>1.03145</td>\n",
       "      <td>1525.450</td>\n",
       "      <td>5758.99</td>\n",
       "      <td>4.654260e-50</td>\n",
       "      <td>chr6_692805_C_G</td>\n",
       "      <td>-334</td>\n",
       "      <td>3847</td>\n",
       "      <td>4871</td>\n",
       "      <td>...</td>\n",
       "      <td>1.004740e-55</td>\n",
       "      <td>-0.293773</td>\n",
       "      <td>0.018501</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>2.476110e-48</td>\n",
       "      <td>1.071110e-47</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000272463.1</td>\n",
       "      <td>7808</td>\n",
       "      <td>1.02158</td>\n",
       "      <td>1816.810</td>\n",
       "      <td>5904.27</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr6_711150_A_C</td>\n",
       "      <td>-255</td>\n",
       "      <td>3861</td>\n",
       "      <td>4826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.810795</td>\n",
       "      <td>0.016028</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0</td>\n",
       "      <td>1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id  num_var  beta_shape1  beta_shape2  true_df  \\\n",
       "0   ENSG00000285578.1     4927      1.02051      994.008  5807.91   \n",
       "1  ENSG00000112679.14     5553      1.04192     1091.760  5735.69   \n",
       "2  ENSG00000137265.15     6093      1.04731     1128.120  5717.04   \n",
       "3  ENSG00000112685.14     7690      1.03145     1525.450  5758.99   \n",
       "4   ENSG00000272463.1     7808      1.02158     1816.810  5904.27   \n",
       "\n",
       "   pval_true_df          variant_id  tss_distance  ma_samples  ma_count  ...  \\\n",
       "0  5.121050e-54     chr6_184740_C_T           401         351       366  ...   \n",
       "1  4.436730e-16  chr6_250373_T_TGAG        -41257         592       623  ...   \n",
       "2  1.010130e-18     chr6_396321_C_T          4569        1464      1642  ...   \n",
       "3  4.654260e-50     chr6_692805_C_G          -334        3847      4871  ...   \n",
       "4  0.000000e+00     chr6_711150_A_C          -255        3861      4826  ...   \n",
       "\n",
       "   pval_nominal     slope  slope_se  pval_perm     pval_beta          qval  \\\n",
       "0  1.180640e-59 -0.892507  0.054226     0.0001  4.693430e-52  2.136940e-51   \n",
       "1  7.488500e-18  0.377827  0.043769     0.0001  1.449020e-13  3.235320e-13   \n",
       "2  7.095080e-21  0.250111  0.026594     0.0001  2.192040e-16  5.316970e-16   \n",
       "3  1.004740e-55 -0.293773  0.018501     0.0001  2.476110e-48  1.071110e-47   \n",
       "4  0.000000e+00  0.810795  0.016028     0.0001  0.000000e+00  0.000000e+00   \n",
       "\n",
       "   pval_nominal_threshold       tissue pcs  maf  \n",
       "0                0.000042  Whole_blood   0   1%  \n",
       "1                0.000042  Whole_blood   0   1%  \n",
       "2                0.000041  Whole_blood   0   1%  \n",
       "3                0.000029  Whole_blood   0   1%  \n",
       "4                0.000023  Whole_blood   0   1%  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load in primary hits\n",
    "CIS_EQTL_PERMUTATIONS_GLOB = glob.glob('../data/scan-results/joint/cis-eqtl/permutations/maf*/Whole_blood*.txt.gz')\n",
    "CIS_SQTL_PERMUTATIONS_GLOB = glob.glob('../data/scan-results/joint/cis-sqtl/permutations/maf*/Whole_blood*.txt.gz')\n",
    "cis_eqtl_permutations_all = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).split('.')[0], pcs=int(os.path.basename(f).split('.')[1]), maf='1%' if 'maf001' in f else '0.1%') for f in CIS_EQTL_PERMUTATIONS_GLOB])\n",
    "cis_sqtl_permutations_all = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).split('.')[0], pcs=int(os.path.basename(f).split('.')[1]), maf='1%' if 'maf001' in f else '0.1%') for f in CIS_SQTL_PERMUTATIONS_GLOB])\n",
    "cis_eqtl_permutations_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:09:55.683275Z",
     "iopub.status.busy": "2024-12-02T19:09:55.682513Z",
     "iopub.status.idle": "2024-12-02T19:09:55.959861Z",
     "shell.execute_reply": "2024-12-02T19:09:55.959321Z"
    }
   },
   "outputs": [],
   "source": [
    "cis_eqtl_permutations = cis_eqtl_permutations_all[cis_eqtl_permutations_all.tissue.map(CIS_EQTL_PCS) == cis_eqtl_permutations_all.pcs]\n",
    "cis_sqtl_permutations = cis_sqtl_permutations_all[cis_sqtl_permutations_all.tissue.map(CIS_SQTL_PCS) == cis_sqtl_permutations_all.pcs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:09:55.962817Z",
     "iopub.status.busy": "2024-12-02T19:09:55.962495Z",
     "iopub.status.idle": "2024-12-02T19:09:55.978852Z",
     "shell.execute_reply": "2024-12-02T19:09:55.977972Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "With the lower MAF threshold, the average number of variants tested against each gene more than doubled (7,866 vs 17,195 for the cis-eQTL scan)\n"
     ]
    }
   ],
   "source": [
    "print('With the lower MAF threshold, the average number of variants tested against each gene more than doubled ({:,} vs {:,} for the cis-eQTL scan)'.format(\n",
    "    round(cis_eqtl_permutations[(cis_eqtl_permutations.maf=='1%')].num_var.mean()),\n",
    "    round(cis_eqtl_permutations[(cis_eqtl_permutations.maf=='0.1%')].num_var.mean())\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:09:55.981273Z",
     "iopub.status.busy": "2024-12-02T19:09:55.980933Z",
     "iopub.status.idle": "2024-12-02T19:09:56.042579Z",
     "shell.execute_reply": "2024-12-02T19:09:56.041753Z"
    }
   },
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(METADATA, sep='\\t')\n",
    "sample_size_dict = metadata[metadata.used_for_scan].groupby('tissue').size().to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:09:56.046210Z",
     "iopub.status.busy": "2024-12-02T19:09:56.045497Z",
     "iopub.status.idle": "2024-12-02T19:09:58.193998Z",
     "shell.execute_reply": "2024-12-02T19:09:58.193441Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>modality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_180573_C_A</td>\n",
       "      <td>0.274643</td>\n",
       "      <td>0.027967</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_184740_C_T</td>\n",
       "      <td>0.725334</td>\n",
       "      <td>0.028355</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_153439_T_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.678416</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_196027_C_T</td>\n",
       "      <td>0.170460</td>\n",
       "      <td>0.066006</td>\n",
       "      <td>3</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_197265_T_C</td>\n",
       "      <td>0.829534</td>\n",
       "      <td>0.073365</td>\n",
       "      <td>3</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phenotype_id       variant_id       pip        af  cs_id       tissue  \\\n",
       "0  ENSG00000285578.1  chr6_180573_C_A  0.274643  0.027967      1  Whole_blood   \n",
       "1  ENSG00000285578.1  chr6_184740_C_T  0.725334  0.028355      1  Whole_blood   \n",
       "2  ENSG00000285578.1  chr6_153439_T_A  1.000000  0.678416      2  Whole_blood   \n",
       "3  ENSG00000285578.1  chr6_196027_C_T  0.170460  0.066006      3  Whole_blood   \n",
       "4  ENSG00000285578.1  chr6_197265_T_C  0.829534  0.073365      3  Whole_blood   \n",
       "\n",
       "  maf  modality  \n",
       "0  1%  cis-eQTL  \n",
       "1  1%  cis-eQTL  \n",
       "2  1%  cis-eQTL  \n",
       "3  1%  cis-eQTL  \n",
       "4  1%  cis-eQTL  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CIS_EQTL_SUSIE_GLOB = glob.glob('../data/scan-results/joint/cis-eqtl/susie/maf*/Whole_blood*.cs.txt')\n",
    "CIS_SQTL_SUSIE_GLOB = glob.glob('../data/scan-results/joint/cis-sqtl/susie/maf*/postprocessed/Whole_blood*.by-gene.cs.txt')\n",
    "cis_eqtl_susie = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).split('.')[0], maf='1%' if 'maf001' in f else '0.1%') for f in CIS_EQTL_SUSIE_GLOB])\n",
    "cis_sqtl_susie = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).split('.')[0], maf='1%' if 'maf001' in f else '0.1%') for f in CIS_SQTL_SUSIE_GLOB])\n",
    "cis_susie = pd.concat([cis_eqtl_susie.assign(modality='cis-eQTL'), cis_sqtl_susie.assign(modality='cis-sQTL')])\n",
    "cis_susie.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:09:58.196536Z",
     "iopub.status.busy": "2024-12-02T19:09:58.196241Z",
     "iopub.status.idle": "2024-12-02T19:10:01.574436Z",
     "shell.execute_reply": "2024-12-02T19:10:01.573707Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>unique_cs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_180573_C_A</td>\n",
       "      <td>0.274643</td>\n",
       "      <td>0.027967</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>ENSG00000285578.1___L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_184740_C_T</td>\n",
       "      <td>0.725334</td>\n",
       "      <td>0.028355</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>ENSG00000285578.1___L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_153439_T_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.678416</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>ENSG00000285578.1___L2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_196027_C_T</td>\n",
       "      <td>0.170460</td>\n",
       "      <td>0.066006</td>\n",
       "      <td>3</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>ENSG00000285578.1___L3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>chr6_197265_T_C</td>\n",
       "      <td>0.829534</td>\n",
       "      <td>0.073365</td>\n",
       "      <td>3</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000285578.1</td>\n",
       "      <td>ENSG00000285578.1___L3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phenotype_id       variant_id       pip        af  cs_id       tissue  \\\n",
       "0  ENSG00000285578.1  chr6_180573_C_A  0.274643  0.027967      1  Whole_blood   \n",
       "1  ENSG00000285578.1  chr6_184740_C_T  0.725334  0.028355      1  Whole_blood   \n",
       "2  ENSG00000285578.1  chr6_153439_T_A  1.000000  0.678416      2  Whole_blood   \n",
       "3  ENSG00000285578.1  chr6_196027_C_T  0.170460  0.066006      3  Whole_blood   \n",
       "4  ENSG00000285578.1  chr6_197265_T_C  0.829534  0.073365      3  Whole_blood   \n",
       "\n",
       "  maf  modality            gene_id            unique_cs_id  \n",
       "0  1%  cis-eQTL  ENSG00000285578.1  ENSG00000285578.1___L1  \n",
       "1  1%  cis-eQTL  ENSG00000285578.1  ENSG00000285578.1___L1  \n",
       "2  1%  cis-eQTL  ENSG00000285578.1  ENSG00000285578.1___L2  \n",
       "3  1%  cis-eQTL  ENSG00000285578.1  ENSG00000285578.1___L3  \n",
       "4  1%  cis-eQTL  ENSG00000285578.1  ENSG00000285578.1___L3  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cis_susie['gene_id'] = cis_susie.phenotype_id.map(phenotype_id_to_gene_id)\n",
    "cis_susie['unique_cs_id'] = cis_susie.phenotype_id + '___L' + cis_susie.cs_id.astype(str)\n",
    "cis_susie.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:01.577021Z",
     "iopub.status.busy": "2024-12-02T19:10:01.576774Z",
     "iopub.status.idle": "2024-12-02T19:10:01.588713Z",
     "shell.execute_reply": "2024-12-02T19:10:01.588197Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19394"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cis_eqtl_permutations[(cis_eqtl_permutations.maf=='0.1%') & (cis_eqtl_permutations.qval<=0.05)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:01.591000Z",
     "iopub.status.busy": "2024-12-02T19:10:01.590683Z",
     "iopub.status.idle": "2024-12-02T19:10:04.623145Z",
     "shell.execute_reply": "2024-12-02T19:10:04.622472Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'While the number of cis-e/sGenes showed little change compared to a 0.01 MAF threshold (19,465 and 19,394 cis-eGenes at 0.01 and 0.001 MAF thresholds, respectively, and 8,795 vs 8,873 cis-sGenes), the number of total cis-eQTL signals detected increased by 9.7% (from 69,766 to 76,545) and the number of cis-sQTL signals increased by 9.0% (from 35,770 to 39,001)'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'While the number of cis-e/sGenes showed little change compared to a 0.01 MAF threshold ({:,} and {:,} cis-eGenes at 0.01 and 0.001 MAF thresholds, respectively, and {:,} vs {:,} cis-sGenes), the number of total cis-eQTL signals detected increased by {}% (from {:,} to {:,}) and the number of cis-sQTL signals increased by {}% (from {:,} to {:,})'.format(\n",
    "    len(cis_eqtl_permutations[(cis_eqtl_permutations.maf=='1%') & (cis_eqtl_permutations.qval<=0.05)]),\n",
    "    len(cis_eqtl_permutations[(cis_eqtl_permutations.maf=='0.1%') & (cis_eqtl_permutations.qval<=0.05)]),\n",
    "    len(cis_sqtl_permutations[(cis_sqtl_permutations.maf=='1%') & (cis_sqtl_permutations.qval<=0.05)]),\n",
    "    len(cis_sqtl_permutations[(cis_sqtl_permutations.maf=='0.1%') & (cis_sqtl_permutations.qval<=0.05)]),\n",
    "    round(100*(cis_susie[(cis_susie.modality=='cis-eQTL') & (cis_susie.maf=='0.1%')].unique_cs_id.nunique() - cis_susie[(cis_susie.modality=='cis-eQTL') & (cis_susie.maf=='1%')].unique_cs_id.nunique()) / cis_susie[(cis_susie.modality=='cis-eQTL') & (cis_susie.maf=='1%')].unique_cs_id.nunique(), 1),\n",
    "    cis_susie[(cis_susie.modality=='cis-eQTL') & (cis_susie.maf=='1%')].unique_cs_id.nunique(),\n",
    "    cis_susie[(cis_susie.modality=='cis-eQTL') & (cis_susie.maf=='0.1%')].unique_cs_id.nunique(),\n",
    "    round(100*(cis_susie[(cis_susie.modality=='cis-sQTL') & (cis_susie.maf=='0.1%')].unique_cs_id.nunique() - cis_susie[(cis_susie.modality=='cis-sQTL') & (cis_susie.maf=='1%')].unique_cs_id.nunique()) / cis_susie[(cis_susie.modality=='cis-sQTL') & (cis_susie.maf=='1%')].unique_cs_id.nunique(), 1),\n",
    "    cis_susie[(cis_susie.modality=='cis-sQTL') & (cis_susie.maf=='1%')].unique_cs_id.nunique(),\n",
    "    cis_susie[(cis_susie.modality=='cis-sQTL') & (cis_susie.maf=='0.1%')].unique_cs_id.nunique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:04.626068Z",
     "iopub.status.busy": "2024-12-02T19:10:04.625646Z",
     "iopub.status.idle": "2024-12-02T19:10:04.630517Z",
     "shell.execute_reply": "2024-12-02T19:10:04.629303Z"
    }
   },
   "outputs": [],
   "source": [
    "# # find signals specific to either scan, or shared\n",
    "# cis_eqtl_signals_at_maf_1 = cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-eQTL')].merge(compare_credible_sets(cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-eQTL')], cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-eQTL')]))\n",
    "# cis_eqtl_signals_at_maf_01 = cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-eQTL')].merge(compare_credible_sets(cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-eQTL')], cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-eQTL')]))\n",
    "# cis_sqtl_signals_at_maf_1 = cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-sQTL')].merge(compare_credible_sets(cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-sQTL')], cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-sQTL')]))\n",
    "# cis_sqtl_signals_at_maf_01 = cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-sQTL')].merge(compare_credible_sets(cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-sQTL')], cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-sQTL')]))\n",
    "# cis_sqtl_signals_at_maf_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:04.633204Z",
     "iopub.status.busy": "2024-12-02T19:10:04.632762Z",
     "iopub.status.idle": "2024-12-02T19:10:06.830082Z",
     "shell.execute_reply": "2024-12-02T19:10:06.829431Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7936530688300891\n",
      "0.8597428012300811\n"
     ]
    }
   ],
   "source": [
    "# how many CS at 1% persist (by overlap)?\n",
    "persist_by_overlap_ciseqtl = compare_credible_sets(cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-eQTL')], cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-eQTL')])\n",
    "\n",
    "cissqtl_maf1_as_gene = cis_susie.loc[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-sQTL'),['gene_id', 'variant_id', 'unique_cs_id']].rename(columns={'gene_id': 'phenotype_id', 'unique_cs_id': 'cs_id'})\n",
    "cissqtl_maf01_as_gene = cis_susie.loc[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-sQTL'),['gene_id', 'variant_id', 'unique_cs_id']].rename(columns={'gene_id': 'phenotype_id', 'unique_cs_id': 'cs_id'})\n",
    "persist_by_overlap_cissqtl = compare_credible_sets(cissqtl_maf1_as_gene, cissqtl_maf01_as_gene)\n",
    "print(persist_by_overlap_ciseqtl.in_other_scan_cs.mean())\n",
    "print(persist_by_overlap_cissqtl.in_other_scan_cs.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:06.833175Z",
     "iopub.status.busy": "2024-12-02T19:10:06.832739Z",
     "iopub.status.idle": "2024-12-02T19:10:10.612899Z",
     "shell.execute_reply": "2024-12-02T19:10:10.611162Z"
    }
   },
   "outputs": [],
   "source": [
    "top_pip_variants_ciseqtl_maf1 = top_pip_variants(cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-eQTL')])\n",
    "top_pip_variants_ciseqtl_maf01 = top_pip_variants(cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-eQTL')])\n",
    "top_pip_variants_cissqtl_maf1 = top_pip_variants(cis_susie[(cis_susie.maf=='1%') & (cis_susie.modality=='cis-sQTL')])\n",
    "top_pip_variants_cissqtl_maf01 = top_pip_variants(cis_susie[(cis_susie.maf=='0.1%') & (cis_susie.modality=='cis-sQTL')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:10.620030Z",
     "iopub.status.busy": "2024-12-02T19:10:10.619101Z",
     "iopub.status.idle": "2024-12-02T19:10:10.622996Z",
     "shell.execute_reply": "2024-12-02T19:10:10.622429Z"
    }
   },
   "outputs": [],
   "source": [
    "# another way to look at this:\n",
    "# for each signal w/ MAF < 1%; what is the best proxy MAF > 1% SNP? Was it a (now lost) CS?\n",
    "# but will not deal with this for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:10:10.627020Z",
     "iopub.status.busy": "2024-12-02T19:10:10.626450Z",
     "iopub.status.idle": "2024-12-02T19:15:11.257598Z",
     "shell.execute_reply": "2024-12-02T19:15:11.256942Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>C1</th>\n",
       "      <th>C2</th>\n",
       "      <th>G0</th>\n",
       "      <th>tissue</th>\n",
       "      <th>ancestry</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>ref_allele_count</th>\n",
       "      <th>alt_allele_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21</td>\n",
       "      <td>chr21_5090094_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>chr21_5090518_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>chr21_5090561_G_A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>chr21_5090588_C_T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21</td>\n",
       "      <td>chr21_5090666_C_A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHR                SNP A1 A2  C1  C2  G0  tissue ancestry ref_allele  \\\n",
       "0   21  chr21_5090094_C_T  T  C   0   2   0  T_cell      EAS          C   \n",
       "1   21  chr21_5090518_G_A  A  G   0   2   0  T_cell      EAS          G   \n",
       "2   21  chr21_5090561_G_A  A  G   0   2   0  T_cell      EAS          G   \n",
       "3   21  chr21_5090588_C_T  T  C   0   2   0  T_cell      EAS          C   \n",
       "4   21  chr21_5090666_C_A  A  C   0   2   0  T_cell      EAS          C   \n",
       "\n",
       "  alt_allele  ref_allele_count  alt_allele_count  \n",
       "0          T                 2                 0  \n",
       "1          A                 2                 0  \n",
       "2          A                 2                 0  \n",
       "3          T                 2                 0  \n",
       "4          A                 2                 0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLINK_MAC_FILES = glob.glob('../work/ancestry-allele-counts-75-AMR-50/cs-variants/results/mac-per-chrom/*')\n",
    "mac = pd.concat([pd.read_csv(f, delim_whitespace=True).assign(tissue=os.path.basename(f).split('.')[0], ancestry=os.path.basename(f).split('.')[1]) for f in PLINK_MAC_FILES])\n",
    "mac['ref_allele'] = mac.SNP.str.split('_', expand=True)[2]\n",
    "mac['alt_allele'] = mac.SNP.str.split('_', expand=True)[3]\n",
    "assert(all(((mac.ref_allele == mac.A1) | (mac.alt_allele == mac.A1))))\n",
    "assert(all(((mac.ref_allele == mac.A2) | (mac.alt_allele == mac.A2))))\n",
    "mac['ref_allele_count'] = np.where(mac.ref_allele == mac.A1, mac.C1, mac.C2)\n",
    "mac['alt_allele_count'] = np.where(mac.alt_allele == mac.A1, mac.C1, mac.C2)\n",
    "mac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:15:11.260301Z",
     "iopub.status.busy": "2024-12-02T19:15:11.260011Z",
     "iopub.status.idle": "2024-12-02T19:15:45.445357Z",
     "shell.execute_reply": "2024-12-02T19:15:45.444799Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ref_allele_count</th>\n",
       "      <th>alt_allele_count</th>\n",
       "      <th>tissue</th>\n",
       "      <th>ancestry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr21_5090094_C_T</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr21_5090518_G_A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr21_5090561_G_A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr21_5090588_C_T</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21_5090666_C_A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>T_cell</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID  ref_allele_count  alt_allele_count  tissue ancestry\n",
       "0  chr21_5090094_C_T                 2                 0  T_cell      EAS\n",
       "1  chr21_5090518_G_A                 2                 0  T_cell      EAS\n",
       "2  chr21_5090561_G_A                 2                 0  T_cell      EAS\n",
       "3  chr21_5090588_C_T                 2                 0  T_cell      EAS\n",
       "4  chr21_5090666_C_A                 2                 0  T_cell      EAS"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ancestry_mafs = mac[['SNP', 'ref_allele_count', 'alt_allele_count', 'tissue', 'ancestry']].rename(columns={'SNP': 'ID'})\n",
    "ancestry_mafs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:15:45.447774Z",
     "iopub.status.busy": "2024-12-02T19:15:45.447504Z",
     "iopub.status.idle": "2024-12-02T19:15:53.973849Z",
     "shell.execute_reply": "2024-12-02T19:15:53.973186Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3183629/1775847148.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  whole_blood_ancestry_mafs['af'] = whole_blood_ancestry_mafs.alt_allele_count / whole_blood_ancestry_mafs[['alt_allele_count', 'ref_allele_count']].sum(axis=1)\n",
      "/tmp/ipykernel_3183629/1775847148.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  whole_blood_ancestry_mafs['maf'] = np.minimum(whole_blood_ancestry_mafs.af, 1-whole_blood_ancestry_mafs.af)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ancestry</th>\n",
       "      <th>AFR_MAF</th>\n",
       "      <th>AMR_MAF</th>\n",
       "      <th>EUR_MAF</th>\n",
       "      <th>variant_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10_100004256_T_A</th>\n",
       "      <td>0.405914</td>\n",
       "      <td>0.381098</td>\n",
       "      <td>0.379968</td>\n",
       "      <td>chr10_100004256_T_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100006605_T_C</th>\n",
       "      <td>0.137097</td>\n",
       "      <td>0.059959</td>\n",
       "      <td>0.081230</td>\n",
       "      <td>chr10_100006605_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100007810_T_A</th>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001735</td>\n",
       "      <td>chr10_100007810_T_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100007968_C_T</th>\n",
       "      <td>0.137634</td>\n",
       "      <td>0.059959</td>\n",
       "      <td>0.081073</td>\n",
       "      <td>chr10_100007968_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100014566_A_G</th>\n",
       "      <td>0.170968</td>\n",
       "      <td>0.108740</td>\n",
       "      <td>0.028707</td>\n",
       "      <td>chr10_100014566_A_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "ancestry              AFR_MAF   AMR_MAF   EUR_MAF           variant_id\n",
       "ID                                                                    \n",
       "chr10_100004256_T_A  0.405914  0.381098  0.379968  chr10_100004256_T_A\n",
       "chr10_100006605_T_C  0.137097  0.059959  0.081230  chr10_100006605_T_C\n",
       "chr10_100007810_T_A  0.000538  0.000000  0.001735  chr10_100007810_T_A\n",
       "chr10_100007968_C_T  0.137634  0.059959  0.081073  chr10_100007968_C_T\n",
       "chr10_100014566_A_G  0.170968  0.108740  0.028707  chr10_100014566_A_G"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_blood_ancestry_mafs = ancestry_mafs[(ancestry_mafs.tissue=='Whole_blood') & (ancestry_mafs.ancestry.isin(['AMR', 'AFR', 'EUR']))]\n",
    "whole_blood_ancestry_mafs['af'] = whole_blood_ancestry_mafs.alt_allele_count / whole_blood_ancestry_mafs[['alt_allele_count', 'ref_allele_count']].sum(axis=1)\n",
    "whole_blood_ancestry_mafs['maf'] = np.minimum(whole_blood_ancestry_mafs.af, 1-whole_blood_ancestry_mafs.af)\n",
    "whole_blood_ancestry_mafs = whole_blood_ancestry_mafs[['ID', 'ancestry', 'maf']].pivot(index='ID', columns='ancestry', values='maf')\n",
    "whole_blood_ancestry_mafs = whole_blood_ancestry_mafs.rename(columns=lambda x: x + '_MAF')\n",
    "whole_blood_ancestry_mafs['variant_id'] = whole_blood_ancestry_mafs.index.to_list()\n",
    "whole_blood_ancestry_mafs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:15:53.976152Z",
     "iopub.status.busy": "2024-12-02T19:15:53.975895Z",
     "iopub.status.idle": "2024-12-02T19:15:54.862898Z",
     "shell.execute_reply": "2024-12-02T19:15:54.862140Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>unique_cs_id</th>\n",
       "      <th>AFR_MAF</th>\n",
       "      <th>AMR_MAF</th>\n",
       "      <th>EUR_MAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.964002</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.967864</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:100286332:100286605:clu_5974_-:ENSG00000...</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.973766</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-sQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10:100286332:100286605:clu_5974_-:ENSG00000...</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        phenotype_id           variant_id  \\\n",
       "0                                 ENSG00000196072.12  chr10_100278884_G_A   \n",
       "1                                 ENSG00000107566.14  chr10_100278884_G_A   \n",
       "2                                 ENSG00000196072.12  chr10_100278884_G_A   \n",
       "3                                 ENSG00000107566.14  chr10_100278884_G_A   \n",
       "4  chr10:100286332:100286605:clu_5974_-:ENSG00000...  chr10_100278884_G_A   \n",
       "\n",
       "        pip        af  cs_id       tissue   maf  modality             gene_id  \\\n",
       "0  1.000000  0.363108      1  Whole_blood    1%  cis-eQTL  ENSG00000196072.12   \n",
       "1  0.964002  0.363108      2  Whole_blood    1%  cis-eQTL  ENSG00000107566.14   \n",
       "2  1.000000  0.363108      1  Whole_blood  0.1%  cis-eQTL  ENSG00000196072.12   \n",
       "3  0.967864  0.363108      2  Whole_blood  0.1%  cis-eQTL  ENSG00000107566.14   \n",
       "4  0.973766  0.363108      1  Whole_blood    1%  cis-sQTL  ENSG00000196072.12   \n",
       "\n",
       "                                        unique_cs_id  AFR_MAF   AMR_MAF  \\\n",
       "0                            ENSG00000196072.12___L1  0.23172  0.231707   \n",
       "1                            ENSG00000107566.14___L2  0.23172  0.231707   \n",
       "2                            ENSG00000196072.12___L1  0.23172  0.231707   \n",
       "3                            ENSG00000107566.14___L2  0.23172  0.231707   \n",
       "4  chr10:100286332:100286605:clu_5974_-:ENSG00000...  0.23172  0.231707   \n",
       "\n",
       "    EUR_MAF  \n",
       "0  0.444479  \n",
       "1  0.444479  \n",
       "2  0.444479  \n",
       "3  0.444479  \n",
       "4  0.444479  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_pip_variants_combined = pd.concat([top_pip_variants_ciseqtl_maf1, top_pip_variants_ciseqtl_maf01, top_pip_variants_cissqtl_maf1, top_pip_variants_cissqtl_maf01])\n",
    "top_pip_variants_combined = top_pip_variants_combined.merge(whole_blood_ancestry_mafs)\n",
    "top_pip_variants_combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:15:54.865382Z",
     "iopub.status.busy": "2024-12-02T19:15:54.864971Z",
     "iopub.status.idle": "2024-12-02T19:15:54.889442Z",
     "shell.execute_reply": "2024-12-02T19:15:54.889001Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>unique_cs_id</th>\n",
       "      <th>AFR_MAF</th>\n",
       "      <th>AMR_MAF</th>\n",
       "      <th>EUR_MAF</th>\n",
       "      <th>variant_maf</th>\n",
       "      <th>max_maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.964002</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.967864</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:100286332:100286605:clu_5974_-:ENSG00000...</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.973766</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-sQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10:100286332:100286605:clu_5974_-:ENSG00000...</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        phenotype_id           variant_id  \\\n",
       "0                                 ENSG00000196072.12  chr10_100278884_G_A   \n",
       "1                                 ENSG00000107566.14  chr10_100278884_G_A   \n",
       "2                                 ENSG00000196072.12  chr10_100278884_G_A   \n",
       "3                                 ENSG00000107566.14  chr10_100278884_G_A   \n",
       "4  chr10:100286332:100286605:clu_5974_-:ENSG00000...  chr10_100278884_G_A   \n",
       "\n",
       "        pip        af  cs_id       tissue   maf  modality             gene_id  \\\n",
       "0  1.000000  0.363108      1  Whole_blood    1%  cis-eQTL  ENSG00000196072.12   \n",
       "1  0.964002  0.363108      2  Whole_blood    1%  cis-eQTL  ENSG00000107566.14   \n",
       "2  1.000000  0.363108      1  Whole_blood  0.1%  cis-eQTL  ENSG00000196072.12   \n",
       "3  0.967864  0.363108      2  Whole_blood  0.1%  cis-eQTL  ENSG00000107566.14   \n",
       "4  0.973766  0.363108      1  Whole_blood    1%  cis-sQTL  ENSG00000196072.12   \n",
       "\n",
       "                                        unique_cs_id  AFR_MAF   AMR_MAF  \\\n",
       "0                            ENSG00000196072.12___L1  0.23172  0.231707   \n",
       "1                            ENSG00000107566.14___L2  0.23172  0.231707   \n",
       "2                            ENSG00000196072.12___L1  0.23172  0.231707   \n",
       "3                            ENSG00000107566.14___L2  0.23172  0.231707   \n",
       "4  chr10:100286332:100286605:clu_5974_-:ENSG00000...  0.23172  0.231707   \n",
       "\n",
       "    EUR_MAF  variant_maf   max_maf  \n",
       "0  0.444479     0.363108  0.444479  \n",
       "1  0.444479     0.363108  0.444479  \n",
       "2  0.444479     0.363108  0.444479  \n",
       "3  0.444479     0.363108  0.444479  \n",
       "4  0.444479     0.363108  0.444479  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_pip_variants_combined['variant_maf'] = np.minimum(top_pip_variants_combined.af, 1 - top_pip_variants_combined.af)\n",
    "top_pip_variants_combined['max_maf'] = top_pip_variants_combined[['EUR_MAF', 'AFR_MAF', 'AMR_MAF']].max(axis=1)\n",
    "top_pip_variants_combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:15:54.891531Z",
     "iopub.status.busy": "2024-12-02T19:15:54.891311Z",
     "iopub.status.idle": "2024-12-02T19:16:29.119618Z",
     "shell.execute_reply": "2024-12-02T19:16:29.118217Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>unique_cs_id</th>\n",
       "      <th>AFR_MAF</th>\n",
       "      <th>AMR_MAF</th>\n",
       "      <th>EUR_MAF</th>\n",
       "      <th>variant_maf</th>\n",
       "      <th>max_maf</th>\n",
       "      <th>driven_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.964002</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>ENSG00000196072.12___L1</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.967864</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>2</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.1%</td>\n",
       "      <td>cis-eQTL</td>\n",
       "      <td>ENSG00000107566.14</td>\n",
       "      <td>ENSG00000107566.14___L2</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:100286332:100286605:clu_5974_-:ENSG00000...</td>\n",
       "      <td>chr10_100278884_G_A</td>\n",
       "      <td>0.973766</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>1</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>cis-sQTL</td>\n",
       "      <td>ENSG00000196072.12</td>\n",
       "      <td>chr10:100286332:100286605:clu_5974_-:ENSG00000...</td>\n",
       "      <td>0.23172</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>0.363108</td>\n",
       "      <td>0.444479</td>\n",
       "      <td>AFR,AMR,EUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        phenotype_id           variant_id  \\\n",
       "0                                 ENSG00000196072.12  chr10_100278884_G_A   \n",
       "1                                 ENSG00000107566.14  chr10_100278884_G_A   \n",
       "2                                 ENSG00000196072.12  chr10_100278884_G_A   \n",
       "3                                 ENSG00000107566.14  chr10_100278884_G_A   \n",
       "4  chr10:100286332:100286605:clu_5974_-:ENSG00000...  chr10_100278884_G_A   \n",
       "\n",
       "        pip        af  cs_id       tissue   maf  modality             gene_id  \\\n",
       "0  1.000000  0.363108      1  Whole_blood    1%  cis-eQTL  ENSG00000196072.12   \n",
       "1  0.964002  0.363108      2  Whole_blood    1%  cis-eQTL  ENSG00000107566.14   \n",
       "2  1.000000  0.363108      1  Whole_blood  0.1%  cis-eQTL  ENSG00000196072.12   \n",
       "3  0.967864  0.363108      2  Whole_blood  0.1%  cis-eQTL  ENSG00000107566.14   \n",
       "4  0.973766  0.363108      1  Whole_blood    1%  cis-sQTL  ENSG00000196072.12   \n",
       "\n",
       "                                        unique_cs_id  AFR_MAF   AMR_MAF  \\\n",
       "0                            ENSG00000196072.12___L1  0.23172  0.231707   \n",
       "1                            ENSG00000107566.14___L2  0.23172  0.231707   \n",
       "2                            ENSG00000196072.12___L1  0.23172  0.231707   \n",
       "3                            ENSG00000107566.14___L2  0.23172  0.231707   \n",
       "4  chr10:100286332:100286605:clu_5974_-:ENSG00000...  0.23172  0.231707   \n",
       "\n",
       "    EUR_MAF  variant_maf   max_maf    driven_by  \n",
       "0  0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "1  0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "2  0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "3  0.444479     0.363108  0.444479  AFR,AMR,EUR  \n",
       "4  0.444479     0.363108  0.444479  AFR,AMR,EUR  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# try to determine which ancestries drive signals\n",
    "ANCESTRY_MAF_COLS = ['AFR_MAF', 'AMR_MAF', 'EUR_MAF']\n",
    "top_pip_variants_combined['driven_by'] = top_pip_variants_combined[ANCESTRY_MAF_COLS].apply(lambda x: ','.join(x[x/max(x)>=0.2].index.to_list()).replace('_MAF', ''), axis=1)\n",
    "top_pip_variants_combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:16:29.123619Z",
     "iopub.status.busy": "2024-12-02T19:16:29.123355Z",
     "iopub.status.idle": "2024-12-02T19:16:29.159280Z",
     "shell.execute_reply": "2024-12-02T19:16:29.158381Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "maf   modality\n",
       "0.1%  cis-eQTL    0.211862\n",
       "      cis-sQTL    0.187226\n",
       "1%    cis-eQTL    0.000000\n",
       "      cis-sQTL    0.000000\n",
       "Name: variant_maf, dtype: float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_pip_variants_combined.groupby(['maf', 'modality']).variant_maf.apply(lambda x: (x<0.01).mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:16:29.162480Z",
     "iopub.status.busy": "2024-12-02T19:16:29.162166Z",
     "iopub.status.idle": "2024-12-02T19:16:29.292980Z",
     "shell.execute_reply": "2024-12-02T19:16:29.292517Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "76545"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(top_pip_variants_combined[(top_pip_variants_combined.maf=='0.1%') & (top_pip_variants_combined.modality=='cis-eQTL')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:16:29.297427Z",
     "iopub.status.busy": "2024-12-02T19:16:29.296673Z",
     "iopub.status.idle": "2024-12-02T19:16:29.376273Z",
     "shell.execute_reply": "2024-12-02T19:16:29.375781Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The top posterior inclusion probability (PIP) variant in 16,217 cis-eQTL and 7,302 cis-sQTL credible sets had MAF < 0.01\n"
     ]
    }
   ],
   "source": [
    "print('The top posterior inclusion probability (PIP) variant in {:,} cis-eQTL and {:,} cis-sQTL credible sets had MAF < 0.01'.format(\n",
    "    (top_pip_variants_combined[(top_pip_variants_combined.maf=='0.1%') & (top_pip_variants_combined.modality=='cis-eQTL')].variant_maf<0.01).sum(),\n",
    "    (top_pip_variants_combined[(top_pip_variants_combined.maf=='0.1%') & (top_pip_variants_combined.modality=='cis-sQTL')].variant_maf<0.01).sum()\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:16:29.381099Z",
     "iopub.status.busy": "2024-12-02T19:16:29.380560Z",
     "iopub.status.idle": "2024-12-02T19:16:29.466125Z",
     "shell.execute_reply": "2024-12-02T19:16:29.465274Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7,479 cis-eQTL and 3,290 cis-sQTL signals had MAF < 0.01 in all three well-represented ancestries\n"
     ]
    }
   ],
   "source": [
    "print('{:,} cis-eQTL and {:,} cis-sQTL signals had MAF < 0.01 in all three well-represented ancestries'.format(\n",
    "    (top_pip_variants_combined[(top_pip_variants_combined.maf=='0.1%') & (top_pip_variants_combined.modality=='cis-eQTL')].max_maf<0.01).sum(),\n",
    "    (top_pip_variants_combined[(top_pip_variants_combined.maf=='0.1%') & (top_pip_variants_combined.modality=='cis-sQTL')].max_maf<0.01).sum()\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-02T19:16:29.468873Z",
     "iopub.status.busy": "2024-12-02T19:16:29.468412Z",
     "iopub.status.idle": "2024-12-02T19:16:31.955172Z",
     "shell.execute_reply": "2024-12-02T19:16:31.954308Z"
    }
   },
   "outputs": [],
   "source": [
    "top_pip_variants_combined.to_csv(f'tables/{PREFIX}maf-breakdown.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2020-09",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
