{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_180573_C_A</td>\n",
       "      <td>0.274643</td>\n",
       "      <td>0.027967</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_184740_C_T</td>\n",
       "      <td>0.725334</td>\n",
       "      <td>0.028355</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_153439_T_A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.678416</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_196027_C_T</td>\n",
       "      <td>0.170460</td>\n",
       "      <td>0.066006</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_197265_T_C</td>\n",
       "      <td>0.829534</td>\n",
       "      <td>0.073365</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id       variant_id       pip        af  cs_id\n",
       "0  ENSG00000285578  chr6_180573_C_A  0.274643  0.027967      1\n",
       "1  ENSG00000285578  chr6_184740_C_T  0.725334  0.028355      1\n",
       "2  ENSG00000285578  chr6_153439_T_A  1.000000  0.678416      2\n",
       "3  ENSG00000285578  chr6_196027_C_T  0.170460  0.066006      3\n",
       "4  ENSG00000285578  chr6_197265_T_C  0.829534  0.073365      3"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "PREFIX = 'signal-in-topmed-gtex-eqtlgen-direct.'\n",
    "\n",
    "def compare_credible_sets(scan_1_cs, scan_2_cs, summarize=True):\n",
    "    \"\"\"\n",
    "    Given a dataframe representing scan_1_cs and scan_2_cs (each having columns ['phenotype_id', 'variant_id', 'cs_id']; can have other columns too),\n",
    "    return a dataframe showing, for each CS in scan_1_cs, whether it overlaps a CS in scan_2_cs (for the same phenotype_id)\n",
    "    If summarize = False, simply returns scan_1_cs with an added column indicating whether each credible set SNP is a credible set SNP \n",
    "    for the same phenotype in the other scan\n",
    "    \"\"\"\n",
    "    # Validate input\n",
    "    if not isinstance(scan_1_cs, pd.DataFrame):\n",
    "        raise TypeError('scan_1_cs must be a DataFrame')\n",
    "    if not isinstance(scan_2_cs, pd.DataFrame):\n",
    "        raise TypeError('scan_2_cs must be a DataFrame')\n",
    "    for i in ['phenotype_id', 'variant_id', 'cs_id']:\n",
    "        if not i in scan_1_cs.columns.to_list():\n",
    "            raise ValueError(f'scan_1_cs must include column {i}')\n",
    "        if not i in scan_2_cs.columns.to_list():\n",
    "            raise ValueError(f'scan_2_cs must include column {i}')\n",
    "\n",
    "    results = scan_1_cs.merge(scan_2_cs[['phenotype_id', 'variant_id']].drop_duplicates().assign(in_other_scan_cs=1), how='left')\n",
    "    results.in_other_scan_cs = results.in_other_scan_cs.fillna(0).astype(int)\n",
    "    assert(len(results) == len(scan_1_cs))\n",
    "\n",
    "    if summarize:\n",
    "        return results.groupby(['phenotype_id', 'cs_id']).in_other_scan_cs.max().reset_index()\n",
    "    else:\n",
    "        return results\n",
    "\n",
    "\n",
    "USE_LD = True\n",
    "\n",
    "topmed_whole_blood_ciseqtl = pd.read_csv('../data/scan-results/joint/cis-eqtl/susie/maf001/Whole_blood.100.cs.txt', sep='\\t')\n",
    "gtex_whole_blood_ciseqtl = pd.read_parquet('/net/topmed11/working/porchard/gtex-preprocessing/data/gtex/GTEx_v8_finemapping_SuSiE/GTEx_v8.Whole_Blood.eQTL.SuSiE_summary.parquet')\n",
    "topmed_whole_blood_ciseqtl.phenotype_id = topmed_whole_blood_ciseqtl.phenotype_id.str.split('.', expand=True)[0]\n",
    "gtex_whole_blood_ciseqtl.phenotype_id = gtex_whole_blood_ciseqtl.phenotype_id.str.split('.', expand=True)[0]\n",
    "gtex_whole_blood_ciseqtl.variant_id = gtex_whole_blood_ciseqtl.variant_id.str.replace('_b38', '')\n",
    "topmed_whole_blood_ciseqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>cs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187642</td>\n",
       "      <td>chr1_953778_G_C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000187608</td>\n",
       "      <td>chr1_1015336_A_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000187583</td>\n",
       "      <td>chr1_953778_G_C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000188290</td>\n",
       "      <td>chr1_995543_A_G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>chr1_1002736_T_G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id        variant_id  cs_id\n",
       "0  ENSG00000187642   chr1_953778_G_C      1\n",
       "1  ENSG00000187608  chr1_1015336_A_T      1\n",
       "2  ENSG00000187583   chr1_953778_G_C      1\n",
       "3  ENSG00000188290   chr1_995543_A_G      1\n",
       "4  ENSG00000188157  chr1_1002736_T_G      1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reshape the eQTLGen primary signals as if they were credible sets\n",
    "eqtlgen_whole_blood_ciseqtl = pd.read_csv('/net/topmed11/working/porchard/eqtlgen-preprocessing/work/top-hit-per-gene-after-lifting/top-per-gene.txt', sep='\\t')\n",
    "eqtlgen_whole_blood_ciseqtl = eqtlgen_whole_blood_ciseqtl[eqtlgen_whole_blood_ciseqtl.FDR<=0.05]\n",
    "eqtlgen_whole_blood_ciseqtl = eqtlgen_whole_blood_ciseqtl[['Gene', 'SNP']]\n",
    "eqtlgen_whole_blood_ciseqtl.columns = ['phenotype_id', 'variant_id']\n",
    "eqtlgen_whole_blood_ciseqtl['cs_id'] = 1\n",
    "eqtlgen_whole_blood_ciseqtl.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-3-19f11488fec4>:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  direct_whole_blood_ciseqtl = pd.read_csv('/net/topmed11/working/porchard/direct-preprocessing/work/lift-and-tabix/cis-eqtl-significant/results/tabixed/direct.txt.gz', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>cs_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000272512</td>\n",
       "      <td>chr1_633714_A_G</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000235373</td>\n",
       "      <td>chr1_757414_CA_C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000160087</td>\n",
       "      <td>chr1_771398_G_A</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>chr1_791564_C_G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000228327</td>\n",
       "      <td>chr1_800909_T_A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id        variant_id  cs_id\n",
       "0  ENSG00000272512   chr1_633714_A_G      6\n",
       "1  ENSG00000235373  chr1_757414_CA_C      2\n",
       "2  ENSG00000160087   chr1_771398_G_A      3\n",
       "3  ENSG00000237491   chr1_791564_C_G      2\n",
       "4  ENSG00000228327   chr1_800909_T_A      1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reshape the DIRECT conditional hits as if they were credible sets\n",
    "direct_whole_blood_ciseqtl = pd.read_csv('/net/topmed11/working/porchard/direct-preprocessing/work/lift-and-tabix/cis-eqtl-significant/results/tabixed/direct.txt.gz', sep='\\t')\n",
    "direct_whole_blood_ciseqtl = direct_whole_blood_ciseqtl[['GeneID', 'SNPid', 'DiscoveryOrder']]\n",
    "direct_whole_blood_ciseqtl.GeneID = direct_whole_blood_ciseqtl.GeneID.str.split('.', expand=True)[0]\n",
    "assert(direct_whole_blood_ciseqtl.groupby(['GeneID', 'DiscoveryOrder']).size().max() == 1)\n",
    "direct_whole_blood_ciseqtl = direct_whole_blood_ciseqtl.rename(columns={'DiscoveryOrder': 'cs_id', 'GeneID': 'phenotype_id', 'SNPid': 'variant_id'})\n",
    "direct_whole_blood_ciseqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "59828"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(direct_whole_blood_ciseqtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ld_buddies = pd.concat([pd.read_csv(f, delim_whitespace=True) for f in glob.glob('../work/eqtlgen-and-direct-ld-buddies/results/ld/*')])\n",
    "ld_buddies = ld_buddies[['SNP_A', 'SNP_B', 'R2']].drop_duplicates()\n",
    "r2_threshold = 0.8\n",
    "assert(r2_threshold >= ld_buddies.R2.min())\n",
    "ld_buddies = ld_buddies[ld_buddies.R2>=r2_threshold]\n",
    "ld_buddies = ld_buddies[ld_buddies.SNP_A!=ld_buddies.SNP_B]\n",
    "ld_buddies_dict = {variant: set() for variant in (ld_buddies.SNP_A.to_list() + ld_buddies.SNP_B.to_list())} # variant --> ld_buddies\n",
    "for row in ld_buddies.itertuples():\n",
    "    ld_buddies_dict[row.SNP_A].add(row.SNP_B)\n",
    "    ld_buddies_dict[row.SNP_B].add(row.SNP_A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expand_cs(cs, ld_buddies_dict):\n",
    "    new_cs = []\n",
    "    for phenotype_id, variant_id, cs_id in zip(cs.phenotype_id, cs.variant_id, cs.cs_id):\n",
    "        new_cs.append([phenotype_id, variant_id, cs_id])\n",
    "        if variant_id in ld_buddies_dict:\n",
    "            for buddy in ld_buddies_dict[variant_id]:\n",
    "                new_cs.append([phenotype_id, buddy, cs_id])\n",
    "    return pd.DataFrame(new_cs, columns=['phenotype_id', 'variant_id', 'cs_id']) \n",
    "\n",
    "eqtlgen_whole_blood_ciseqtl_expanded = expand_cs(eqtlgen_whole_blood_ciseqtl, ld_buddies_dict)\n",
    "direct_whole_blood_ciseqtl_expanded = expand_cs(direct_whole_blood_ciseqtl, ld_buddies_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(compare_credible_sets(eqtlgen_whole_blood_ciseqtl_expanded, eqtlgen_whole_blood_ciseqtl).in_other_scan_cs.min() == 1)\n",
    "assert(compare_credible_sets(eqtlgen_whole_blood_ciseqtl, eqtlgen_whole_blood_ciseqtl_expanded).in_other_scan_cs.min() == 1)\n",
    "assert(compare_credible_sets(direct_whole_blood_ciseqtl_expanded, direct_whole_blood_ciseqtl).in_other_scan_cs.min() == 1)\n",
    "assert(compare_credible_sets(direct_whole_blood_ciseqtl, direct_whole_blood_ciseqtl_expanded).in_other_scan_cs.min() == 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_180573_C_A</td>\n",
       "      <td>1</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_184740_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_153439_T_A</td>\n",
       "      <td>2</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_196027_C_T</td>\n",
       "      <td>3</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000285578</td>\n",
       "      <td>chr6_197265_T_C</td>\n",
       "      <td>3</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id       variant_id cs_id     src\n",
       "0  ENSG00000285578  chr6_180573_C_A     1  TOPMed\n",
       "1  ENSG00000285578  chr6_184740_C_T     1  TOPMed\n",
       "2  ENSG00000285578  chr6_153439_T_A     2  TOPMed\n",
       "3  ENSG00000285578  chr6_196027_C_T     3  TOPMed\n",
       "4  ENSG00000285578  chr6_197265_T_C     3  TOPMed"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if USE_LD:\n",
    "    combined = pd.concat([topmed_whole_blood_ciseqtl.drop(columns=['pip', 'af']).assign(src='TOPMed'), gtex_whole_blood_ciseqtl.drop(columns=['pip', 'af']).assign(src='GTEx'), eqtlgen_whole_blood_ciseqtl_expanded.assign(src='eQTLGen'), direct_whole_blood_ciseqtl_expanded.assign(src='DIRECT')])\n",
    "else:\n",
    "    combined = pd.concat([topmed_whole_blood_ciseqtl.drop(columns=['pip', 'af']).assign(src='TOPMed'), gtex_whole_blood_ciseqtl.drop(columns=['pip', 'af']).assign(src='GTEx'), eqtlgen_whole_blood_ciseqtl.assign(src='eQTLGen'), direct_whole_blood_ciseqtl.assign(src='DIRECT')])\n",
    "combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "comparisons = [] # source 1, source 2, fraction from source 1 in source 2\n",
    "comparisons_against_all = [] # source, fraction from source 1 in any of the other sources\n",
    "comparisons_against_all_full = [] # source, fraction from source 1 in any of the other sources\n",
    "sources = combined.src.unique()\n",
    "\n",
    "for src1 in sources:\n",
    "    source_1_data = combined[combined.src==src1]\n",
    "    other_source_data = combined[combined.src!=src1]\n",
    "    comparisons_against_all.append([src1, compare_credible_sets(source_1_data, other_source_data).in_other_scan_cs.mean()])\n",
    "    comparisons_against_all_full.append(compare_credible_sets(source_1_data, other_source_data).assign(src=src1))\n",
    "    for src2, source_2_data in other_source_data.groupby('src'):\n",
    "        src1_in_src2 = compare_credible_sets(source_1_data, source_2_data)\n",
    "        comparisons.append([src1, src2, src1_in_src2.in_other_scan_cs.mean()])\n",
    "    \n",
    "comparisons = pd.DataFrame(comparisons, columns=['source_1', 'source_2', 'fraction_cs_in_source_2'])\n",
    "comparisons_against_all = pd.DataFrame(comparisons_against_all, columns=['source', 'fraction_cs_in_other_sources'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TOPMed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      phenotype_id cs_id  in_other_scan_cs     src\n",
       "0  ENSG00000000003     1                 1  TOPMed\n",
       "1  ENSG00000000003     2                 1  TOPMed\n",
       "2  ENSG00000000419     1                 0  TOPMed\n",
       "3  ENSG00000000419     2                 0  TOPMed\n",
       "4  ENSG00000000457     1                 1  TOPMed"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comparisons_against_all_full = pd.concat(comparisons_against_all_full)\n",
    "comparisons_against_all_full.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    51037\n",
      "1    18729\n",
      "Name: in_other_scan_cs, dtype: int64\n",
      "69766\n"
     ]
    }
   ],
   "source": [
    "print(comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts())\n",
    "print(len(comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "comparisons_against_all_full.loc[comparisons_against_all_full.src=='TOPMed',['phenotype_id', 'cs_id', 'in_other_scan_cs']].to_csv(f'{PREFIX}topmed-signal-in-gtex-eqtlgen-direct.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>source_2</th>\n",
       "      <th>DIRECT</th>\n",
       "      <th>GTEx</th>\n",
       "      <th>TOPMed</th>\n",
       "      <th>eQTLGen</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source_1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DIRECT</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.081066</td>\n",
       "      <td>0.204469</td>\n",
       "      <td>0.084124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEx</th>\n",
       "      <td>0.407709</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.646079</td>\n",
       "      <td>0.282410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOPMed</th>\n",
       "      <td>0.173810</td>\n",
       "      <td>0.107201</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.095806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eQTLGen</th>\n",
       "      <td>0.287455</td>\n",
       "      <td>0.187614</td>\n",
       "      <td>0.388238</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "source_2    DIRECT      GTEx    TOPMed   eQTLGen\n",
       "source_1                                        \n",
       "DIRECT    1.000000  0.081066  0.204469  0.084124\n",
       "GTEx      0.407709  1.000000  0.646079  0.282410\n",
       "TOPMed    0.173810  0.107201  1.000000  0.095806\n",
       "eQTLGen   0.287455  0.187614  0.388238  1.000000"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comparisons.pivot(index='source_1', columns='source_2', values='fraction_cs_in_source_2').fillna(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>fraction_cs_in_other_sources</th>\n",
       "      <th>fraction_not_in_other_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TOPMed</td>\n",
       "      <td>0.268455</td>\n",
       "      <td>0.731545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEx</td>\n",
       "      <td>0.756580</td>\n",
       "      <td>0.243420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>eQTLGen</td>\n",
       "      <td>0.512745</td>\n",
       "      <td>0.487255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DIRECT</td>\n",
       "      <td>0.244284</td>\n",
       "      <td>0.755716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    source  fraction_cs_in_other_sources  fraction_not_in_other_sources\n",
       "0   TOPMed                      0.268455                       0.731545\n",
       "1     GTEx                      0.756580                       0.243420\n",
       "2  eQTLGen                      0.512745                       0.487255\n",
       "3   DIRECT                      0.244284                       0.755716"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comparisons_against_all['fraction_not_in_other_sources'] = 1 - comparisons_against_all.fraction_cs_in_other_sources\n",
    "comparisons_against_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16987"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(eqtlgen_whole_blood_ciseqtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_modality</th>\n",
       "      <th>xqtl_tissue</th>\n",
       "      <th>xqtl_maf</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2162</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.663501e-14</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.366463e-15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000162777</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000162777.17</td>\n",
       "      <td>chr1_110947722_111447723</td>\n",
       "      <td>EUR</td>\n",
       "      <td>biomarkers-30750-both_sexes-irnt</td>\n",
       "      <td>biomarkers-30750-both_sexes-irnt___EUR___chr1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>6415</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.198827e-177</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000140416</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000140416.21</td>\n",
       "      <td>chr15_62392636_63970660</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30100-both_sexes-irnt</td>\n",
       "      <td>continuous-30100-both_sexes-irnt___EUR___chr15...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>2026</td>\n",
       "      <td>chr1_159205564_G_A</td>\n",
       "      <td>chr1_159205564_G_A</td>\n",
       "      <td>1.501781e-170</td>\n",
       "      <td>9.969781e-52</td>\n",
       "      <td>3.012665e-122</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000213088.11</td>\n",
       "      <td>chr1_158955703_159455704</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30130-both_sexes-irnt</td>\n",
       "      <td>continuous-30130-both_sexes-irnt___EUR___chr1_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1562</td>\n",
       "      <td>chr6_396321_C_T</td>\n",
       "      <td>chr6_396321_C_T</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>9.685085e-299</td>\n",
       "      <td>6.254239e-167</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137265</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137265.15</td>\n",
       "      <td>chr6_146320_646321</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-1757-both_sexes</td>\n",
       "      <td>continuous-1757-both_sexes___EUR___chr6_146320...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>3868</td>\n",
       "      <td>chr7_28237488_C_G</td>\n",
       "      <td>chr7_28237488_C_G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.197983e-60</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.638423e-14</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000153814</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000153814.13</td>\n",
       "      <td>chr7_27987487_28934757</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30180-both_sexes-irnt</td>\n",
       "      <td>continuous-30180-both_sexes-irnt___EUR___chr7_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    nsnps                hit1                hit2      PP.H0.abf  \\\n",
       "9    2162  chr1_111197723_C_G  chr1_111197723_C_G   0.000000e+00   \n",
       "26   6415  chr15_63049797_T_C  chr15_63049797_T_C   0.000000e+00   \n",
       "30   2026  chr1_159205564_G_A  chr1_159205564_G_A  1.501781e-170   \n",
       "33   1562     chr6_396321_C_T     chr6_396321_C_T   0.000000e+00   \n",
       "43   3868   chr7_28237488_C_G   chr7_28237488_C_G   0.000000e+00   \n",
       "\n",
       "        PP.H1.abf      PP.H2.abf     PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "9    2.663501e-14   0.000000e+00  6.366463e-15        1.0     4     1  ...   \n",
       "26  3.198827e-177   0.000000e+00  0.000000e+00        1.0     4     1  ...   \n",
       "30   9.969781e-52  3.012665e-122  0.000000e+00        1.0     2     2  ...   \n",
       "33  9.685085e-299  6.254239e-167  0.000000e+00        1.0     1     1  ...   \n",
       "43   1.197983e-60   0.000000e+00  4.638423e-14        1.0     1     2  ...   \n",
       "\n",
       "   xqtl_modality  xqtl_tissue  xqtl_maf   xqtl_phenotype xqtl_ancestry  \\\n",
       "9        ciseqtl  Whole_blood        1%  ENSG00000162777         joint   \n",
       "26       ciseqtl  Whole_blood        1%  ENSG00000140416         joint   \n",
       "30       ciseqtl  Whole_blood        1%  ENSG00000213088         joint   \n",
       "33       ciseqtl  Whole_blood        1%  ENSG00000137265         joint   \n",
       "43       ciseqtl  Whole_blood        1%  ENSG00000153814         joint   \n",
       "\n",
       "             xqtl_gene               gwas_region  gwas_ancestry  \\\n",
       "9   ENSG00000162777.17  chr1_110947722_111447723            EUR   \n",
       "26  ENSG00000140416.21   chr15_62392636_63970660            EUR   \n",
       "30  ENSG00000213088.11  chr1_158955703_159455704            EUR   \n",
       "33  ENSG00000137265.15        chr6_146320_646321            EUR   \n",
       "43  ENSG00000153814.13    chr7_27987487_28934757            EUR   \n",
       "\n",
       "                          gwas_trait  \\\n",
       "9   biomarkers-30750-both_sexes-irnt   \n",
       "26  continuous-30100-both_sexes-irnt   \n",
       "30  continuous-30130-both_sexes-irnt   \n",
       "33        continuous-1757-both_sexes   \n",
       "43  continuous-30180-both_sexes-irnt   \n",
       "\n",
       "                                          gwas_signal  \n",
       "9   biomarkers-30750-both_sexes-irnt___EUR___chr1_...  \n",
       "26  continuous-30100-both_sexes-irnt___EUR___chr15...  \n",
       "30  continuous-30130-both_sexes-irnt___EUR___chr1_...  \n",
       "33  continuous-1757-both_sexes___EUR___chr6_146320...  \n",
       "43  continuous-30180-both_sexes-irnt___EUR___chr7_...  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs = pd.read_csv('../work/coloc/panukbb/joint/postprocessed.txt', sep='\\t')\n",
    "colocs = colocs[colocs.gwas_ancestry=='EUR']\n",
    "#colocs = colocs[(colocs.xqtl_maf!='0.1%')]\n",
    "colocs = colocs[(colocs.xqtl_tissue=='Whole_blood') & (colocs.xqtl_maf=='1%') & (colocs.xqtl_modality=='ciseqtl')]\n",
    "colocs.xqtl_phenotype = colocs.xqtl_phenotype.str.split('.', expand=True)[0]\n",
    "colocs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8641\n",
      "8641\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-17-1e6a32bb7d85>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  for_merge['xqtl_phenotype'] = for_merge.phenotype_id\n",
      "<ipython-input-17-1e6a32bb7d85>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  for_merge['xqtl_cs'] = 'L' + for_merge.cs_id.astype(str)\n"
     ]
    }
   ],
   "source": [
    "print(len(colocs))\n",
    "for_merge = comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed']\n",
    "for_merge['xqtl_phenotype'] = for_merge.phenotype_id\n",
    "for_merge['xqtl_cs'] = 'L' + for_merge.cs_id.astype(str)\n",
    "for_merge = for_merge[['xqtl_cs', 'xqtl_phenotype', 'in_other_scan_cs']]\n",
    "colocs = colocs.merge(for_merge)\n",
    "print(len(colocs))\n",
    "#print(len(for_merge))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_tissue</th>\n",
       "      <th>xqtl_maf</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2162</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.663501e-14</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.366463e-15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000162777</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000162777.17</td>\n",
       "      <td>chr1_110947722_111447723</td>\n",
       "      <td>EUR</td>\n",
       "      <td>biomarkers-30750-both_sexes-irnt</td>\n",
       "      <td>biomarkers-30750-both_sexes-irnt___EUR___chr1_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2160</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.076595e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.130320e-11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000162777</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000162777.17</td>\n",
       "      <td>chr1_110943971_111443972</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30080-both_sexes-irnt</td>\n",
       "      <td>continuous-30080-both_sexes-irnt___EUR___chr1_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6415</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.198827e-177</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000140416</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000140416.21</td>\n",
       "      <td>chr15_62392636_63970660</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30100-both_sexes-irnt</td>\n",
       "      <td>continuous-30100-both_sexes-irnt___EUR___chr15...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5194</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.783741e-110</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000140416</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000140416.21</td>\n",
       "      <td>chr15_62799038_65581301</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30080-both_sexes-irnt</td>\n",
       "      <td>continuous-30080-both_sexes-irnt___EUR___chr15...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2026</td>\n",
       "      <td>chr1_159205564_G_A</td>\n",
       "      <td>chr1_159205564_G_A</td>\n",
       "      <td>1.501781e-170</td>\n",
       "      <td>9.969781e-52</td>\n",
       "      <td>3.012665e-122</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000213088.11</td>\n",
       "      <td>chr1_158955703_159455704</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30130-both_sexes-irnt</td>\n",
       "      <td>continuous-30130-both_sexes-irnt___EUR___chr1_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   nsnps                hit1                hit2      PP.H0.abf  \\\n",
       "0   2162  chr1_111197723_C_G  chr1_111197723_C_G   0.000000e+00   \n",
       "1   2160  chr1_111197723_C_G  chr1_111197723_C_G   0.000000e+00   \n",
       "2   6415  chr15_63049797_T_C  chr15_63049797_T_C   0.000000e+00   \n",
       "3   5194  chr15_63049797_T_C  chr15_63049797_T_C   0.000000e+00   \n",
       "4   2026  chr1_159205564_G_A  chr1_159205564_G_A  1.501781e-170   \n",
       "\n",
       "       PP.H1.abf      PP.H2.abf     PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "0   2.663501e-14   0.000000e+00  6.366463e-15        1.0     4     1  ...   \n",
       "1   2.076595e-11   0.000000e+00  1.130320e-11        1.0     4     2  ...   \n",
       "2  3.198827e-177   0.000000e+00  0.000000e+00        1.0     4     1  ...   \n",
       "3  5.783741e-110   0.000000e+00  0.000000e+00        1.0     4     2  ...   \n",
       "4   9.969781e-52  3.012665e-122  0.000000e+00        1.0     2     2  ...   \n",
       "\n",
       "   xqtl_tissue  xqtl_maf   xqtl_phenotype xqtl_ancestry           xqtl_gene  \\\n",
       "0  Whole_blood        1%  ENSG00000162777         joint  ENSG00000162777.17   \n",
       "1  Whole_blood        1%  ENSG00000162777         joint  ENSG00000162777.17   \n",
       "2  Whole_blood        1%  ENSG00000140416         joint  ENSG00000140416.21   \n",
       "3  Whole_blood        1%  ENSG00000140416         joint  ENSG00000140416.21   \n",
       "4  Whole_blood        1%  ENSG00000213088         joint  ENSG00000213088.11   \n",
       "\n",
       "                gwas_region  gwas_ancestry                        gwas_trait  \\\n",
       "0  chr1_110947722_111447723            EUR  biomarkers-30750-both_sexes-irnt   \n",
       "1  chr1_110943971_111443972            EUR  continuous-30080-both_sexes-irnt   \n",
       "2   chr15_62392636_63970660            EUR  continuous-30100-both_sexes-irnt   \n",
       "3   chr15_62799038_65581301            EUR  continuous-30080-both_sexes-irnt   \n",
       "4  chr1_158955703_159455704            EUR  continuous-30130-both_sexes-irnt   \n",
       "\n",
       "                                         gwas_signal in_other_scan_cs  \n",
       "0  biomarkers-30750-both_sexes-irnt___EUR___chr1_...                1  \n",
       "1  continuous-30080-both_sexes-irnt___EUR___chr1_...                1  \n",
       "2  continuous-30100-both_sexes-irnt___EUR___chr15...                1  \n",
       "3  continuous-30080-both_sexes-irnt___EUR___chr15...                1  \n",
       "4  continuous-30130-both_sexes-irnt___EUR___chr1_...                1  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5872"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(colocs[['gwas_cs', 'gwas_signal']].drop_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    2760\n",
       "1    2369\n",
       "Name: in_other_scan_cs, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs[['xqtl_phenotype', 'xqtl_cs', 'in_other_scan_cs']].drop_duplicates().in_other_scan_cs.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'For example, in whole blood we detect 69,766 total cis-eQTL signals  for 19,468 genes, 51,037 (73.2%) of which were not detected in GTEx, DIRECT, or eQTLGen. Of these 51,037, 2,760 (5.4%) colocalize with at least one PanUKBB GWAS signal, emphasizing their potential biological relevance and usefulness in downstream analysis'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"For example, in whole blood we detect {:,} total cis-eQTL signals  for 19,468 genes, {:,} ({}%) of which were not detected in GTEx, DIRECT, or eQTLGen. Of these {:,}, {:,} ({}%) colocalize with at least one PanUKBB GWAS signal, emphasizing their potential biological relevance and usefulness in downstream analysis\".format(\n",
    "len(topmed_whole_blood_ciseqtl[['phenotype_id', 'cs_id']].drop_duplicates()),\n",
    "comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts()[0],\n",
    "round(100*comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts()[0] / comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts().sum(), 1),\n",
    "comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts()[0],\n",
    "colocs[['xqtl_phenotype', 'xqtl_cs', 'in_other_scan_cs']].drop_duplicates().in_other_scan_cs.value_counts()[0],\n",
    "round(100*colocs[['xqtl_phenotype', 'xqtl_cs', 'in_other_scan_cs']].drop_duplicates().in_other_scan_cs.value_counts()[0] / comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts()[0], 1)\n",
    "\n",
    "\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51037"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comparisons_against_all_full[comparisons_against_all_full.src=='TOPMed'].in_other_scan_cs.value_counts()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_tissue</th>\n",
       "      <th>xqtl_maf</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2162</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.663501e-14</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.366463e-15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000162777</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000162777.17</td>\n",
       "      <td>chr1_110947722_111447723</td>\n",
       "      <td>EUR</td>\n",
       "      <td>biomarkers-30750-both_sexes-irnt</td>\n",
       "      <td>biomarkers-30750-both_sexes-irnt___EUR___chr1_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2160</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>chr1_111197723_C_G</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.076595e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.130320e-11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000162777</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000162777.17</td>\n",
       "      <td>chr1_110943971_111443972</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30080-both_sexes-irnt</td>\n",
       "      <td>continuous-30080-both_sexes-irnt___EUR___chr1_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6415</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.198827e-177</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000140416</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000140416.21</td>\n",
       "      <td>chr15_62392636_63970660</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30100-both_sexes-irnt</td>\n",
       "      <td>continuous-30100-both_sexes-irnt___EUR___chr15...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5194</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>chr15_63049797_T_C</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.783741e-110</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000140416</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000140416.21</td>\n",
       "      <td>chr15_62799038_65581301</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30080-both_sexes-irnt</td>\n",
       "      <td>continuous-30080-both_sexes-irnt___EUR___chr15...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2026</td>\n",
       "      <td>chr1_159205564_G_A</td>\n",
       "      <td>chr1_159205564_G_A</td>\n",
       "      <td>1.501781e-170</td>\n",
       "      <td>9.969781e-52</td>\n",
       "      <td>3.012665e-122</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000213088.11</td>\n",
       "      <td>chr1_158955703_159455704</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30130-both_sexes-irnt</td>\n",
       "      <td>continuous-30130-both_sexes-irnt___EUR___chr1_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   nsnps                hit1                hit2      PP.H0.abf  \\\n",
       "0   2162  chr1_111197723_C_G  chr1_111197723_C_G   0.000000e+00   \n",
       "1   2160  chr1_111197723_C_G  chr1_111197723_C_G   0.000000e+00   \n",
       "2   6415  chr15_63049797_T_C  chr15_63049797_T_C   0.000000e+00   \n",
       "3   5194  chr15_63049797_T_C  chr15_63049797_T_C   0.000000e+00   \n",
       "4   2026  chr1_159205564_G_A  chr1_159205564_G_A  1.501781e-170   \n",
       "\n",
       "       PP.H1.abf      PP.H2.abf     PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "0   2.663501e-14   0.000000e+00  6.366463e-15        1.0     4     1  ...   \n",
       "1   2.076595e-11   0.000000e+00  1.130320e-11        1.0     4     2  ...   \n",
       "2  3.198827e-177   0.000000e+00  0.000000e+00        1.0     4     1  ...   \n",
       "3  5.783741e-110   0.000000e+00  0.000000e+00        1.0     4     2  ...   \n",
       "4   9.969781e-52  3.012665e-122  0.000000e+00        1.0     2     2  ...   \n",
       "\n",
       "   xqtl_tissue  xqtl_maf   xqtl_phenotype xqtl_ancestry           xqtl_gene  \\\n",
       "0  Whole_blood        1%  ENSG00000162777         joint  ENSG00000162777.17   \n",
       "1  Whole_blood        1%  ENSG00000162777         joint  ENSG00000162777.17   \n",
       "2  Whole_blood        1%  ENSG00000140416         joint  ENSG00000140416.21   \n",
       "3  Whole_blood        1%  ENSG00000140416         joint  ENSG00000140416.21   \n",
       "4  Whole_blood        1%  ENSG00000213088         joint  ENSG00000213088.11   \n",
       "\n",
       "                gwas_region  gwas_ancestry                        gwas_trait  \\\n",
       "0  chr1_110947722_111447723            EUR  biomarkers-30750-both_sexes-irnt   \n",
       "1  chr1_110943971_111443972            EUR  continuous-30080-both_sexes-irnt   \n",
       "2   chr15_62392636_63970660            EUR  continuous-30100-both_sexes-irnt   \n",
       "3   chr15_62799038_65581301            EUR  continuous-30080-both_sexes-irnt   \n",
       "4  chr1_158955703_159455704            EUR  continuous-30130-both_sexes-irnt   \n",
       "\n",
       "                                         gwas_signal in_other_scan_cs  \n",
       "0  biomarkers-30750-both_sexes-irnt___EUR___chr1_...                1  \n",
       "1  continuous-30080-both_sexes-irnt___EUR___chr1_...                1  \n",
       "2  continuous-30100-both_sexes-irnt___EUR___chr15...                1  \n",
       "3  continuous-30080-both_sexes-irnt___EUR___chr15...                1  \n",
       "4  continuous-30130-both_sexes-irnt___EUR___chr1_...                1  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_tissue</th>\n",
       "      <th>xqtl_maf</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1831</th>\n",
       "      <td>2942</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>6.523500e-81</td>\n",
       "      <td>6.786296e-56</td>\n",
       "      <td>6.895906e-28</td>\n",
       "      <td>0.005184</td>\n",
       "      <td>0.994816</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76074762_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-6152-both_sexes-100</td>\n",
       "      <td>categorical-6152-both_sexes-100___EUR___chr11_...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1832</th>\n",
       "      <td>2934</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>3.905879e-125</td>\n",
       "      <td>4.063226e-100</td>\n",
       "      <td>7.508850e-28</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>0.994177</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76071628_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30210-both_sexes-irnt</td>\n",
       "      <td>continuous-30210-both_sexes-irnt___EUR___chr11...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1833</th>\n",
       "      <td>1496</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>1.961336e-30</td>\n",
       "      <td>2.040348e-05</td>\n",
       "      <td>1.044228e-27</td>\n",
       "      <td>0.008881</td>\n",
       "      <td>0.991099</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76332713_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-20003-both_sexes-1140881856</td>\n",
       "      <td>categorical-20003-both_sexes-1140881856___EUR_...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1834</th>\n",
       "      <td>1675</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582761_G_C</td>\n",
       "      <td>1.154040e-31</td>\n",
       "      <td>1.200530e-06</td>\n",
       "      <td>1.952618e-27</td>\n",
       "      <td>0.018349</td>\n",
       "      <td>0.981649</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76338386_76838387</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-20003-both_sexes-1140861998</td>\n",
       "      <td>categorical-20003-both_sexes-1140861998___EUR_...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1835</th>\n",
       "      <td>1852</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76584554_G_A</td>\n",
       "      <td>1.063491e-35</td>\n",
       "      <td>1.106334e-10</td>\n",
       "      <td>2.404550e-27</td>\n",
       "      <td>0.023060</td>\n",
       "      <td>0.976940</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76334553_76834554</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-135-both_sexes</td>\n",
       "      <td>continuous-135-both_sexes___EUR___chr11_763345...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>1842</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76570549_C_T</td>\n",
       "      <td>3.314461e-30</td>\n",
       "      <td>3.447983e-05</td>\n",
       "      <td>5.979166e-27</td>\n",
       "      <td>0.060321</td>\n",
       "      <td>0.939644</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>1%</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76334553_76834554</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-AG-both_sexes-irnt</td>\n",
       "      <td>continuous-AG-both_sexes-irnt___EUR___chr11_76...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      nsnps                hit1                hit2      PP.H0.abf  \\\n",
       "1831   2942  chr11_76582483_G_T  chr11_76582714_G_A   6.523500e-81   \n",
       "1832   2934  chr11_76582483_G_T  chr11_76582714_G_A  3.905879e-125   \n",
       "1833   1496  chr11_76582483_G_T  chr11_76582714_G_A   1.961336e-30   \n",
       "1834   1675  chr11_76582483_G_T  chr11_76582761_G_C   1.154040e-31   \n",
       "1835   1852  chr11_76582483_G_T  chr11_76584554_G_A   1.063491e-35   \n",
       "1836   1842  chr11_76582483_G_T  chr11_76570549_C_T   3.314461e-30   \n",
       "\n",
       "          PP.H1.abf     PP.H2.abf  PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "1831   6.786296e-56  6.895906e-28   0.005184   0.994816     2     1  ...   \n",
       "1832  4.063226e-100  7.508850e-28   0.005823   0.994177     2     1  ...   \n",
       "1833   2.040348e-05  1.044228e-27   0.008881   0.991099     2     1  ...   \n",
       "1834   1.200530e-06  1.952618e-27   0.018349   0.981649     2     1  ...   \n",
       "1835   1.106334e-10  2.404550e-27   0.023060   0.976940     2     1  ...   \n",
       "1836   3.447983e-05  5.979166e-27   0.060321   0.939644     2     1  ...   \n",
       "\n",
       "      xqtl_tissue  xqtl_maf   xqtl_phenotype xqtl_ancestry  \\\n",
       "1831  Whole_blood        1%  ENSG00000137507         joint   \n",
       "1832  Whole_blood        1%  ENSG00000137507         joint   \n",
       "1833  Whole_blood        1%  ENSG00000137507         joint   \n",
       "1834  Whole_blood        1%  ENSG00000137507         joint   \n",
       "1835  Whole_blood        1%  ENSG00000137507         joint   \n",
       "1836  Whole_blood        1%  ENSG00000137507         joint   \n",
       "\n",
       "               xqtl_gene              gwas_region  gwas_ancestry  \\\n",
       "1831  ENSG00000137507.11  chr11_76074762_76832714            EUR   \n",
       "1832  ENSG00000137507.11  chr11_76071628_76832714            EUR   \n",
       "1833  ENSG00000137507.11  chr11_76332713_76832714            EUR   \n",
       "1834  ENSG00000137507.11  chr11_76338386_76838387            EUR   \n",
       "1835  ENSG00000137507.11  chr11_76334553_76834554            EUR   \n",
       "1836  ENSG00000137507.11  chr11_76334553_76834554            EUR   \n",
       "\n",
       "                                   gwas_trait  \\\n",
       "1831          categorical-6152-both_sexes-100   \n",
       "1832         continuous-30210-both_sexes-irnt   \n",
       "1833  categorical-20003-both_sexes-1140881856   \n",
       "1834  categorical-20003-both_sexes-1140861998   \n",
       "1835                continuous-135-both_sexes   \n",
       "1836            continuous-AG-both_sexes-irnt   \n",
       "\n",
       "                                            gwas_signal in_other_scan_cs  \n",
       "1831  categorical-6152-both_sexes-100___EUR___chr11_...                0  \n",
       "1832  continuous-30210-both_sexes-irnt___EUR___chr11...                0  \n",
       "1833  categorical-20003-both_sexes-1140881856___EUR_...                0  \n",
       "1834  categorical-20003-both_sexes-1140861998___EUR_...                0  \n",
       "1835  continuous-135-both_sexes___EUR___chr11_763345...                0  \n",
       "1836  continuous-AG-both_sexes-irnt___EUR___chr11_76...                0  \n",
       "\n",
       "[6 rows x 31 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs[colocs.xqtl_gene=='ENSG00000137507.11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_type</th>\n",
       "      <th>phenocode</th>\n",
       "      <th>coding</th>\n",
       "      <th>modifier</th>\n",
       "      <th>description</th>\n",
       "      <th>description_more</th>\n",
       "      <th>coding_description</th>\n",
       "      <th>category</th>\n",
       "      <th>n_cases_EUR</th>\n",
       "      <th>n_controls_EUR</th>\n",
       "      <th>sldsc_25bin_h2_observed_EUR</th>\n",
       "      <th>sldsc_25bin_h2_observed_se_EUR</th>\n",
       "      <th>sldsc_25bin_h2_liability_EUR</th>\n",
       "      <th>sldsc_25bin_h2_liability_se_EUR</th>\n",
       "      <th>sldsc_25bin_h2_z_EUR</th>\n",
       "      <th>lambda_gc_EUR</th>\n",
       "      <th>aws_path</th>\n",
       "      <th>aws_path_tabix</th>\n",
       "      <th>gwas_trait</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>biomarkers</td>\n",
       "      <td>30600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>irnt</td>\n",
       "      <td>Albumin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biological samples &gt; Assay results &gt; Blood ass...</td>\n",
       "      <td>367192.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1448</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0.1448</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>16.6440</td>\n",
       "      <td>1.3680</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files/bio...</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files_tab...</td>\n",
       "      <td>biomarkers-30600-both_sexes-irnt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>biomarkers</td>\n",
       "      <td>30610</td>\n",
       "      <td>NaN</td>\n",
       "      <td>irnt</td>\n",
       "      <td>Alkaline phosphatase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biological samples &gt; Assay results &gt; Blood ass...</td>\n",
       "      <td>400988.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2047</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>0.2047</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>11.9710</td>\n",
       "      <td>1.6682</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files/bio...</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files_tab...</td>\n",
       "      <td>biomarkers-30610-both_sexes-irnt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>biomarkers</td>\n",
       "      <td>30620</td>\n",
       "      <td>NaN</td>\n",
       "      <td>irnt</td>\n",
       "      <td>Alanine aminotransferase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biological samples &gt; Assay results &gt; Blood ass...</td>\n",
       "      <td>400822.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1235</td>\n",
       "      <td>0.0091</td>\n",
       "      <td>0.1235</td>\n",
       "      <td>0.0091</td>\n",
       "      <td>13.5710</td>\n",
       "      <td>1.4150</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files/bio...</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files_tab...</td>\n",
       "      <td>biomarkers-30620-both_sexes-irnt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>biomarkers</td>\n",
       "      <td>30630</td>\n",
       "      <td>NaN</td>\n",
       "      <td>irnt</td>\n",
       "      <td>Apolipoprotein A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biological samples &gt; Assay results &gt; Blood ass...</td>\n",
       "      <td>364987.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1822</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>0.1822</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>15.1830</td>\n",
       "      <td>1.6285</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files/bio...</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files_tab...</td>\n",
       "      <td>biomarkers-30630-both_sexes-irnt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biomarkers</td>\n",
       "      <td>30640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>irnt</td>\n",
       "      <td>Apolipoprotein B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biological samples &gt; Assay results &gt; Blood ass...</td>\n",
       "      <td>399003.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1020</td>\n",
       "      <td>0.0114</td>\n",
       "      <td>0.1020</td>\n",
       "      <td>0.0114</td>\n",
       "      <td>8.9474</td>\n",
       "      <td>1.4551</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files/bio...</td>\n",
       "      <td>s3://pan-ukb-us-east-1/sumstats_flat_files_tab...</td>\n",
       "      <td>biomarkers-30640-both_sexes-irnt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   trait_type phenocode coding modifier               description  \\\n",
       "0  biomarkers     30600    NaN     irnt                   Albumin   \n",
       "1  biomarkers     30610    NaN     irnt      Alkaline phosphatase   \n",
       "2  biomarkers     30620    NaN     irnt  Alanine aminotransferase   \n",
       "3  biomarkers     30630    NaN     irnt          Apolipoprotein A   \n",
       "4  biomarkers     30640    NaN     irnt          Apolipoprotein B   \n",
       "\n",
       "  description_more coding_description  \\\n",
       "0              NaN                NaN   \n",
       "1              NaN                NaN   \n",
       "2              NaN                NaN   \n",
       "3              NaN                NaN   \n",
       "4              NaN                NaN   \n",
       "\n",
       "                                            category  n_cases_EUR  \\\n",
       "0  Biological samples > Assay results > Blood ass...     367192.0   \n",
       "1  Biological samples > Assay results > Blood ass...     400988.0   \n",
       "2  Biological samples > Assay results > Blood ass...     400822.0   \n",
       "3  Biological samples > Assay results > Blood ass...     364987.0   \n",
       "4  Biological samples > Assay results > Blood ass...     399003.0   \n",
       "\n",
       "   n_controls_EUR  sldsc_25bin_h2_observed_EUR  \\\n",
       "0             NaN                       0.1448   \n",
       "1             NaN                       0.2047   \n",
       "2             NaN                       0.1235   \n",
       "3             NaN                       0.1822   \n",
       "4             NaN                       0.1020   \n",
       "\n",
       "   sldsc_25bin_h2_observed_se_EUR  sldsc_25bin_h2_liability_EUR  \\\n",
       "0                          0.0087                        0.1448   \n",
       "1                          0.0171                        0.2047   \n",
       "2                          0.0091                        0.1235   \n",
       "3                          0.0120                        0.1822   \n",
       "4                          0.0114                        0.1020   \n",
       "\n",
       "   sldsc_25bin_h2_liability_se_EUR  sldsc_25bin_h2_z_EUR  lambda_gc_EUR  \\\n",
       "0                           0.0087               16.6440         1.3680   \n",
       "1                           0.0171               11.9710         1.6682   \n",
       "2                           0.0091               13.5710         1.4150   \n",
       "3                           0.0120               15.1830         1.6285   \n",
       "4                           0.0114                8.9474         1.4551   \n",
       "\n",
       "                                            aws_path  \\\n",
       "0  s3://pan-ukb-us-east-1/sumstats_flat_files/bio...   \n",
       "1  s3://pan-ukb-us-east-1/sumstats_flat_files/bio...   \n",
       "2  s3://pan-ukb-us-east-1/sumstats_flat_files/bio...   \n",
       "3  s3://pan-ukb-us-east-1/sumstats_flat_files/bio...   \n",
       "4  s3://pan-ukb-us-east-1/sumstats_flat_files/bio...   \n",
       "\n",
       "                                      aws_path_tabix  \\\n",
       "0  s3://pan-ukb-us-east-1/sumstats_flat_files_tab...   \n",
       "1  s3://pan-ukb-us-east-1/sumstats_flat_files_tab...   \n",
       "2  s3://pan-ukb-us-east-1/sumstats_flat_files_tab...   \n",
       "3  s3://pan-ukb-us-east-1/sumstats_flat_files_tab...   \n",
       "4  s3://pan-ukb-us-east-1/sumstats_flat_files_tab...   \n",
       "\n",
       "                         gwas_trait  \n",
       "0  biomarkers-30600-both_sexes-irnt  \n",
       "1  biomarkers-30610-both_sexes-irnt  \n",
       "2  biomarkers-30620-both_sexes-irnt  \n",
       "3  biomarkers-30630-both_sexes-irnt  \n",
       "4  biomarkers-30640-both_sexes-irnt  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_traits = pd.read_csv('/net/topmed11/working/porchard/panukbb-finemapping/work/selected-traits/manifest.EUR.txt', sep='\\t').rename(columns={'trait_id': 'gwas_trait'})\n",
    "gwas_traits.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8641\n",
      "8641\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "      <th>description</th>\n",
       "      <th>description_more</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2355</th>\n",
       "      <td>2934</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>3.905879e-125</td>\n",
       "      <td>4.063226e-100</td>\n",
       "      <td>7.508850e-28</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>0.994177</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76071628_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-30210-both_sexes-irnt</td>\n",
       "      <td>continuous-30210-both_sexes-irnt___EUR___chr11...</td>\n",
       "      <td>0</td>\n",
       "      <td>Eosinophill percentage</td>\n",
       "      <td>Result of \"Eosinophils Percentage\" assay, perf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2722</th>\n",
       "      <td>1842</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76570549_C_T</td>\n",
       "      <td>3.314461e-30</td>\n",
       "      <td>3.447983e-05</td>\n",
       "      <td>5.979166e-27</td>\n",
       "      <td>0.060321</td>\n",
       "      <td>0.939644</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76334553_76834554</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-AG-both_sexes-irnt</td>\n",
       "      <td>continuous-AG-both_sexes-irnt___EUR___chr11_76...</td>\n",
       "      <td>0</td>\n",
       "      <td>Albumin/Globulin ratio</td>\n",
       "      <td>Albumin (30600) / (Total Protein [30860] - Alb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7190</th>\n",
       "      <td>2942</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>6.523500e-81</td>\n",
       "      <td>6.786296e-56</td>\n",
       "      <td>6.895906e-28</td>\n",
       "      <td>0.005184</td>\n",
       "      <td>0.994816</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76074762_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-6152-both_sexes-100</td>\n",
       "      <td>categorical-6152-both_sexes-100___EUR___chr11_...</td>\n",
       "      <td>0</td>\n",
       "      <td>Blood clot, DVT, bronchitis, emphysema, asthma...</td>\n",
       "      <td>ACE touchscreen question \"Has a doctor ever to...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7539</th>\n",
       "      <td>1675</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582761_G_C</td>\n",
       "      <td>1.154040e-31</td>\n",
       "      <td>1.200530e-06</td>\n",
       "      <td>1.952618e-27</td>\n",
       "      <td>0.018349</td>\n",
       "      <td>0.981649</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76338386_76838387</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-20003-both_sexes-1140861998</td>\n",
       "      <td>categorical-20003-both_sexes-1140861998___EUR_...</td>\n",
       "      <td>0</td>\n",
       "      <td>Treatment/medication code</td>\n",
       "      <td>Code for treatment Negative codes indicate fre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7548</th>\n",
       "      <td>1496</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>1.961336e-30</td>\n",
       "      <td>2.040348e-05</td>\n",
       "      <td>1.044228e-27</td>\n",
       "      <td>0.008881</td>\n",
       "      <td>0.991099</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76332713_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-20003-both_sexes-1140881856</td>\n",
       "      <td>categorical-20003-both_sexes-1140881856___EUR_...</td>\n",
       "      <td>0</td>\n",
       "      <td>Treatment/medication code</td>\n",
       "      <td>Code for treatment Negative codes indicate fre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8132</th>\n",
       "      <td>1852</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76584554_G_A</td>\n",
       "      <td>1.063491e-35</td>\n",
       "      <td>1.106334e-10</td>\n",
       "      <td>2.404550e-27</td>\n",
       "      <td>0.023060</td>\n",
       "      <td>0.976940</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76334553_76834554</td>\n",
       "      <td>EUR</td>\n",
       "      <td>continuous-135-both_sexes</td>\n",
       "      <td>continuous-135-both_sexes___EUR___chr11_763345...</td>\n",
       "      <td>0</td>\n",
       "      <td>Number of self-reported non-cancer illnesses</td>\n",
       "      <td>Number of non-cancer illnesses entered</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      nsnps                hit1                hit2      PP.H0.abf  \\\n",
       "2355   2934  chr11_76582483_G_T  chr11_76582714_G_A  3.905879e-125   \n",
       "2722   1842  chr11_76582483_G_T  chr11_76570549_C_T   3.314461e-30   \n",
       "7190   2942  chr11_76582483_G_T  chr11_76582714_G_A   6.523500e-81   \n",
       "7539   1675  chr11_76582483_G_T  chr11_76582761_G_C   1.154040e-31   \n",
       "7548   1496  chr11_76582483_G_T  chr11_76582714_G_A   1.961336e-30   \n",
       "8132   1852  chr11_76582483_G_T  chr11_76584554_G_A   1.063491e-35   \n",
       "\n",
       "          PP.H1.abf     PP.H2.abf  PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "2355  4.063226e-100  7.508850e-28   0.005823   0.994177     2     1  ...   \n",
       "2722   3.447983e-05  5.979166e-27   0.060321   0.939644     2     1  ...   \n",
       "7190   6.786296e-56  6.895906e-28   0.005184   0.994816     2     1  ...   \n",
       "7539   1.200530e-06  1.952618e-27   0.018349   0.981649     2     1  ...   \n",
       "7548   2.040348e-05  1.044228e-27   0.008881   0.991099     2     1  ...   \n",
       "8132   1.106334e-10  2.404550e-27   0.023060   0.976940     2     1  ...   \n",
       "\n",
       "       xqtl_phenotype  xqtl_ancestry           xqtl_gene  \\\n",
       "2355  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "2722  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "7190  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "7539  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "7548  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "8132  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "\n",
       "                  gwas_region gwas_ancestry  \\\n",
       "2355  chr11_76071628_76832714           EUR   \n",
       "2722  chr11_76334553_76834554           EUR   \n",
       "7190  chr11_76074762_76832714           EUR   \n",
       "7539  chr11_76338386_76838387           EUR   \n",
       "7548  chr11_76332713_76832714           EUR   \n",
       "8132  chr11_76334553_76834554           EUR   \n",
       "\n",
       "                                   gwas_trait  \\\n",
       "2355         continuous-30210-both_sexes-irnt   \n",
       "2722            continuous-AG-both_sexes-irnt   \n",
       "7190          categorical-6152-both_sexes-100   \n",
       "7539  categorical-20003-both_sexes-1140861998   \n",
       "7548  categorical-20003-both_sexes-1140881856   \n",
       "8132                continuous-135-both_sexes   \n",
       "\n",
       "                                            gwas_signal  in_other_scan_cs  \\\n",
       "2355  continuous-30210-both_sexes-irnt___EUR___chr11...                 0   \n",
       "2722  continuous-AG-both_sexes-irnt___EUR___chr11_76...                 0   \n",
       "7190  categorical-6152-both_sexes-100___EUR___chr11_...                 0   \n",
       "7539  categorical-20003-both_sexes-1140861998___EUR_...                 0   \n",
       "7548  categorical-20003-both_sexes-1140881856___EUR_...                 0   \n",
       "8132  continuous-135-both_sexes___EUR___chr11_763345...                 0   \n",
       "\n",
       "                                            description  \\\n",
       "2355                             Eosinophill percentage   \n",
       "2722                             Albumin/Globulin ratio   \n",
       "7190  Blood clot, DVT, bronchitis, emphysema, asthma...   \n",
       "7539                          Treatment/medication code   \n",
       "7548                          Treatment/medication code   \n",
       "8132       Number of self-reported non-cancer illnesses   \n",
       "\n",
       "                                       description_more  \n",
       "2355  Result of \"Eosinophils Percentage\" assay, perf...  \n",
       "2722  Albumin (30600) / (Total Protein [30860] - Alb...  \n",
       "7190  ACE touchscreen question \"Has a doctor ever to...  \n",
       "7539  Code for treatment Negative codes indicate fre...  \n",
       "7548  Code for treatment Negative codes indicate fre...  \n",
       "8132             Number of non-cancer illnesses entered  \n",
       "\n",
       "[6 rows x 33 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(colocs))\n",
    "colocs = colocs.merge(gwas_traits[['description', 'description_more', 'gwas_trait']])\n",
    "print(len(colocs))\n",
    "colocs[colocs.xqtl_gene=='ENSG00000137507.11']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['continuous-30210-both_sexes-irnt___EUR___chr11_76071628_76832714',\n",
       "       'continuous-AG-both_sexes-irnt___EUR___chr11_76334553_76834554',\n",
       "       'categorical-6152-both_sexes-100___EUR___chr11_76074762_76832714',\n",
       "       'categorical-20003-both_sexes-1140861998___EUR___chr11_76338386_76838387',\n",
       "       'categorical-20003-both_sexes-1140881856___EUR___chr11_76332713_76832714',\n",
       "       'continuous-135-both_sexes___EUR___chr11_76334553_76834554'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs[colocs.xqtl_gene=='ENSG00000137507.11'].gwas_signal.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "      <th>description</th>\n",
       "      <th>description_more</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7190</th>\n",
       "      <td>2942</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>6.523500e-81</td>\n",
       "      <td>6.786296e-56</td>\n",
       "      <td>6.895906e-28</td>\n",
       "      <td>0.005184</td>\n",
       "      <td>0.994816</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76074762_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-6152-both_sexes-100</td>\n",
       "      <td>categorical-6152-both_sexes-100___EUR___chr11_...</td>\n",
       "      <td>0</td>\n",
       "      <td>Blood clot, DVT, bronchitis, emphysema, asthma...</td>\n",
       "      <td>ACE touchscreen question \"Has a doctor ever to...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      nsnps                hit1                hit2     PP.H0.abf  \\\n",
       "7190   2942  chr11_76582483_G_T  chr11_76582714_G_A  6.523500e-81   \n",
       "\n",
       "         PP.H1.abf     PP.H2.abf  PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "7190  6.786296e-56  6.895906e-28   0.005184   0.994816     2     1  ...   \n",
       "\n",
       "       xqtl_phenotype  xqtl_ancestry           xqtl_gene  \\\n",
       "7190  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "\n",
       "                  gwas_region gwas_ancestry                       gwas_trait  \\\n",
       "7190  chr11_76074762_76832714           EUR  categorical-6152-both_sexes-100   \n",
       "\n",
       "                                            gwas_signal  in_other_scan_cs  \\\n",
       "7190  categorical-6152-both_sexes-100___EUR___chr11_...                 0   \n",
       "\n",
       "                                            description  \\\n",
       "7190  Blood clot, DVT, bronchitis, emphysema, asthma...   \n",
       "\n",
       "                                       description_more  \n",
       "7190  ACE touchscreen question \"Has a doctor ever to...  \n",
       "\n",
       "[1 rows x 33 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs[(colocs.xqtl_gene=='ENSG00000137507.11') & (colocs.gwas_signal=='categorical-6152-both_sexes-100___EUR___chr11_76074762_76832714')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nsnps</th>\n",
       "      <th>hit1</th>\n",
       "      <th>hit2</th>\n",
       "      <th>PP.H0.abf</th>\n",
       "      <th>PP.H1.abf</th>\n",
       "      <th>PP.H2.abf</th>\n",
       "      <th>PP.H3.abf</th>\n",
       "      <th>PP.H4.abf</th>\n",
       "      <th>idx1</th>\n",
       "      <th>idx2</th>\n",
       "      <th>...</th>\n",
       "      <th>xqtl_phenotype</th>\n",
       "      <th>xqtl_ancestry</th>\n",
       "      <th>xqtl_gene</th>\n",
       "      <th>gwas_region</th>\n",
       "      <th>gwas_ancestry</th>\n",
       "      <th>gwas_trait</th>\n",
       "      <th>gwas_signal</th>\n",
       "      <th>in_other_scan_cs</th>\n",
       "      <th>description</th>\n",
       "      <th>description_more</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7190</th>\n",
       "      <td>2942</td>\n",
       "      <td>chr11_76582483_G_T</td>\n",
       "      <td>chr11_76582714_G_A</td>\n",
       "      <td>6.523500e-81</td>\n",
       "      <td>6.786296e-56</td>\n",
       "      <td>6.895906e-28</td>\n",
       "      <td>0.005184</td>\n",
       "      <td>0.994816</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000137507</td>\n",
       "      <td>joint</td>\n",
       "      <td>ENSG00000137507.11</td>\n",
       "      <td>chr11_76074762_76832714</td>\n",
       "      <td>EUR</td>\n",
       "      <td>categorical-6152-both_sexes-100</td>\n",
       "      <td>categorical-6152-both_sexes-100___EUR___chr11_...</td>\n",
       "      <td>0</td>\n",
       "      <td>Blood clot, DVT, bronchitis, emphysema, asthma...</td>\n",
       "      <td>ACE touchscreen question \"Has a doctor ever to...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      nsnps                hit1                hit2     PP.H0.abf  \\\n",
       "7190   2942  chr11_76582483_G_T  chr11_76582714_G_A  6.523500e-81   \n",
       "\n",
       "         PP.H1.abf     PP.H2.abf  PP.H3.abf  PP.H4.abf  idx1  idx2  ...  \\\n",
       "7190  6.786296e-56  6.895906e-28   0.005184   0.994816     2     1  ...   \n",
       "\n",
       "       xqtl_phenotype  xqtl_ancestry           xqtl_gene  \\\n",
       "7190  ENSG00000137507          joint  ENSG00000137507.11   \n",
       "\n",
       "                  gwas_region gwas_ancestry                       gwas_trait  \\\n",
       "7190  chr11_76074762_76832714           EUR  categorical-6152-both_sexes-100   \n",
       "\n",
       "                                            gwas_signal  in_other_scan_cs  \\\n",
       "7190  categorical-6152-both_sexes-100___EUR___chr11_...                 0   \n",
       "\n",
       "                                            description  \\\n",
       "7190  Blood clot, DVT, bronchitis, emphysema, asthma...   \n",
       "\n",
       "                                       description_more  \n",
       "7190  ACE touchscreen question \"Has a doctor ever to...  \n",
       "\n",
       "[1 rows x 33 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs[(colocs.gwas_signal=='categorical-6152-both_sexes-100___EUR___chr11_76074762_76832714')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "nsnps                                                            2942\n",
       "hit1                                               chr11_76582483_G_T\n",
       "hit2                                               chr11_76582714_G_A\n",
       "PP.H0.abf                                                         0.0\n",
       "PP.H1.abf                                                         0.0\n",
       "PP.H2.abf                                                         0.0\n",
       "PP.H3.abf                                                    0.005184\n",
       "PP.H4.abf                                                    0.994816\n",
       "idx1                                                                2\n",
       "idx2                                                                1\n",
       "chrom                                                           chr11\n",
       "region_start                                                 76075261\n",
       "region_end                                                   76832430\n",
       "xqtl_cs                                                            L1\n",
       "gwas_cs                                                            L1\n",
       "p1                                                             0.0001\n",
       "p2                                                             0.0001\n",
       "p12                                                          0.000005\n",
       "xqtl_file           /tmp/rnaseq-2024-10-16-sample-update/stage/col...\n",
       "gwas_file           /tmp/rnaseq-2024-10-16-sample-update/stage/col...\n",
       "xqtl_modality                                                 ciseqtl\n",
       "xqtl_tissue                                               Whole_blood\n",
       "xqtl_maf                                                           1%\n",
       "xqtl_phenotype                                        ENSG00000137507\n",
       "xqtl_ancestry                                                   joint\n",
       "xqtl_gene                                          ENSG00000137507.11\n",
       "gwas_region                                   chr11_76074762_76832714\n",
       "gwas_ancestry                                                     EUR\n",
       "gwas_trait                            categorical-6152-both_sexes-100\n",
       "gwas_signal         categorical-6152-both_sexes-100___EUR___chr11_...\n",
       "in_other_scan_cs                                                    0\n",
       "description         Blood clot, DVT, bronchitis, emphysema, asthma...\n",
       "description_more    ACE touchscreen question \"Has a doctor ever to...\n",
       "Name: 7190, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colocs.loc[7190]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colocs[coloced.gwas_signal=='']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2020-09",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
