{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:15:54.769412Z",
     "iopub.status.busy": "2024-12-03T19:15:54.758685Z",
     "iopub.status.idle": "2024-12-03T19:15:56.032367Z",
     "shell.execute_reply": "2024-12-03T19:15:56.032677Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/net/snowwhite/home/porchard/anaconda3/envs/2020-09/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n",
      "/net/snowwhite/home/porchard/anaconda3/envs/2020-09/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import logging\n",
    "import glob\n",
    "import os\n",
    "from topmed_manuscript_clean import phenotype_id_to_gene_id\n",
    "from scipy.stats import fisher_exact as fe\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s: %(message)s')\n",
    "import random\n",
    "import re\n",
    "\n",
    "random.seed(42)\n",
    "\n",
    "def shuffle_dict(d):\n",
    "    \"\"\"\n",
    "    Shuffles the key - value relationships of a dictionary\n",
    "    \n",
    "    x = {'a': 0, 'b': 1, 'c': 2}\n",
    "    print(x)\n",
    "    print(shuffle_dict(x))\n",
    "    \n",
    "    \"\"\"\n",
    "    keys = list(d.keys())\n",
    "    values = [d[k] for k in keys]\n",
    "    random.shuffle(keys)\n",
    "    return dict(zip(keys, values))\n",
    "\n",
    "\n",
    "def parse_ensembl_id(x):\n",
    "    \"\"\"\n",
    "    See https://m.ensembl.org/info/genome/stable_ids/index.html:\n",
    "    ENS[species prefix][feature type prefix][a unique eleven digit number]\n",
    "    \"\"\"\n",
    "    FEATURE_TYPE_PREFICES = ['E', 'FM', 'G', 'GT', 'P', 'R', 'T']\n",
    "    ENSEMBL_RE = '^ENS(.*?)({})(\\d+)\\.?(.*?)$'.format('|'.join(FEATURE_TYPE_PREFICES))\n",
    "    species_prefix, feature_type_prefix, digits, version = re.match(ENSEMBL_RE, x).groups()\n",
    "    return {'species_prefix': species_prefix, 'feature_type_prefix': feature_type_prefix, 'digits': digits, 'version': version}\n",
    "\n",
    "    \n",
    "def strip_version_from_ensembl_gene_id(x, keep_par_y=False):\n",
    "    \"\"\"\n",
    "    ENSG00000280767.3 --> ENSG00000280767\n",
    "    if keep_par_y: ENSG00000280767.3_PAR_Y --> ENSG00000280767.PAR_Y\n",
    "    if !keep_par_y: ENSG00000280767.3_PAR_Y --> ENSG00000280767\n",
    "    \"\"\"\n",
    "    x = parse_ensembl_id(x)\n",
    "    if keep_par_y and 'PAR_Y' in x['version']:\n",
    "        return 'ENS{species_prefix}{feature_type_prefix}{digits}.PAR_Y'.format(**x)\n",
    "    else:\n",
    "        return 'ENS{species_prefix}{feature_type_prefix}{digits}'.format(**x)\n",
    "\n",
    "\n",
    "def strip_version(x):\n",
    "    return strip_version_from_ensembl_gene_id(x, keep_par_y=True)\n",
    "\n",
    "\n",
    "def fishers_exact(variants, control_variants, annotation_table):\n",
    "    \"\"\"\n",
    "    variants: list/array of variant IDs\n",
    "    control_variants: list/array of control variant IDs\n",
    "    annotation_table: binary matrix indicating if each variant (row) overlaps each annotation (column). Index must contain all variants and control variants\n",
    "\n",
    "    returns: fishers exact test results comparing variant to control variant overlap of each annotation\n",
    "    \"\"\"\n",
    "    n_hits_overlapping_annotation = annotation_table.loc[variants].sum()\n",
    "    n_hits_not_overlapping_annotation = len(variants) - n_hits_overlapping_annotation\n",
    "    n_controls_overlapping_annotation = annotation_table.loc[control_variants].sum()\n",
    "    n_controls_not_overlapping_annotation = len(control_variants) - n_controls_overlapping_annotation\n",
    "    \n",
    "    fet_tables = {annotation: \n",
    "                  np.array([\n",
    "                      [n_hits_overlapping_annotation[annotation], n_hits_not_overlapping_annotation[annotation]], \n",
    "                      [n_controls_overlapping_annotation[annotation], n_controls_not_overlapping_annotation[annotation]]\n",
    "                      ]) for annotation in annotation_table.columns}\n",
    "    fet = {annotation: fe(v) for annotation, v in fet_tables.items()}\n",
    "    \n",
    "    results = pd.DataFrame([[annotation, v[0], v[1]] for annotation, v in fet.items()], columns=['annotation', 'odds_ratio', 'p'])\n",
    "    results['n_hits_overlapping_annotation'] = n_hits_overlapping_annotation.loc[results.annotation.to_list()].to_list()\n",
    "    results['n_hits_not_overlapping_annotation'] = n_hits_not_overlapping_annotation.loc[results.annotation.to_list()].to_list()\n",
    "    results['n_controls_overlapping_annotation'] = n_controls_overlapping_annotation.loc[results.annotation.to_list()].to_list()\n",
    "    results['n_controls_not_overlapping_annotation'] = n_controls_not_overlapping_annotation.loc[results.annotation.to_list()].to_list()\n",
    "\n",
    "    return results\n",
    "\n",
    "def parse_attribute(attribute_series: pd.Series, attribute_name: str) -> pd.Series:\n",
    "    \"\"\"\n",
    "    Parse the attributes column of a (GENCODE/RefSeq) GTF file.\n",
    "\n",
    "    Input:\n",
    "    * a [str]: the attributes element (column 9 of the GTF file)\n",
    "    * regex [str]: a regular expression that will be iteratively applied to the attribute string to capture attribute key, val pairs. Default should work for GENCODE/RefSeq\n",
    "    \"\"\"\n",
    "    if not isinstance(attribute_series, pd.Series):\n",
    "        raise TypeError('attribute_series must be a pandas Series')\n",
    "    if not isinstance(attribute_name, str):\n",
    "        raise TypeError('attribute_name must be a string')\n",
    "    \n",
    "    return attribute_series.str.extract(f'{attribute_name} \"(.*?)\"')\n",
    "\n",
    "\n",
    "def gtf_to_df(gtf: str, parse_attributes: list=None) -> pd.DataFrame:\n",
    "    df = pd.read_csv(gtf, sep='\\t', header=None, names=['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes'], comment='#')\n",
    "    if parse_attributes is not None:\n",
    "        for a in parse_attributes:\n",
    "            df[a] = parse_attribute(df.attributes, a)\n",
    "    return df\n",
    "\n",
    "PREFIX = 'tables/trans-enrichment-in-cis.'\n",
    "CLUMPED_TRANS_EQTL = '../work/clump-trans-variants/clump-trans-signals.significant-trans-eqtl-clumped.tsv'\n",
    "CLUMPED_TRANS_SQTL = '../work/clump-trans-variants/clump-trans-signals.significant-trans-sqtl-clumped.tsv'\n",
    "CONTROL_SNPS_TRANS_EQTL = glob.glob('../work/control-snps/trans-eqtl/maf005/results/controls/*.controls.txt')\n",
    "CONTROL_SNPS_TRANS_SQTL = glob.glob('../work/control-snps/trans-sqtl/maf005/results/controls/*.controls.txt')\n",
    "GTF = '../data/gtf/gencode.v30.GRCh38.ERCC.genes.collapsed_only.gtf.gz'\n",
    "USE_TISSUES = ['Whole_blood']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:15:56.038906Z",
     "iopub.status.busy": "2024-12-03T19:15:56.038541Z",
     "iopub.status.idle": "2024-12-03T19:15:59.817868Z",
     "shell.execute_reply": "2024-12-03T19:15:59.818749Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278267.1\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243485.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284332.1\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   source feature  start    end score strand frame  \\\n",
       "0   chr1   HAVANA    gene  11869  14409     .      +     .   \n",
       "6   chr1   HAVANA    gene  14404  29570     .      -     .   \n",
       "19  chr1  ENSEMBL    gene  17369  17436     .      -     .   \n",
       "22  chr1   HAVANA    gene  29554  31109     .      +     .   \n",
       "27  chr1  ENSEMBL    gene  30366  30503     .      +     .   \n",
       "\n",
       "                                           attributes            gene_id  \\\n",
       "0   gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  ENSG00000223972.5   \n",
       "6   gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  ENSG00000227232.5   \n",
       "19  gene_id \"ENSG00000278267.1\"; transcript_id \"EN...  ENSG00000278267.1   \n",
       "22  gene_id \"ENSG00000243485.5\"; transcript_id \"EN...  ENSG00000243485.5   \n",
       "27  gene_id \"ENSG00000284332.1\"; transcript_id \"EN...  ENSG00000284332.1   \n",
       "\n",
       "      gene_name  \n",
       "0       DDX11L1  \n",
       "6        WASH7P  \n",
       "19    MIR6859-1  \n",
       "22  MIR1302-2HG  \n",
       "27    MIR1302-2  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df = gtf_to_df(GTF, parse_attributes=['gene_id', 'gene_name'])\n",
    "gtf_df = gtf_df[gtf_df.feature=='gene']\n",
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:15:59.825300Z",
     "iopub.status.busy": "2024-12-03T19:15:59.824395Z",
     "iopub.status.idle": "2024-12-03T19:15:59.840816Z",
     "shell.execute_reply": "2024-12-03T19:15:59.840376Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7_50342615_A_G</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.235203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_156344836_A_G</td>\n",
       "      <td>ENSG00000002330.13</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.500775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6_144036619_C_A</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.055934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6_122440739_T_C</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.706539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22_46290431_C_G</td>\n",
       "      <td>ENSG00000004799.8</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.107530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           variant_id        phenotype_id       tissue        af\n",
       "0   chr7_50342615_A_G  ENSG00000000938.13  Whole_blood  0.235203\n",
       "1  chr1_156344836_A_G  ENSG00000002330.13  Whole_blood  0.500775\n",
       "2  chr6_144036619_C_A  ENSG00000004059.11  Whole_blood  0.055934\n",
       "3  chr6_122440739_T_C   ENSG00000004478.8  Whole_blood  0.706539\n",
       "4  chr22_46290431_C_G   ENSG00000004799.8  Whole_blood  0.107530"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_eqtl = pd.read_csv(CLUMPED_TRANS_EQTL, sep='\\t')\n",
    "trans_eqtl = trans_eqtl.loc[(trans_eqtl.qvalue<=0.05),['clumped_variant_id', 'phenotype_id', 'tissue', 'af']].rename(columns={'clumped_variant_id': 'variant_id'})\n",
    "trans_eqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:15:59.845072Z",
     "iopub.status.busy": "2024-12-03T19:15:59.844427Z",
     "iopub.status.idle": "2024-12-03T19:15:59.854521Z",
     "shell.execute_reply": "2024-12-03T19:15:59.854117Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_156344313_T_C</td>\n",
       "      <td>chr19:35907762:35908175:clu_21928_-:ENSG000000...</td>\n",
       "      <td>ENSG00000011600.11</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.497753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7_50360284_G_A</td>\n",
       "      <td>chrX:48467925:48468305:clu_44487_-:ENSG0000001...</td>\n",
       "      <td>ENSG00000017483.15</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.498528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3_187117653_G_A</td>\n",
       "      <td>chr4:25256674:25259037:clu_33710_+:ENSG0000003...</td>\n",
       "      <td>ENSG00000038210.13</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.633561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6_163408503_T_C</td>\n",
       "      <td>chr11:63903158:63904786:clu_8082_+:ENSG0000007...</td>\n",
       "      <td>ENSG00000072518.20</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.933065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6_163408503_T_C</td>\n",
       "      <td>chr11:85983973:85990250:clu_7229_-:ENSG0000007...</td>\n",
       "      <td>ENSG00000073921.18</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>0.932987</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           variant_id                                       phenotype_id  \\\n",
       "0  chr1_156344313_T_C  chr19:35907762:35908175:clu_21928_-:ENSG000000...   \n",
       "1   chr7_50360284_G_A  chrX:48467925:48468305:clu_44487_-:ENSG0000001...   \n",
       "2  chr3_187117653_G_A  chr4:25256674:25259037:clu_33710_+:ENSG0000003...   \n",
       "3  chr6_163408503_T_C  chr11:63903158:63904786:clu_8082_+:ENSG0000007...   \n",
       "4  chr6_163408503_T_C  chr11:85983973:85990250:clu_7229_-:ENSG0000007...   \n",
       "\n",
       "              gene_id       tissue        af  \n",
       "0  ENSG00000011600.11  Whole_blood  0.497753  \n",
       "1  ENSG00000017483.15  Whole_blood  0.498528  \n",
       "2  ENSG00000038210.13  Whole_blood  0.633561  \n",
       "3  ENSG00000072518.20  Whole_blood  0.933065  \n",
       "4  ENSG00000073921.18  Whole_blood  0.932987  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_sqtl = pd.read_csv(CLUMPED_TRANS_SQTL, sep='\\t')\n",
    "trans_sqtl = trans_sqtl.loc[(trans_sqtl.qvalue<=0.05),['clumped_variant_id', 'phenotype_id', 'gene_id', 'tissue', 'af']].rename(columns={'clumped_variant_id': 'variant_id'})\n",
    "trans_sqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:15:59.858711Z",
     "iopub.status.busy": "2024-12-03T19:15:59.858045Z",
     "iopub.status.idle": "2024-12-03T19:16:02.936981Z",
     "shell.execute_reply": "2024-12-03T19:16:02.935982Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>af</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr20:50941209:50942031:clu_35052_-:ENSG000000...</td>\n",
       "      <td>chr20_50879604_A_C</td>\n",
       "      <td>0.054397</td>\n",
       "      <td>0.037568</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20:50941209:50942031:clu_35052_-:ENSG000000...</td>\n",
       "      <td>chr20_50882479_T_C</td>\n",
       "      <td>0.055092</td>\n",
       "      <td>0.025174</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20:50941209:50942031:clu_35052_-:ENSG000000...</td>\n",
       "      <td>chr20_50885892_G_C</td>\n",
       "      <td>0.237507</td>\n",
       "      <td>0.019752</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20:50941209:50942031:clu_35052_-:ENSG000000...</td>\n",
       "      <td>chr20_50888836_C_A</td>\n",
       "      <td>0.030341</td>\n",
       "      <td>0.024400</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20:50941209:50942031:clu_35052_-:ENSG000000...</td>\n",
       "      <td>chr20_50892638_C_T</td>\n",
       "      <td>0.079246</td>\n",
       "      <td>0.020139</td>\n",
       "      <td>1</td>\n",
       "      <td>Lung</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        phenotype_id          variant_id  \\\n",
       "0  chr20:50941209:50942031:clu_35052_-:ENSG000000...  chr20_50879604_A_C   \n",
       "1  chr20:50941209:50942031:clu_35052_-:ENSG000000...  chr20_50882479_T_C   \n",
       "2  chr20:50941209:50942031:clu_35052_-:ENSG000000...  chr20_50885892_G_C   \n",
       "3  chr20:50941209:50942031:clu_35052_-:ENSG000000...  chr20_50888836_C_A   \n",
       "4  chr20:50941209:50942031:clu_35052_-:ENSG000000...  chr20_50892638_C_T   \n",
       "\n",
       "        pip        af  cs_id tissue  \n",
       "0  0.054397  0.037568      1   Lung  \n",
       "1  0.055092  0.025174      1   Lung  \n",
       "2  0.237507  0.019752      1   Lung  \n",
       "3  0.030341  0.024400      1   Lung  \n",
       "4  0.079246  0.020139      1   Lung  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cis_eqtl = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).split('.')[0]) for f in glob.glob('../data/scan-results/joint/cis-eqtl/susie/maf001/*.cs.txt')])\n",
    "cis_sqtl = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).split('.')[0]) for f in glob.glob('../data/scan-results/joint/cis-sqtl/susie/maf001/postprocessed/*.by-gene.cs.txt')])\n",
    "cis_sqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:02.945509Z",
     "iopub.status.busy": "2024-12-03T19:16:02.944366Z",
     "iopub.status.idle": "2024-12-03T19:16:02.976484Z",
     "shell.execute_reply": "2024-12-03T19:16:02.977289Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>control_variant</th>\n",
       "      <th>variant_maf</th>\n",
       "      <th>control_variant_maf</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18_79466098_G_A</td>\n",
       "      <td>chr18_60801235_T_C</td>\n",
       "      <td>0.05562</td>\n",
       "      <td>0.05803</td>\n",
       "      <td>Whole_blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18_2646148_A_G</td>\n",
       "      <td>chr18_8927091_A_G</td>\n",
       "      <td>0.07251</td>\n",
       "      <td>0.07282</td>\n",
       "      <td>Whole_blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr18_76821755_T_C</td>\n",
       "      <td>chr18_74410805_T_G</td>\n",
       "      <td>0.18720</td>\n",
       "      <td>0.19210</td>\n",
       "      <td>Whole_blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr18_72760883_T_C</td>\n",
       "      <td>chr18_77018215_A_G</td>\n",
       "      <td>0.30860</td>\n",
       "      <td>0.30450</td>\n",
       "      <td>Whole_blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr18_74608983_G_A</td>\n",
       "      <td>chr18_8792525_C_CT</td>\n",
       "      <td>0.31100</td>\n",
       "      <td>0.31600</td>\n",
       "      <td>Whole_blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               variant     control_variant  variant_maf  control_variant_maf  \\\n",
       "0   chr18_79466098_G_A  chr18_60801235_T_C      0.05562              0.05803   \n",
       "5    chr18_2646148_A_G   chr18_8927091_A_G      0.07251              0.07282   \n",
       "10  chr18_76821755_T_C  chr18_74410805_T_G      0.18720              0.19210   \n",
       "15  chr18_72760883_T_C  chr18_77018215_A_G      0.30860              0.30450   \n",
       "20  chr18_74608983_G_A  chr18_8792525_C_CT      0.31100              0.31600   \n",
       "\n",
       "         tissue  \n",
       "0   Whole_blood  \n",
       "5   Whole_blood  \n",
       "10  Whole_blood  \n",
       "15  Whole_blood  \n",
       "20  Whole_blood  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "controls_trans_eqtl = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).replace('.controls.txt', '')) for f in CONTROL_SNPS_TRANS_EQTL])\n",
    "controls_trans_eqtl = controls_trans_eqtl[controls_trans_eqtl.tissue.isin(USE_TISSUES)]\n",
    "controls_trans_eqtl = controls_trans_eqtl.groupby(['tissue', 'variant']).head(1)\n",
    "controls_trans_sqtl = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=os.path.basename(f).replace('.controls.txt', '')) for f in CONTROL_SNPS_TRANS_SQTL])\n",
    "controls_trans_sqtl = controls_trans_sqtl[controls_trans_sqtl.tissue.isin(USE_TISSUES)]\n",
    "controls_trans_sqtl = controls_trans_sqtl.groupby(['tissue', 'variant']).head(1)\n",
    "controls_trans_eqtl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:02.981815Z",
     "iopub.status.busy": "2024-12-03T19:16:02.981126Z",
     "iopub.status.idle": "2024-12-03T19:16:02.983834Z",
     "shell.execute_reply": "2024-12-03T19:16:02.983470Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "614"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(controls_trans_eqtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:02.988051Z",
     "iopub.status.busy": "2024-12-03T19:16:02.987382Z",
     "iopub.status.idle": "2024-12-03T19:16:02.991025Z",
     "shell.execute_reply": "2024-12-03T19:16:02.990482Z"
    }
   },
   "outputs": [],
   "source": [
    "assert(list(sorted(trans_eqtl[trans_eqtl.tissue=='Whole_blood'].variant_id.unique().tolist())) == list(sorted(controls_trans_eqtl[controls_trans_eqtl.tissue=='Whole_blood'].variant.to_list())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.096149Z",
     "iopub.status.busy": "2024-12-03T19:16:03.095006Z",
     "iopub.status.idle": "2024-12-03T19:16:03.668103Z",
     "shell.execute_reply": "2024-12-03T19:16:03.668928Z"
    }
   },
   "outputs": [],
   "source": [
    "# make annotation matrices\n",
    "annotation_matrices = dict()\n",
    "for tissue in USE_TISSUES:\n",
    "    tissue_trans_variants_and_controls = controls_trans_eqtl.loc[controls_trans_eqtl.tissue==tissue,['variant', 'control_variant']].values.flatten().tolist()\n",
    "    tissue_trans_variants_and_controls += controls_trans_sqtl.loc[controls_trans_sqtl.tissue==tissue,['variant', 'control_variant']].values.flatten().tolist()\n",
    "    tissue_trans_variants_and_controls = list(set(tissue_trans_variants_and_controls))\n",
    "    m = pd.DataFrame({'variant': tissue_trans_variants_and_controls})\n",
    "    m['ciseqtl'] = m.variant.isin(cis_eqtl[cis_eqtl.tissue==tissue].variant_id.unique())\n",
    "    m['cissqtl'] = m.variant.isin(cis_sqtl[cis_sqtl.tissue==tissue].variant_id.unique())\n",
    "    m['ciseqtl_or_cissqtl'] = m.ciseqtl | m.cissqtl\n",
    "    annotation_matrices[tissue] = m.set_index('variant').astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.695584Z",
     "iopub.status.busy": "2024-12-03T19:16:03.694470Z",
     "iopub.status.idle": "2024-12-03T19:16:03.698687Z",
     "shell.execute_reply": "2024-12-03T19:16:03.698352Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ciseqtl</th>\n",
       "      <th>cissqtl</th>\n",
       "      <th>ciseqtl_or_cissqtl</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr2_49050684_G_A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr16_10815232_C_A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX_43226118_G_T</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3_77722227_A_G</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr19_12419069_C_G</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ciseqtl  cissqtl  ciseqtl_or_cissqtl\n",
       "variant                                                 \n",
       "chr2_49050684_G_A         0        0                   0\n",
       "chr16_10815232_C_A        0        0                   0\n",
       "chrX_43226118_G_T         0        0                   0\n",
       "chr3_77722227_A_G         0        0                   0\n",
       "chr19_12419069_C_G        0        0                   0"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_matrices['Whole_blood'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.703681Z",
     "iopub.status.busy": "2024-12-03T19:16:03.702991Z",
     "iopub.status.idle": "2024-12-03T19:16:03.705289Z",
     "shell.execute_reply": "2024-12-03T19:16:03.704600Z"
    }
   },
   "outputs": [],
   "source": [
    "assert(all((annotation_matrices['Whole_blood'].loc[:,['ciseqtl', 'cissqtl']].sum(axis=1)>0).astype(int) == annotation_matrices['Whole_blood'].loc[:,'ciseqtl_or_cissqtl']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.713050Z",
     "iopub.status.busy": "2024-12-03T19:16:03.712083Z",
     "iopub.status.idle": "2024-12-03T19:16:03.761948Z",
     "shell.execute_reply": "2024-12-03T19:16:03.760864Z"
    }
   },
   "outputs": [],
   "source": [
    "# test for enrichment w/ cis-e/sQTL\n",
    "results = []\n",
    "for tissue in USE_TISSUES:\n",
    "    trans_eqtl_enrichment = fishers_exact(controls_trans_eqtl[controls_trans_eqtl.tissue==tissue].variant.to_list(), controls_trans_eqtl[controls_trans_eqtl.tissue==tissue].control_variant.to_list(), annotation_matrices[tissue])\n",
    "    trans_sqtl_enrichment = fishers_exact(controls_trans_sqtl[controls_trans_sqtl.tissue==tissue].variant.to_list(), controls_trans_sqtl[controls_trans_sqtl.tissue==tissue].control_variant.to_list(), annotation_matrices[tissue])\n",
    "    results.append(trans_eqtl_enrichment.assign(tissue=tissue, modality='transeqtl'))\n",
    "    results.append(trans_sqtl_enrichment.assign(tissue=tissue, modality='transsqtl'))\n",
    "results = pd.concat(results)\n",
    "results.to_csv(f'{PREFIX}basic-trans-enrichment-in-cis.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.771567Z",
     "iopub.status.busy": "2024-12-03T19:16:03.771004Z",
     "iopub.status.idle": "2024-12-03T19:16:03.773750Z",
     "shell.execute_reply": "2024-12-03T19:16:03.774020Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotation</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p</th>\n",
       "      <th>n_hits_overlapping_annotation</th>\n",
       "      <th>n_hits_not_overlapping_annotation</th>\n",
       "      <th>n_controls_overlapping_annotation</th>\n",
       "      <th>n_controls_not_overlapping_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>7.035625</td>\n",
       "      <td>4.978788e-27</td>\n",
       "      <td>163</td>\n",
       "      <td>451</td>\n",
       "      <td>30</td>\n",
       "      <td>584</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transeqtl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cissqtl</td>\n",
       "      <td>3.109282</td>\n",
       "      <td>1.398202e-07</td>\n",
       "      <td>82</td>\n",
       "      <td>532</td>\n",
       "      <td>29</td>\n",
       "      <td>585</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transeqtl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ciseqtl_or_cissqtl</td>\n",
       "      <td>5.781351</td>\n",
       "      <td>6.342585e-29</td>\n",
       "      <td>202</td>\n",
       "      <td>412</td>\n",
       "      <td>48</td>\n",
       "      <td>566</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transeqtl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ciseqtl</td>\n",
       "      <td>19.565217</td>\n",
       "      <td>2.188607e-04</td>\n",
       "      <td>15</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transsqtl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cissqtl</td>\n",
       "      <td>5.105769</td>\n",
       "      <td>5.366637e-02</td>\n",
       "      <td>9</td>\n",
       "      <td>52</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transsqtl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ciseqtl_or_cissqtl</td>\n",
       "      <td>8.746032</td>\n",
       "      <td>2.529502e-04</td>\n",
       "      <td>19</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>58</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transsqtl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           annotation  odds_ratio             p  \\\n",
       "0             ciseqtl    7.035625  4.978788e-27   \n",
       "1             cissqtl    3.109282  1.398202e-07   \n",
       "2  ciseqtl_or_cissqtl    5.781351  6.342585e-29   \n",
       "0             ciseqtl   19.565217  2.188607e-04   \n",
       "1             cissqtl    5.105769  5.366637e-02   \n",
       "2  ciseqtl_or_cissqtl    8.746032  2.529502e-04   \n",
       "\n",
       "   n_hits_overlapping_annotation  n_hits_not_overlapping_annotation  \\\n",
       "0                            163                                451   \n",
       "1                             82                                532   \n",
       "2                            202                                412   \n",
       "0                             15                                 46   \n",
       "1                              9                                 52   \n",
       "2                             19                                 42   \n",
       "\n",
       "   n_controls_overlapping_annotation  n_controls_not_overlapping_annotation  \\\n",
       "0                                 30                                    584   \n",
       "1                                 29                                    585   \n",
       "2                                 48                                    566   \n",
       "0                                  1                                     60   \n",
       "1                                  2                                     59   \n",
       "2                                  3                                     58   \n",
       "\n",
       "        tissue   modality  \n",
       "0  Whole_blood  transeqtl  \n",
       "1  Whole_blood  transeqtl  \n",
       "2  Whole_blood  transeqtl  \n",
       "0  Whole_blood  transsqtl  \n",
       "1  Whole_blood  transsqtl  \n",
       "2  Whole_blood  transsqtl  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.782552Z",
     "iopub.status.busy": "2024-12-03T19:16:03.782195Z",
     "iopub.status.idle": "2024-12-03T19:16:03.785272Z",
     "shell.execute_reply": "2024-12-03T19:16:03.784944Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotation</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p</th>\n",
       "      <th>n_hits_overlapping_annotation</th>\n",
       "      <th>n_hits_not_overlapping_annotation</th>\n",
       "      <th>n_controls_overlapping_annotation</th>\n",
       "      <th>n_controls_not_overlapping_annotation</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ciseqtl_or_cissqtl</td>\n",
       "      <td>5.781351</td>\n",
       "      <td>6.342585e-29</td>\n",
       "      <td>202</td>\n",
       "      <td>412</td>\n",
       "      <td>48</td>\n",
       "      <td>566</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>transeqtl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           annotation  odds_ratio             p  \\\n",
       "2  ciseqtl_or_cissqtl    5.781351  6.342585e-29   \n",
       "\n",
       "   n_hits_overlapping_annotation  n_hits_not_overlapping_annotation  \\\n",
       "2                            202                                412   \n",
       "\n",
       "   n_controls_overlapping_annotation  n_controls_not_overlapping_annotation  \\\n",
       "2                                 48                                    566   \n",
       "\n",
       "        tissue   modality  \n",
       "2  Whole_blood  transeqtl  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[(results.tissue=='Whole_blood') & (results.modality=='transeqtl') & (results.annotation=='ciseqtl_or_cissqtl')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.792373Z",
     "iopub.status.busy": "2024-12-03T19:16:03.791907Z",
     "iopub.status.idle": "2024-12-03T19:16:03.794512Z",
     "shell.execute_reply": "2024-12-03T19:16:03.794821Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32.9% of unique trans-eVariants overlap at least one cis-eQTL or cis-sQTL credible set in one tissue (31.1% for trans-sVariants)\n"
     ]
    }
   ],
   "source": [
    "x = results[(results.tissue=='Whole_blood') & (results.modality=='transeqtl') & (results.annotation=='ciseqtl_or_cissqtl')]\n",
    "y = results[(results.tissue=='Whole_blood') & (results.modality=='transsqtl') & (results.annotation=='ciseqtl_or_cissqtl')]\n",
    "assert(len(x) == 1)\n",
    "assert(len(y) == 1)\n",
    "print('{}% of unique trans-eVariants overlap at least one cis-eQTL or cis-sQTL credible set in one tissue ({}% for trans-sVariants)'.format(\n",
    "    round(100*x.n_hits_overlapping_annotation.values[0] / (x.n_hits_overlapping_annotation.values[0] + x.n_hits_not_overlapping_annotation.values[0]), 1),\n",
    "    round(100*y.n_hits_overlapping_annotation.values[0] / (y.n_hits_overlapping_annotation.values[0] + y.n_hits_not_overlapping_annotation.values[0]), 1)\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.799588Z",
     "iopub.status.busy": "2024-12-03T19:16:03.798482Z",
     "iopub.status.idle": "2024-12-03T19:16:03.800701Z",
     "shell.execute_reply": "2024-12-03T19:16:03.801474Z"
    }
   },
   "outputs": [],
   "source": [
    "# compare enrichment in the case where trans-eVariant is trans-eVariant for more than 1 trans-eGene vs only 1 trans-eGene\n",
    "# results = []\n",
    "\n",
    "# for tissue in USE_TISSUES:\n",
    "#     trans_evariants = trans_eqtl[trans_eqtl.tissue==tissue].variant_id.value_counts()\n",
    "#     trans_eqtl_for_more_than_one_gene = trans_evariants[trans_evariants>1].index.tolist()\n",
    "#     trans_eqtl_for_one_gene = trans_evariants[trans_evariants==1].index.tolist()\n",
    "\n",
    "#     trans_svariants = trans_sqtl[trans_sqtl.tissue==tissue].variant_id.value_counts()\n",
    "#     trans_sqtl_for_more_than_one_gene = trans_svariants[trans_svariants>1].index.tolist()\n",
    "#     trans_sqtl_for_one_gene = trans_svariants[trans_svariants==1].index.tolist()\n",
    "\n",
    "#     trans_eqtl_enrichment = fishers_exact(trans_eqtl_for_more_than_one_gene, trans_eqtl_for_one_gene, annotation_matrices[tissue])\n",
    "#     trans_sqtl_enrichment = fishers_exact(trans_sqtl_for_more_than_one_gene, trans_sqtl_for_one_gene, annotation_matrices[tissue])\n",
    "#     results.append(trans_eqtl_enrichment.assign(tissue=tissue, modality='transeqtl'))\n",
    "#     results.append(trans_sqtl_enrichment.assign(tissue=tissue, modality='transsqtl'))\n",
    "\n",
    "# results = pd.concat(results)\n",
    "# results.to_csv(f'{PREFIX}trans-enrichment-in-cis-more-than-one-transegene-vs-only-one-transgene.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.805324Z",
     "iopub.status.busy": "2024-12-03T19:16:03.804954Z",
     "iopub.status.idle": "2024-12-03T19:16:03.807254Z",
     "shell.execute_reply": "2024-12-03T19:16:03.806671Z"
    }
   },
   "outputs": [],
   "source": [
    "# tmp = trans_eqtl.groupby(['tissue', 'variant_id', 'af']).phenotype_id.nunique().rename('n_egenes').reset_index()\n",
    "# tmp['maf'] = np.minimum(tmp.af, 1-tmp.af)\n",
    "# tmp['evariant_for_more_than_one_gene'] = tmp.n_egenes > 1\n",
    "# tmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.810495Z",
     "iopub.status.busy": "2024-12-03T19:16:03.810007Z",
     "iopub.status.idle": "2024-12-03T19:16:03.812441Z",
     "shell.execute_reply": "2024-12-03T19:16:03.811954Z"
    }
   },
   "outputs": [],
   "source": [
    "# fig, ax = plt.subplots()\n",
    "\n",
    "# sns.stripplot(x='evariant_for_more_than_one_gene', y='maf', data=tmp, ax=ax)\n",
    "# sns.boxplot(x='evariant_for_more_than_one_gene', y='maf', data=tmp, ax=ax, color='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:03.816764Z",
     "iopub.status.busy": "2024-12-03T19:16:03.815666Z",
     "iopub.status.idle": "2024-12-03T19:16:04.660937Z",
     "shell.execute_reply": "2024-12-03T19:16:04.661783Z"
    }
   },
   "outputs": [],
   "source": [
    "# TODO: subset to genes that were considered in the paper that selected TFs. If gene isn't in the list, not clear if it's because it was considered and wasn't a TF or because it wasn't considered at all\n",
    "# TODO: subset to cis tested genes only?\n",
    "tf_list = set(pd.read_csv('/net/topmed10/working/porchard/rnaseq/data/tfs/TFs_Ensembl_v_1.01.txt', sep='\\t', header=None)[0])\n",
    "\n",
    "gene_is_tf = {i: strip_version(i) in tf_list for i in gtf_df.gene_id.unique() if 'ENS' in i}\n",
    "# not losing much by subsetting to genes that appear in the GO mappings\n",
    "# tmp = cis_eqtl.phenotype_id.isin(set(gene_is_tf.keys())).value_counts()\n",
    "# tmp/tmp.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:04.666827Z",
     "iopub.status.busy": "2024-12-03T19:16:04.665996Z",
     "iopub.status.idle": "2024-12-03T19:16:06.882871Z",
     "shell.execute_reply": "2024-12-03T19:16:06.882358Z"
    }
   },
   "outputs": [],
   "source": [
    "cis_eqtl['gene_id'] = cis_eqtl.phenotype_id\n",
    "cis_sqtl['gene_id'] = cis_sqtl.phenotype_id.map(phenotype_id_to_gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T19:16:06.891981Z",
     "iopub.status.busy": "2024-12-03T19:16:06.890832Z",
     "iopub.status.idle": "2024-12-03T19:16:45.435460Z",
     "shell.execute_reply": "2024-12-03T19:16:45.436143Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-12-03 14:16:07,477 - INFO: Permutation 0 of 1000\n",
      "2024-12-03 14:16:08,695 - INFO: Permutation 100 of 1000\n",
      "2024-12-03 14:16:09,914 - INFO: Permutation 200 of 1000\n",
      "2024-12-03 14:16:11,122 - INFO: Permutation 300 of 1000\n",
      "2024-12-03 14:16:12,332 - INFO: Permutation 400 of 1000\n",
      "2024-12-03 14:16:13,546 - INFO: Permutation 500 of 1000\n",
      "2024-12-03 14:16:14,742 - INFO: Permutation 600 of 1000\n",
      "2024-12-03 14:16:15,955 - INFO: Permutation 700 of 1000\n",
      "2024-12-03 14:16:17,167 - INFO: Permutation 800 of 1000\n",
      "2024-12-03 14:16:18,387 - INFO: Permutation 900 of 1000\n",
      "2024-12-03 14:16:20,096 - INFO: Permutation 0 of 1000\n",
      "2024-12-03 14:16:20,736 - INFO: Permutation 100 of 1000\n",
      "2024-12-03 14:16:21,373 - INFO: Permutation 200 of 1000\n",
      "2024-12-03 14:16:22,016 - INFO: Permutation 300 of 1000\n",
      "2024-12-03 14:16:22,657 - INFO: Permutation 400 of 1000\n",
      "2024-12-03 14:16:23,292 - INFO: Permutation 500 of 1000\n",
      "2024-12-03 14:16:23,929 - INFO: Permutation 600 of 1000\n",
      "2024-12-03 14:16:24,570 - INFO: Permutation 700 of 1000\n",
      "2024-12-03 14:16:25,204 - INFO: Permutation 800 of 1000\n",
      "2024-12-03 14:16:25,839 - INFO: Permutation 900 of 1000\n",
      "2024-12-03 14:16:26,841 - INFO: Permutation 0 of 1000\n",
      "2024-12-03 14:16:28,039 - INFO: Permutation 100 of 1000\n",
      "2024-12-03 14:16:29,233 - INFO: Permutation 200 of 1000\n",
      "2024-12-03 14:16:30,422 - INFO: Permutation 300 of 1000\n",
      "2024-12-03 14:16:31,614 - INFO: Permutation 400 of 1000\n",
      "2024-12-03 14:16:32,809 - INFO: Permutation 500 of 1000\n",
      "2024-12-03 14:16:34,007 - INFO: Permutation 600 of 1000\n",
      "2024-12-03 14:16:35,220 - INFO: Permutation 700 of 1000\n",
      "2024-12-03 14:16:36,414 - INFO: Permutation 800 of 1000\n",
      "2024-12-03 14:16:37,597 - INFO: Permutation 900 of 1000\n",
      "2024-12-03 14:16:39,044 - INFO: Permutation 0 of 1000\n",
      "2024-12-03 14:16:39,669 - INFO: Permutation 100 of 1000\n",
      "2024-12-03 14:16:40,293 - INFO: Permutation 200 of 1000\n",
      "2024-12-03 14:16:40,919 - INFO: Permutation 300 of 1000\n",
      "2024-12-03 14:16:41,548 - INFO: Permutation 400 of 1000\n",
      "2024-12-03 14:16:42,176 - INFO: Permutation 500 of 1000\n",
      "2024-12-03 14:16:42,801 - INFO: Permutation 600 of 1000\n",
      "2024-12-03 14:16:43,518 - INFO: Permutation 700 of 1000\n",
      "2024-12-03 14:16:44,147 - INFO: Permutation 800 of 1000\n",
      "2024-12-03 14:16:44,786 - INFO: Permutation 900 of 1000\n"
     ]
    }
   ],
   "source": [
    "# use only unique variants -- otherwise, if trans-eVariant w/ many trans-eGenes becomes cis-eQTL for TF in subset of permutations, get weird upper tail density\n",
    "# Amongst genes w/ at least one eQTL credible set, TF genes don't have more credible sets or bigger credible sets\n",
    "# so can use permutation test amongst cis-eGenes to test for significance here\n",
    "# use whole blood only?\n",
    "\n",
    "REPS = 1000\n",
    "\n",
    "observed_all = []\n",
    "permuted_all = []\n",
    "\n",
    "for tissue in USE_TISSUES:\n",
    "    for trans in ['trans_eqtl', 'trans_sqtl']:\n",
    "        for cis in ['cis_eqtl', 'cis_sqtl']:\n",
    "            trans_df = trans_eqtl if trans == 'trans_eqtl' else trans_sqtl\n",
    "            cis_df = cis_eqtl if cis == 'cis_eqtl' else cis_sqtl\n",
    "            tissue_trans_variants = trans_df[(trans_df.tissue==tissue)].variant_id.unique()\n",
    "            tissue_trans_variants_to_cis_genes = cis_df.loc[(cis_df.tissue==tissue) & (cis_df.variant_id.isin(tissue_trans_variants)),['variant_id', 'gene_id']].drop_duplicates()\n",
    "\n",
    "            cis_egene_is_tf = {i: gene_is_tf[i] for i in cis_df[cis_df.tissue==tissue].gene_id.unique() if i in gene_is_tf}\n",
    "            tissue_trans_variants_to_cis_genes['gene_is_tf'] = tissue_trans_variants_to_cis_genes.gene_id.map(cis_egene_is_tf).astype(bool)\n",
    "\n",
    "            observed = tissue_trans_variants_to_cis_genes.groupby('variant_id').gene_is_tf.max().sum()\n",
    "            observed_all.append([tissue, trans, cis, observed])\n",
    "            \n",
    "            for i in range(REPS):\n",
    "                if i % 100 == 0:\n",
    "                    logging.info(f'Permutation {i} of {REPS}')\n",
    "                # shuffle gene --> is_tf assignments\n",
    "                tmp = shuffle_dict(cis_egene_is_tf)\n",
    "                tissue_trans_variants_to_cis_genes['gene_is_tf'] = tissue_trans_variants_to_cis_genes.gene_id.map(tmp).astype(bool)\n",
    "                stat = tissue_trans_variants_to_cis_genes.groupby('variant_id').gene_is_tf.max().sum()\n",
    "                permuted_all.append([tissue, trans, cis, stat])\n",
    "\n",
    "observed = pd.DataFrame(observed_all, columns=['tissue', 'trans_modality', 'cis_modality', 'n']).assign(c='observed')\n",
    "permuted = pd.DataFrame(permuted_all, columns=['tissue', 'trans_modality', 'cis_modality', 'n']).assign(c='permuted')\n",
    "cisgene_is_tf_permutations = pd.concat([observed, permuted])\n",
    "cisgene_is_tf_permutations.to_csv(f'{PREFIX}cisgene_is_tf_permutations.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2020-09",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
