{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:00.726149Z",
     "iopub.status.busy": "2024-12-03T18:33:00.723096Z",
     "iopub.status.idle": "2024-12-03T18:33:05.345451Z",
     "shell.execute_reply": "2024-12-03T18:33:05.345950Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/net/snowwhite/home/porchard/anaconda3/envs/2020-09/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n",
      "/net/snowwhite/home/porchard/anaconda3/envs/2020-09/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from topmed_manuscript_clean import gtf_to_df, phenotype_id_to_gene_id, format_modality\n",
    "import sys\n",
    "from scipy.stats import hypergeom\n",
    "from gprofiler import GProfiler\n",
    "gp = GProfiler(return_dataframe=True)\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import re\n",
    "import glob\n",
    "\n",
    "\n",
    "def parse_modality(s):\n",
    "    RE = '((cis|trans)-?([es]qtl))'\n",
    "    m = re.search(RE, s, flags=re.IGNORECASE)\n",
    "    assert(m is not None)\n",
    "    return m.group(0)\n",
    "\n",
    "\n",
    "def parse_tissue(s):\n",
    "    tissues = ['Whole_blood', 'Lung', 'Nasal_epithelial', 'PBMC', 'Monocyte', 'T_cell']\n",
    "    RE = '(' + '|'.join(tissues) + ')'\n",
    "    m = re.search(RE, s)\n",
    "    assert(m is not None)\n",
    "    return m.group(0)\n",
    "\n",
    "def genrich(genes, background):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    #from scipy.stats import hypergeom\n",
    "    # M is the total number of objects, n is total number of Type I objects. The random variate represents the number of Type I objects in N drawn without replacement from the total population.\n",
    "    enrich = gp.profile(organism='hsapiens', query=genes, all_results=True, background=background)\n",
    "    enrich['nominal_p'] = hypergeom.sf(enrich.intersection_size-1, enrich.effective_domain_size, enrich.query_size, enrich.term_size)\n",
    "    enrich['nominal_p'] = [nominal_p if nominal_p > 0 else p_value for nominal_p, p_value in zip(enrich.nominal_p, enrich.p_value)]\n",
    "    enrich['obs_exp'] = (enrich.intersection_size / enrich.term_size) / (enrich.query_size / enrich.effective_domain_size)\n",
    "    enrich['log2_enrich'] = np.log2(enrich.obs_exp)\n",
    "    enrich['-log10(q)'] = -1*np.log10(enrich.p_value)\n",
    "    enrich['-log10(p)'] = -1*np.log10(enrich.nominal_p)\n",
    "    if enrich[enrich.significant].obs_exp.min() <= 1:\n",
    "        sys.stderr.write('Warning: observed / expected for significant enriched items not always > 1?')\n",
    "        sys.stderr.flush()\n",
    "    if enrich[enrich.significant].nominal_p.max() > 0.05:\n",
    "        sys.stderr.write('Warning: nominal p for significant enriched itens not always < 0.05?')\n",
    "        sys.stderr.flush()\n",
    "    if not all(enrich[enrich.significant].nominal_p <= enrich[enrich.significant].p_value):\n",
    "        sys.stderr.write('Warning: nominal p for significant enriched items sometimes > corrected p?')\n",
    "        sys.stderr.flush()\n",
    "    return enrich\n",
    "\n",
    "\n",
    "PREFIX = 'trans-go-enrichment.'\n",
    "\n",
    "GTF = '../data/gtf/gencode.v30.GRCh38.ERCC.genes.collapsed_only.gtf.gz'\n",
    "gtf_df = gtf_to_df(GTF, parse_attributes=['gene_id', 'gene_name'])\n",
    "gtf_df = gtf_df[gtf_df.feature=='gene']\n",
    "gene_id_to_gene_name = dict(zip(gtf_df.gene_id, gtf_df.gene_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:05.350698Z",
     "iopub.status.busy": "2024-12-03T18:33:05.350196Z",
     "iopub.status.idle": "2024-12-03T18:33:05.356028Z",
     "shell.execute_reply": "2024-12-03T18:33:05.355463Z"
    }
   },
   "outputs": [],
   "source": [
    "full_trans_permutation_files = pd.DataFrame({'file': glob.glob('../../manuscript-intermediate-processing/data/scan-results/joint/trans-*qtl/maf005/trans-top/*')})\n",
    "full_trans_permutation_files['tissue'] = full_trans_permutation_files.file.map(parse_tissue)\n",
    "full_trans_permutation_files['modality'] = full_trans_permutation_files.file.map(lambda x: format_modality(parse_modality(x)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:05.361308Z",
     "iopub.status.busy": "2024-12-03T18:33:05.360293Z",
     "iopub.status.idle": "2024-12-03T18:33:06.007823Z",
     "shell.execute_reply": "2024-12-03T18:33:06.008142Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>af</th>\n",
       "      <th>gene_mappability</th>\n",
       "      <th>gene_crossmaps_to_gene_near_variant</th>\n",
       "      <th>biotype</th>\n",
       "      <th>...</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>pval_beta_no_zero</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>phenotypes_tested_for_gene</th>\n",
       "      <th>pval_beta_corrected_across_phenotypes</th>\n",
       "      <th>pval_beta_corrected_across_phenotypes_no_zero</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr14_83387951_A_T</td>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>7.924750e-06</td>\n",
       "      <td>0.101930</td>\n",
       "      <td>0.022799</td>\n",
       "      <td>0.003120</td>\n",
       "      <td>0.190889</td>\n",
       "      <td>0.958185</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.999982e-01</td>\n",
       "      <td>9.999982e-01</td>\n",
       "      <td>9.999999e-01</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7_132431526_AG_A</td>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>3.256160e-07</td>\n",
       "      <td>-0.144682</td>\n",
       "      <td>0.028295</td>\n",
       "      <td>0.004078</td>\n",
       "      <td>0.119538</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>0.469577</td>\n",
       "      <td>4.675975e-01</td>\n",
       "      <td>4.675975e-01</td>\n",
       "      <td>9.999999e-01</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX_26802742_G_A</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>3.075770e-06</td>\n",
       "      <td>0.076452</td>\n",
       "      <td>0.016371</td>\n",
       "      <td>0.003403</td>\n",
       "      <td>0.657887</td>\n",
       "      <td>0.967695</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>0.994150</td>\n",
       "      <td>9.940261e-01</td>\n",
       "      <td>9.940261e-01</td>\n",
       "      <td>9.999999e-01</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4_106906657_G_A</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>3.760130e-07</td>\n",
       "      <td>-0.125686</td>\n",
       "      <td>0.024712</td>\n",
       "      <td>0.004034</td>\n",
       "      <td>0.155330</td>\n",
       "      <td>0.973241</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>0.496775</td>\n",
       "      <td>4.957646e-01</td>\n",
       "      <td>4.957646e-01</td>\n",
       "      <td>9.999999e-01</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7_50330658_C_T</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>3.503050e-25</td>\n",
       "      <td>-0.234596</td>\n",
       "      <td>0.022533</td>\n",
       "      <td>0.016690</td>\n",
       "      <td>0.235203</td>\n",
       "      <td>0.999762</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>3.457262e-18</td>\n",
       "      <td>3.457262e-18</td>\n",
       "      <td>1.392881e-16</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            variant_id        phenotype_id          pval         b      b_se  \\\n",
       "0   chr14_83387951_A_T  ENSG00000000003.14  7.924750e-06  0.101930  0.022799   \n",
       "1  chr7_132431526_AG_A  ENSG00000000419.12  3.256160e-07 -0.144682  0.028295   \n",
       "2    chrX_26802742_G_A  ENSG00000000457.14  3.075770e-06  0.076452  0.016371   \n",
       "3   chr4_106906657_G_A  ENSG00000000460.17  3.760130e-07 -0.125686  0.024712   \n",
       "4    chr7_50330658_C_T  ENSG00000000938.13  3.503050e-25 -0.234596  0.022533   \n",
       "\n",
       "         r2        af  gene_mappability  gene_crossmaps_to_gene_near_variant  \\\n",
       "0  0.003120  0.190889          0.958185                                False   \n",
       "1  0.004078  0.119538          1.000000                                False   \n",
       "2  0.003403  0.657887          0.967695                                False   \n",
       "3  0.004034  0.155330          0.973241                                False   \n",
       "4  0.016690  0.235203          0.999762                                False   \n",
       "\n",
       "          biotype  ... pval_perm     pval_beta  pval_beta_no_zero  \\\n",
       "0  protein_coding  ...  1.000000  9.999982e-01       9.999982e-01   \n",
       "1  protein_coding  ...  0.469577  4.675975e-01       4.675975e-01   \n",
       "2  protein_coding  ...  0.994150  9.940261e-01       9.940261e-01   \n",
       "3  protein_coding  ...  0.496775  4.957646e-01       4.957646e-01   \n",
       "4  protein_coding  ...  0.000050  3.457262e-18       3.457262e-18   \n",
       "\n",
       "         qvalue       tissue    modality  gene_id  phenotypes_tested_for_gene  \\\n",
       "0  9.999999e-01  Whole_blood  trans-eQTL      NaN                         NaN   \n",
       "1  9.999999e-01  Whole_blood  trans-eQTL      NaN                         NaN   \n",
       "2  9.999999e-01  Whole_blood  trans-eQTL      NaN                         NaN   \n",
       "3  9.999999e-01  Whole_blood  trans-eQTL      NaN                         NaN   \n",
       "4  1.392881e-16  Whole_blood  trans-eQTL      NaN                         NaN   \n",
       "\n",
       "   pval_beta_corrected_across_phenotypes  \\\n",
       "0                                    NaN   \n",
       "1                                    NaN   \n",
       "2                                    NaN   \n",
       "3                                    NaN   \n",
       "4                                    NaN   \n",
       "\n",
       "   pval_beta_corrected_across_phenotypes_no_zero  \n",
       "0                                            NaN  \n",
       "1                                            NaN  \n",
       "2                                            NaN  \n",
       "3                                            NaN  \n",
       "4                                            NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_trans_permutations = pd.concat([pd.read_csv(f, sep='\\t').assign(tissue=tissue, modality=modality) for f, tissue, modality in zip(full_trans_permutation_files.file, full_trans_permutation_files.tissue, full_trans_permutation_files.modality)])\n",
    "full_trans_permutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:06.022837Z",
     "iopub.status.busy": "2024-12-03T18:33:06.022472Z",
     "iopub.status.idle": "2024-12-03T18:33:06.065961Z",
     "shell.execute_reply": "2024-12-03T18:33:06.064926Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9.98102e-06"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_trans_permutations[full_trans_permutations.tissue=='Whole_blood'].pval.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:06.071580Z",
     "iopub.status.busy": "2024-12-03T18:33:06.070644Z",
     "iopub.status.idle": "2024-12-03T18:33:06.098574Z",
     "shell.execute_reply": "2024-12-03T18:33:06.099334Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>af</th>\n",
       "      <th>gene_mappability</th>\n",
       "      <th>gene_crossmaps_to_gene_near_variant</th>\n",
       "      <th>biotype</th>\n",
       "      <th>...</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>true_df</th>\n",
       "      <th>pval_true_df</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>pval_beta_no_zero</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>tissue</th>\n",
       "      <th>clumped_variant_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7_50330658_C_T</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>3.503050e-25</td>\n",
       "      <td>-0.234596</td>\n",
       "      <td>0.022533</td>\n",
       "      <td>0.016690</td>\n",
       "      <td>0.235203</td>\n",
       "      <td>0.999762</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1.005841</td>\n",
       "      <td>9.795177e+05</td>\n",
       "      <td>6086.276059</td>\n",
       "      <td>4.468785e-24</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>3.457262e-18</td>\n",
       "      <td>3.457262e-18</td>\n",
       "      <td>1.392881e-16</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr7_50342615_A_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_156302480_C_T</td>\n",
       "      <td>ENSG00000002330.13</td>\n",
       "      <td>1.027770e-13</td>\n",
       "      <td>0.134956</td>\n",
       "      <td>0.018106</td>\n",
       "      <td>0.008625</td>\n",
       "      <td>0.500775</td>\n",
       "      <td>0.999556</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1.005125</td>\n",
       "      <td>1.008630e+06</td>\n",
       "      <td>6081.962583</td>\n",
       "      <td>3.926652e-13</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>3.664337e-07</td>\n",
       "      <td>3.664337e-07</td>\n",
       "      <td>5.779733e-06</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr1_156344836_A_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6_144036619_C_A</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>1.984040e-17</td>\n",
       "      <td>0.331091</td>\n",
       "      <td>0.038865</td>\n",
       "      <td>0.011237</td>\n",
       "      <td>0.055934</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1.012491</td>\n",
       "      <td>9.587406e+05</td>\n",
       "      <td>6060.399614</td>\n",
       "      <td>1.281332e-16</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>9.188711e-11</td>\n",
       "      <td>9.188711e-11</td>\n",
       "      <td>2.100000e-09</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr6_144036619_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6_122440739_T_C</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>1.034670e-10</td>\n",
       "      <td>0.126184</td>\n",
       "      <td>0.019495</td>\n",
       "      <td>0.006518</td>\n",
       "      <td>0.706539</td>\n",
       "      <td>0.960425</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1.003425</td>\n",
       "      <td>1.105596e+06</td>\n",
       "      <td>6130.610651</td>\n",
       "      <td>2.431723e-10</td>\n",
       "      <td>0.00025</td>\n",
       "      <td>2.609707e-04</td>\n",
       "      <td>2.609707e-04</td>\n",
       "      <td>2.773133e-03</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr6_122440739_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22_46290431_C_G</td>\n",
       "      <td>ENSG00000004799.8</td>\n",
       "      <td>7.740950e-13</td>\n",
       "      <td>0.207077</td>\n",
       "      <td>0.028839</td>\n",
       "      <td>0.008009</td>\n",
       "      <td>0.107530</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1.012491</td>\n",
       "      <td>9.587406e+05</td>\n",
       "      <td>6060.399614</td>\n",
       "      <td>2.939505e-12</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>2.389651e-06</td>\n",
       "      <td>2.389651e-06</td>\n",
       "      <td>3.403684e-05</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr22_46290431_C_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           variant_id        phenotype_id          pval         b      b_se  \\\n",
       "0   chr7_50330658_C_T  ENSG00000000938.13  3.503050e-25 -0.234596  0.022533   \n",
       "1  chr1_156302480_C_T  ENSG00000002330.13  1.027770e-13  0.134956  0.018106   \n",
       "2  chr6_144036619_C_A  ENSG00000004059.11  1.984040e-17  0.331091  0.038865   \n",
       "3  chr6_122440739_T_C   ENSG00000004478.8  1.034670e-10  0.126184  0.019495   \n",
       "4  chr22_46290431_C_G   ENSG00000004799.8  7.740950e-13  0.207077  0.028839   \n",
       "\n",
       "         r2        af  gene_mappability  gene_crossmaps_to_gene_near_variant  \\\n",
       "0  0.016690  0.235203          0.999762                                False   \n",
       "1  0.008625  0.500775          0.999556                                False   \n",
       "2  0.011237  0.055934          1.000000                                False   \n",
       "3  0.006518  0.706539          0.960425                                False   \n",
       "4  0.008009  0.107530          1.000000                                False   \n",
       "\n",
       "          biotype  ... beta_shape1   beta_shape2      true_df  pval_true_df  \\\n",
       "0  protein_coding  ...    1.005841  9.795177e+05  6086.276059  4.468785e-24   \n",
       "1  protein_coding  ...    1.005125  1.008630e+06  6081.962583  3.926652e-13   \n",
       "2  protein_coding  ...    1.012491  9.587406e+05  6060.399614  1.281332e-16   \n",
       "3  protein_coding  ...    1.003425  1.105596e+06  6130.610651  2.431723e-10   \n",
       "4  protein_coding  ...    1.012491  9.587406e+05  6060.399614  2.939505e-12   \n",
       "\n",
       "   pval_perm     pval_beta  pval_beta_no_zero        qvalue       tissue  \\\n",
       "0    0.00005  3.457262e-18       3.457262e-18  1.392881e-16  Whole_blood   \n",
       "1    0.00005  3.664337e-07       3.664337e-07  5.779733e-06  Whole_blood   \n",
       "2    0.00005  9.188711e-11       9.188711e-11  2.100000e-09  Whole_blood   \n",
       "3    0.00025  2.609707e-04       2.609707e-04  2.773133e-03  Whole_blood   \n",
       "4    0.00005  2.389651e-06       2.389651e-06  3.403684e-05  Whole_blood   \n",
       "\n",
       "   clumped_variant_id  \n",
       "0   chr7_50342615_A_G  \n",
       "1  chr1_156344836_A_G  \n",
       "2  chr6_144036619_C_A  \n",
       "3  chr6_122440739_T_C  \n",
       "4  chr22_46290431_C_G  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_eqtl_clumped = pd.read_csv('../work/clump-trans-variants/clump-trans-signals.significant-trans-eqtl-clumped.tsv', sep='\\t')\n",
    "trans_sqtl_clumped = pd.read_csv('../work/clump-trans-variants//clump-trans-signals.significant-trans-sqtl-clumped.tsv', sep='\\t')\n",
    "trans_eqtl_clumped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:06.105144Z",
     "iopub.status.busy": "2024-12-03T18:33:06.103685Z",
     "iopub.status.idle": "2024-12-03T18:33:06.119414Z",
     "shell.execute_reply": "2024-12-03T18:33:06.118868Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>tissue</th>\n",
       "      <th>clumped_variant_id</th>\n",
       "      <th>modality</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr7_50342615_A_G</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000002330.13</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr1_156344836_A_G</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>ENSG00000002330.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr6_144036619_C_A</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr6_122440739_T_C</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004799.8</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>chr22_46290431_C_G</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>ENSG00000004799.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         phenotype_id       tissue  clumped_variant_id    modality  \\\n",
       "0  ENSG00000000938.13  Whole_blood   chr7_50342615_A_G  trans-eQTL   \n",
       "1  ENSG00000002330.13  Whole_blood  chr1_156344836_A_G  trans-eQTL   \n",
       "2  ENSG00000004059.11  Whole_blood  chr6_144036619_C_A  trans-eQTL   \n",
       "3   ENSG00000004478.8  Whole_blood  chr6_122440739_T_C  trans-eQTL   \n",
       "4   ENSG00000004799.8  Whole_blood  chr22_46290431_C_G  trans-eQTL   \n",
       "\n",
       "              gene_id  \n",
       "0  ENSG00000000938.13  \n",
       "1  ENSG00000002330.13  \n",
       "2  ENSG00000004059.11  \n",
       "3   ENSG00000004478.8  \n",
       "4   ENSG00000004799.8  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_clumped = pd.concat([trans_eqtl_clumped[['phenotype_id', 'tissue', 'clumped_variant_id']].assign(modality='trans-eQTL'), trans_sqtl_clumped[['phenotype_id', 'tissue', 'clumped_variant_id']].assign(modality='trans-sQTL')])\n",
    "trans_clumped['gene_id'] = trans_clumped.phenotype_id.map(phenotype_id_to_gene_id)\n",
    "trans_clumped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:33:06.130998Z",
     "iopub.status.busy": "2024-12-03T18:33:06.130534Z",
     "iopub.status.idle": "2024-12-03T18:36:04.058427Z",
     "shell.execute_reply": "2024-12-03T18:36:04.057851Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Whole_blood trans-eQTL chr10_63278270_G_C\n",
      "Whole_blood trans-eQTL chr12_54292096_C_T\n",
      "Whole_blood trans-eQTL chr14_35132881_T_G\n",
      "Whole_blood trans-eQTL chr17_16268373_CTT_C\n",
      "Whole_blood trans-eQTL chr17_28862756_A_G\n",
      "Whole_blood trans-eQTL chr17_35548243_T_A\n",
      "Whole_blood trans-eQTL chr17_64066984_C_A\n",
      "Whole_blood trans-eQTL chr1_156344836_A_G\n",
      "Whole_blood trans-eQTL chr1_158840522_C_T\n",
      "Whole_blood trans-eQTL chr1_159204893_T_C\n",
      "Whole_blood trans-eQTL chr2_159571135_C_T\n",
      "Whole_blood trans-eQTL chr3_101391301_T_A\n",
      "Whole_blood trans-eQTL chr3_56815721_T_C\n",
      "Whole_blood trans-eQTL chr5_173778315_G_A\n",
      "Whole_blood trans-eQTL chr6_139520516_T_C\n",
      "Whole_blood trans-eQTL chr6_144036619_C_A\n",
      "Whole_blood trans-eQTL chr6_154833905_C_T\n",
      "Whole_blood trans-eQTL chr7_28684757_G_T\n",
      "Whole_blood trans-eQTL chr7_50255709_CT_C\n",
      "Whole_blood trans-eQTL chr7_50336475_T_C\n",
      "Whole_blood trans-eQTL chr7_50337162_A_G\n",
      "Whole_blood trans-eQTL chr7_50342615_A_G\n",
      "Whole_blood trans-eQTL chr7_50360284_G_A\n",
      "Whole_blood trans-eQTL chr8_60733792_C_CA\n",
      "Whole_blood trans-eQTL chr9_96430637_A_C\n",
      "Whole_blood trans-sQTL chr3_56815721_T_C\n",
      "Whole_blood trans-sQTL chr7_50342615_A_G\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>native</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>term_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>query</th>\n",
       "      <th>parents</th>\n",
       "      <th>nominal_p</th>\n",
       "      <th>obs_exp</th>\n",
       "      <th>log2_enrich</th>\n",
       "      <th>-log10(p)</th>\n",
       "      <th>qval</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>clumped_variant_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:05418</td>\n",
       "      <td>Fluid shear stress and atherosclerosis</td>\n",
       "      <td>Fluid shear stress and atherosclerosis</td>\n",
       "      <td>102</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.019608</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>0.002266</td>\n",
       "      <td>27.519608</td>\n",
       "      <td>4.782388</td>\n",
       "      <td>2.644773</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr10_63278270_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:04510</td>\n",
       "      <td>Focal adhesion</td>\n",
       "      <td>Focal adhesion</td>\n",
       "      <td>144</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.013889</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>0.004457</td>\n",
       "      <td>19.493056</td>\n",
       "      <td>4.284888</td>\n",
       "      <td>2.350934</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr10_63278270_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2738</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:04015</td>\n",
       "      <td>Rap1 signaling pathway</td>\n",
       "      <td>Rap1 signaling pathway</td>\n",
       "      <td>149</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.006711</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>0.101263</td>\n",
       "      <td>9.419463</td>\n",
       "      <td>3.235645</td>\n",
       "      <td>0.994548</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr10_63278270_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2740</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:2000353</td>\n",
       "      <td>positive regulation of endothelial cell apopto...</td>\n",
       "      <td>\"Any process that activates or increases the f...</td>\n",
       "      <td>18</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0072577, GO:1904037, GO:2000351]</td>\n",
       "      <td>0.012755</td>\n",
       "      <td>77.972222</td>\n",
       "      <td>6.284888</td>\n",
       "      <td>1.894307</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr10_63278270_G_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2741</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:2000351</td>\n",
       "      <td>regulation of endothelial cell apoptotic process</td>\n",
       "      <td>\"Any process that modulates the frequency, rat...</td>\n",
       "      <td>38</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0072577, GO:1904035]</td>\n",
       "      <td>0.026756</td>\n",
       "      <td>36.934211</td>\n",
       "      <td>5.206886</td>\n",
       "      <td>1.572577</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr10_63278270_G_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source      native                                               name  \\\n",
       "3      KEGG  KEGG:05418             Fluid shear stress and atherosclerosis   \n",
       "14     KEGG  KEGG:04510                                     Focal adhesion   \n",
       "2738   KEGG  KEGG:04015                             Rap1 signaling pathway   \n",
       "2740  GO:BP  GO:2000353  positive regulation of endothelial cell apopto...   \n",
       "2741  GO:BP  GO:2000351   regulation of endothelial cell apoptotic process   \n",
       "\n",
       "                                            description  term_size  \\\n",
       "3                Fluid shear stress and atherosclerosis        102   \n",
       "14                                       Focal adhesion        144   \n",
       "2738                             Rap1 signaling pathway        149   \n",
       "2740  \"Any process that activates or increases the f...         18   \n",
       "2741  \"Any process that modulates the frequency, rat...         38   \n",
       "\n",
       "      query_size  intersection_size  effective_domain_size  precision  \\\n",
       "3             10                  2                  14035        0.2   \n",
       "14            10                  2                  14035        0.2   \n",
       "2738          10                  1                  14035        0.1   \n",
       "2740          10                  1                  14035        0.1   \n",
       "2741          10                  1                  14035        0.1   \n",
       "\n",
       "        recall    query                               parents  nominal_p  \\\n",
       "3     0.019608  query_1                          [KEGG:00000]   0.002266   \n",
       "14    0.013889  query_1                          [KEGG:00000]   0.004457   \n",
       "2738  0.006711  query_1                          [KEGG:00000]   0.101263   \n",
       "2740  0.055556  query_1  [GO:0072577, GO:1904037, GO:2000351]   0.012755   \n",
       "2741  0.026316  query_1              [GO:0072577, GO:1904035]   0.026756   \n",
       "\n",
       "        obs_exp  log2_enrich  -log10(p)  qval       tissue    modality  \\\n",
       "3     27.519608     4.782388   2.644773   1.0  Whole_blood  trans-eQTL   \n",
       "14    19.493056     4.284888   2.350934   1.0  Whole_blood  trans-eQTL   \n",
       "2738   9.419463     3.235645   0.994548   1.0  Whole_blood  trans-eQTL   \n",
       "2740  77.972222     6.284888   1.894307   1.0  Whole_blood  trans-eQTL   \n",
       "2741  36.934211     5.206886   1.572577   1.0  Whole_blood  trans-eQTL   \n",
       "\n",
       "      clumped_variant_id  \n",
       "3     chr10_63278270_G_C  \n",
       "14    chr10_63278270_G_C  \n",
       "2738  chr10_63278270_G_C  \n",
       "2740  chr10_63278270_G_C  \n",
       "2741  chr10_63278270_G_C  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: use bonferroni correction\n",
    "results = []\n",
    "for (tissue, modality, clumped_variant_id), df in trans_clumped.groupby(['tissue', 'modality', 'clumped_variant_id']):\n",
    "    genes = df['gene_id'].to_list()\n",
    "    if len(genes) >= 10 and tissue == 'Whole_blood':\n",
    "        print(tissue, modality, clumped_variant_id)\n",
    "        background = full_trans_permutations[(full_trans_permutations.tissue==tissue) & (full_trans_permutations.modality==modality)].phenotype_id.map(phenotype_id_to_gene_id).to_list()\n",
    "        go = genrich([i.split('.')[0] for i in genes], [i.split('.')[0] for i in background])\n",
    "        go = go[go.source.isin(['GO:BP', 'KEGG'])]\n",
    "        go = go[go.term_size>=5]\n",
    "        go = go[go.term_size<=1000]\n",
    "        go['qval'] = multipletests(go.nominal_p, method='bonferroni')[1]\n",
    "        go = go.drop(columns=['-log10(q)', 'significant', 'p_value'])\n",
    "        results.append(go.assign(tissue=tissue, modality=modality, clumped_variant_id=clumped_variant_id))\n",
    "results = pd.concat(results)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.067057Z",
     "iopub.status.busy": "2024-12-03T18:36:04.066049Z",
     "iopub.status.idle": "2024-12-03T18:36:04.078993Z",
     "shell.execute_reply": "2024-12-03T18:36:04.077866Z"
    }
   },
   "outputs": [],
   "source": [
    "results = results[(results.tissue=='Whole_blood') & (results.modality=='trans-eQTL')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.086768Z",
     "iopub.status.busy": "2024-12-03T18:36:04.085633Z",
     "iopub.status.idle": "2024-12-03T18:36:04.089419Z",
     "shell.execute_reply": "2024-12-03T18:36:04.089697Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 of 25 such groups showed significant GO:BP or KEGG pathway enrichments\n"
     ]
    }
   ],
   "source": [
    "print('{} of {} such groups showed significant GO:BP or KEGG pathway enrichments'.format(results[results.qval<=0.05].clumped_variant_id.nunique(), results.clumped_variant_id.nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.094846Z",
     "iopub.status.busy": "2024-12-03T18:36:04.093876Z",
     "iopub.status.idle": "2024-12-03T18:36:04.148838Z",
     "shell.execute_reply": "2024-12-03T18:36:04.148274Z"
    }
   },
   "outputs": [],
   "source": [
    "#pruned_results = results[(results.qval<=0.05) & (results.tissue=='Whole_blood') & (results.term_size<=1000) & (results.intersection_size>=(np.maximum(results.query_size/10, 3)))]\n",
    "#pruned_results.to_csv(f'tables/{PREFIX}results.tsv', sep='\\t', index=False)\n",
    "results[(results.qval<=0.05)].rename(columns={'qval': 'bonferroni_p'}).to_csv(f'tables/{PREFIX}results.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.157773Z",
     "iopub.status.busy": "2024-12-03T18:36:04.155486Z",
     "iopub.status.idle": "2024-12-03T18:36:04.160779Z",
     "shell.execute_reply": "2024-12-03T18:36:04.161411Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tissue       clumped_variant_id  \n",
       "Whole_blood  chr12_54292096_C_T       1\n",
       "             chr14_35132881_T_G       1\n",
       "             chr17_16268373_CTT_C    34\n",
       "             chr17_35548243_T_A       5\n",
       "             chr17_64066984_C_A       2\n",
       "             chr1_156344836_A_G       3\n",
       "             chr1_158840522_C_T       5\n",
       "             chr3_101391301_T_A       1\n",
       "             chr3_56815721_T_C       19\n",
       "             chr6_139520516_T_C       1\n",
       "             chr6_144036619_C_A       2\n",
       "             chr7_50255709_CT_C       3\n",
       "             chr7_50342615_A_G        1\n",
       "             chr9_96430637_A_C        3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[(results.qval<=0.05)].groupby(['tissue', 'clumped_variant_id']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.169240Z",
     "iopub.status.busy": "2024-12-03T18:36:04.165882Z",
     "iopub.status.idle": "2024-12-03T18:36:04.181637Z",
     "shell.execute_reply": "2024-12-03T18:36:04.180866Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>native</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>term_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>query</th>\n",
       "      <th>parents</th>\n",
       "      <th>nominal_p</th>\n",
       "      <th>obs_exp</th>\n",
       "      <th>log2_enrich</th>\n",
       "      <th>-log10(p)</th>\n",
       "      <th>qval</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>clumped_variant_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:04141</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>138</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>0.036232</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>1.032438e-07</td>\n",
       "      <td>39.116499</td>\n",
       "      <td>5.289705</td>\n",
       "      <td>6.986136</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr17_64066984_C_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0034976</td>\n",
       "      <td>response to endoplasmic reticulum stress</td>\n",
       "      <td>\"Any process that results in a change in state...</td>\n",
       "      <td>181</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>0.027624</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0033554]</td>\n",
       "      <td>3.995242e-07</td>\n",
       "      <td>29.823629</td>\n",
       "      <td>4.898384</td>\n",
       "      <td>6.398457</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr17_64066984_C_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source      native                                         name  \\\n",
       "0   KEGG  KEGG:04141  Protein processing in endoplasmic reticulum   \n",
       "6  GO:BP  GO:0034976     response to endoplasmic reticulum stress   \n",
       "\n",
       "                                         description  term_size  query_size  \\\n",
       "0        Protein processing in endoplasmic reticulum        138          13   \n",
       "6  \"Any process that results in a change in state...        181          13   \n",
       "\n",
       "   intersection_size  effective_domain_size  precision    recall    query  \\\n",
       "0                  5                  14035   0.384615  0.036232  query_1   \n",
       "6                  5                  14035   0.384615  0.027624  query_1   \n",
       "\n",
       "        parents     nominal_p    obs_exp  log2_enrich  -log10(p)      qval  \\\n",
       "0  [KEGG:00000]  1.032438e-07  39.116499     5.289705   6.986136  0.000067   \n",
       "6  [GO:0033554]  3.995242e-07  29.823629     4.898384   6.398457  0.000260   \n",
       "\n",
       "        tissue    modality  clumped_variant_id  \n",
       "0  Whole_blood  trans-eQTL  chr17_64066984_C_A  \n",
       "6  Whole_blood  trans-eQTL  chr17_64066984_C_A  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[(results.qval<=0.05) & (results.clumped_variant_id=='chr17_64066984_C_A')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.195108Z",
     "iopub.status.busy": "2024-12-03T18:36:04.189081Z",
     "iopub.status.idle": "2024-12-03T18:36:04.197764Z",
     "shell.execute_reply": "2024-12-03T18:36:04.197028Z"
    }
   },
   "outputs": [],
   "source": [
    "tmp = results[(results.qval<=0.05) & (results.clumped_variant_id=='chr17_64066984_C_A') & (results.source=='KEGG')].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.202900Z",
     "iopub.status.busy": "2024-12-03T18:36:04.202352Z",
     "iopub.status.idle": "2024-12-03T18:36:04.205019Z",
     "shell.execute_reply": "2024-12-03T18:36:04.205612Z"
    }
   },
   "outputs": [],
   "source": [
    "assert(tmp.query_size == trans_eqtl_clumped[trans_eqtl_clumped.clumped_variant_id=='chr17_64066984_C_A'].phenotype_id.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.212151Z",
     "iopub.status.busy": "2024-12-03T18:36:04.211365Z",
     "iopub.status.idle": "2024-12-03T18:36:04.214776Z",
     "shell.execute_reply": "2024-12-03T18:36:04.214094Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'For example, one cis-eVariant (chr17_64066984_C_A) for ERN1, which encodes endonuclease IRE1a, was a trans-eVariant for 13 trans-eGenes. ERN1 is a regulator of the endoplasmic reticulum (ER) stress response (Shen et al., 2001; Yoshida et al., 2001), and 5 of the trans-eGenes were in the \"Protein processing in endoplasmic reticulum\" KEGG pathway (39.1-fold enrichment; nominal p=1.0324375176906071e-07)'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'For example, one cis-eVariant (chr17_64066984_C_A) for ERN1, which encodes endonuclease IRE1a, was a trans-eVariant for {} trans-eGenes. ERN1 is a regulator of the endoplasmic reticulum (ER) stress response (Shen et al., 2001; Yoshida et al., 2001), and {} of the trans-eGenes were in the \"{}\" KEGG pathway ({}-fold enrichment; nominal p={})'.format(tmp.query_size, tmp.intersection_size, tmp['name'], round(tmp.obs_exp, 1), tmp.nominal_p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-12-03T18:36:04.221435Z",
     "iopub.status.busy": "2024-12-03T18:36:04.221077Z",
     "iopub.status.idle": "2024-12-03T18:36:04.242507Z",
     "shell.execute_reply": "2024-12-03T18:36:04.241731Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>native</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>term_size</th>\n",
       "      <th>query_size</th>\n",
       "      <th>intersection_size</th>\n",
       "      <th>effective_domain_size</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>query</th>\n",
       "      <th>parents</th>\n",
       "      <th>nominal_p</th>\n",
       "      <th>obs_exp</th>\n",
       "      <th>log2_enrich</th>\n",
       "      <th>-log10(p)</th>\n",
       "      <th>qval</th>\n",
       "      <th>tissue</th>\n",
       "      <th>modality</th>\n",
       "      <th>clumped_variant_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:04613</td>\n",
       "      <td>Neutrophil extracellular trap formation</td>\n",
       "      <td>Neutrophil extracellular trap formation</td>\n",
       "      <td>126</td>\n",
       "      <td>81</td>\n",
       "      <td>13</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.160494</td>\n",
       "      <td>0.103175</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>2.935704e-13</td>\n",
       "      <td>17.877229</td>\n",
       "      <td>4.160051</td>\n",
       "      <td>12.532288</td>\n",
       "      <td>4.879140e-10</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:05034</td>\n",
       "      <td>Alcoholism</td>\n",
       "      <td>Alcoholism</td>\n",
       "      <td>117</td>\n",
       "      <td>81</td>\n",
       "      <td>11</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.135802</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>6.223410e-11</td>\n",
       "      <td>16.290493</td>\n",
       "      <td>4.025958</td>\n",
       "      <td>10.205972</td>\n",
       "      <td>1.034331e-07</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>KEGG:05322</td>\n",
       "      <td>Systemic lupus erythematosus</td>\n",
       "      <td>Systemic lupus erythematosus</td>\n",
       "      <td>80</td>\n",
       "      <td>81</td>\n",
       "      <td>9</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.112500</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[KEGG:00000]</td>\n",
       "      <td>7.496328e-10</td>\n",
       "      <td>19.493056</td>\n",
       "      <td>4.284888</td>\n",
       "      <td>9.125151</td>\n",
       "      <td>1.245890e-06</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0030168</td>\n",
       "      <td>platelet activation</td>\n",
       "      <td>\"A series of progressive, overlapping events t...</td>\n",
       "      <td>105</td>\n",
       "      <td>81</td>\n",
       "      <td>11</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.135802</td>\n",
       "      <td>0.104762</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0001775, GO:0007596]</td>\n",
       "      <td>1.888524e-11</td>\n",
       "      <td>18.152263</td>\n",
       "      <td>4.182078</td>\n",
       "      <td>10.723877</td>\n",
       "      <td>3.138727e-08</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0007596</td>\n",
       "      <td>blood coagulation</td>\n",
       "      <td>\"The sequential process in which the multiple ...</td>\n",
       "      <td>165</td>\n",
       "      <td>81</td>\n",
       "      <td>12</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.148148</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0007599, GO:0042060, GO:0050817]</td>\n",
       "      <td>1.635314e-10</td>\n",
       "      <td>12.601571</td>\n",
       "      <td>3.655532</td>\n",
       "      <td>9.786399</td>\n",
       "      <td>2.717892e-07</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9114</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0014812</td>\n",
       "      <td>muscle cell migration</td>\n",
       "      <td>\"The orderly movement of a muscle cell from on...</td>\n",
       "      <td>60</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.016667</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0016477]</td>\n",
       "      <td>2.939113e-01</td>\n",
       "      <td>2.887860</td>\n",
       "      <td>1.530001</td>\n",
       "      <td>0.531784</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9115</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0014823</td>\n",
       "      <td>response to activity</td>\n",
       "      <td>\"Any process that results in a change in state...</td>\n",
       "      <td>45</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.044444</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0050896]</td>\n",
       "      <td>2.774748e-02</td>\n",
       "      <td>7.700960</td>\n",
       "      <td>2.945038</td>\n",
       "      <td>1.556776</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9116</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0014831</td>\n",
       "      <td>gastro-intestinal system smooth muscle contrac...</td>\n",
       "      <td>\"A process in which force is generated within ...</td>\n",
       "      <td>7</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0006939]</td>\n",
       "      <td>3.971457e-02</td>\n",
       "      <td>24.753086</td>\n",
       "      <td>4.629537</td>\n",
       "      <td>1.401050</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9117</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0014896</td>\n",
       "      <td>muscle hypertrophy</td>\n",
       "      <td>\"The muscle system process that results in enl...</td>\n",
       "      <td>53</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.018868</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0003012]</td>\n",
       "      <td>2.645962e-01</td>\n",
       "      <td>3.269276</td>\n",
       "      <td>1.708971</td>\n",
       "      <td>0.577416</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9118</th>\n",
       "      <td>GO:BP</td>\n",
       "      <td>GO:0014897</td>\n",
       "      <td>striated muscle hypertrophy</td>\n",
       "      <td>\"The enlargement or overgrowth of all or part ...</td>\n",
       "      <td>52</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>14035</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.019231</td>\n",
       "      <td>query_1</td>\n",
       "      <td>[GO:0014896]</td>\n",
       "      <td>2.603114e-01</td>\n",
       "      <td>3.332146</td>\n",
       "      <td>1.736452</td>\n",
       "      <td>0.584507</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>Whole_blood</td>\n",
       "      <td>trans-eQTL</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1662 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     source      native                                               name  \\\n",
       "1      KEGG  KEGG:04613            Neutrophil extracellular trap formation   \n",
       "4      KEGG  KEGG:05034                                         Alcoholism   \n",
       "17     KEGG  KEGG:05322                       Systemic lupus erythematosus   \n",
       "32    GO:BP  GO:0030168                                platelet activation   \n",
       "49    GO:BP  GO:0007596                                  blood coagulation   \n",
       "...     ...         ...                                                ...   \n",
       "9114  GO:BP  GO:0014812                              muscle cell migration   \n",
       "9115  GO:BP  GO:0014823                               response to activity   \n",
       "9116  GO:BP  GO:0014831  gastro-intestinal system smooth muscle contrac...   \n",
       "9117  GO:BP  GO:0014896                                 muscle hypertrophy   \n",
       "9118  GO:BP  GO:0014897                        striated muscle hypertrophy   \n",
       "\n",
       "                                            description  term_size  \\\n",
       "1               Neutrophil extracellular trap formation        126   \n",
       "4                                            Alcoholism        117   \n",
       "17                         Systemic lupus erythematosus         80   \n",
       "32    \"A series of progressive, overlapping events t...        105   \n",
       "49    \"The sequential process in which the multiple ...        165   \n",
       "...                                                 ...        ...   \n",
       "9114  \"The orderly movement of a muscle cell from on...         60   \n",
       "9115  \"Any process that results in a change in state...         45   \n",
       "9116  \"A process in which force is generated within ...          7   \n",
       "9117  \"The muscle system process that results in enl...         53   \n",
       "9118  \"The enlargement or overgrowth of all or part ...         52   \n",
       "\n",
       "      query_size  intersection_size  effective_domain_size  precision  \\\n",
       "1             81                 13                  14035   0.160494   \n",
       "4             81                 11                  14035   0.135802   \n",
       "17            81                  9                  14035   0.111111   \n",
       "32            81                 11                  14035   0.135802   \n",
       "49            81                 12                  14035   0.148148   \n",
       "...          ...                ...                    ...        ...   \n",
       "9114          81                  1                  14035   0.012346   \n",
       "9115          81                  2                  14035   0.024691   \n",
       "9116          81                  1                  14035   0.012346   \n",
       "9117          81                  1                  14035   0.012346   \n",
       "9118          81                  1                  14035   0.012346   \n",
       "\n",
       "        recall    query                               parents     nominal_p  \\\n",
       "1     0.103175  query_1                          [KEGG:00000]  2.935704e-13   \n",
       "4     0.094017  query_1                          [KEGG:00000]  6.223410e-11   \n",
       "17    0.112500  query_1                          [KEGG:00000]  7.496328e-10   \n",
       "32    0.104762  query_1              [GO:0001775, GO:0007596]  1.888524e-11   \n",
       "49    0.072727  query_1  [GO:0007599, GO:0042060, GO:0050817]  1.635314e-10   \n",
       "...        ...      ...                                   ...           ...   \n",
       "9114  0.016667  query_1                          [GO:0016477]  2.939113e-01   \n",
       "9115  0.044444  query_1                          [GO:0050896]  2.774748e-02   \n",
       "9116  0.142857  query_1                          [GO:0006939]  3.971457e-02   \n",
       "9117  0.018868  query_1                          [GO:0003012]  2.645962e-01   \n",
       "9118  0.019231  query_1                          [GO:0014896]  2.603114e-01   \n",
       "\n",
       "        obs_exp  log2_enrich  -log10(p)          qval       tissue  \\\n",
       "1     17.877229     4.160051  12.532288  4.879140e-10  Whole_blood   \n",
       "4     16.290493     4.025958  10.205972  1.034331e-07  Whole_blood   \n",
       "17    19.493056     4.284888   9.125151  1.245890e-06  Whole_blood   \n",
       "32    18.152263     4.182078  10.723877  3.138727e-08  Whole_blood   \n",
       "49    12.601571     3.655532   9.786399  2.717892e-07  Whole_blood   \n",
       "...         ...          ...        ...           ...          ...   \n",
       "9114   2.887860     1.530001   0.531784  1.000000e+00  Whole_blood   \n",
       "9115   7.700960     2.945038   1.556776  1.000000e+00  Whole_blood   \n",
       "9116  24.753086     4.629537   1.401050  1.000000e+00  Whole_blood   \n",
       "9117   3.269276     1.708971   0.577416  1.000000e+00  Whole_blood   \n",
       "9118   3.332146     1.736452   0.584507  1.000000e+00  Whole_blood   \n",
       "\n",
       "        modality clumped_variant_id  \n",
       "1     trans-eQTL  chr3_56815721_T_C  \n",
       "4     trans-eQTL  chr3_56815721_T_C  \n",
       "17    trans-eQTL  chr3_56815721_T_C  \n",
       "32    trans-eQTL  chr3_56815721_T_C  \n",
       "49    trans-eQTL  chr3_56815721_T_C  \n",
       "...          ...                ...  \n",
       "9114  trans-eQTL  chr3_56815721_T_C  \n",
       "9115  trans-eQTL  chr3_56815721_T_C  \n",
       "9116  trans-eQTL  chr3_56815721_T_C  \n",
       "9117  trans-eQTL  chr3_56815721_T_C  \n",
       "9118  trans-eQTL  chr3_56815721_T_C  \n",
       "\n",
       "[1662 rows x 20 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[results.clumped_variant_id=='chr3_56815721_T_C']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2020-09",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
